<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyranges.get_fasta &mdash; pyranges  0.0.124 documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="pyranges.helpers" href="../helpers/index.html" />
    <link rel="prev" title="pyranges.genomicfeatures" href="../genomicfeatures/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            pyranges
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial of PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../how_to_book.html">How-to-book</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../data/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../genomicfeatures/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.genomicfeatures</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.get_fasta</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../helpers/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.helpers</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../multioverlap/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.multioverlap</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../multithreaded/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.multithreaded</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../pyranges/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.pyranges</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../readers/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.readers</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../statistics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.statistics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../stats/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.stats</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../version/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.version</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">pyranges</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.get_fasta</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/autoapi/pyranges/get_fasta/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-pyranges.get_fasta">
<span id="pyranges-get-fasta"></span><h1><a class="reference internal" href="#module-pyranges.get_fasta" title="pyranges.get_fasta"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.get_fasta</span></code></a><a class="headerlink" href="#module-pyranges.get_fasta" title="Permalink to this heading"></a></h1>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.get_fasta.get_sequence" title="pyranges.get_fasta.get_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequence</span></code></a>(gr[, path, pyfaidx_fasta])</p></td>
<td><p>Get the sequence of the intervals from a fasta file</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.get_fasta.get_fasta" title="pyranges.get_fasta.get_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_fasta</span></code></a>(*args, **kwargs)</p></td>
<td><p>Deprecated: this function has been moved to Pyranges.get_sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.get_fasta.get_transcript_sequence" title="pyranges.get_fasta.get_transcript_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_transcript_sequence</span></code></a>(gr, group_by[, path, ...])</p></td>
<td><p>Get the sequence of mRNAs, e.g. joining intervals corresponding to exons of the same transcript</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="pyranges.get_fasta.get_sequence">
<span class="sig-prename descclassname"><span class="pre">pyranges.get_fasta.</span></span><span class="sig-name descname"><span class="pre">get_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pyfaidx_fasta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.get_fasta.get_sequence" title="Permalink to this definition"></a></dt>
<dd><p>Get the sequence of the intervals from a fasta file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gr</strong> (<a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Coordinates.</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to fasta file. It will be indexed using pyfaidx if an index is not found</p></li>
<li><p><strong>pyfaidx_fasta</strong> (<em>pyfaidx.Fasta</em>) – Alternative method to provide fasta target, as a pyfaidx.Fasta object</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequences, one per interval.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Series</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function requires the library pyfaidx, it can be installed with
<code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">bioconda</span> <span class="pre">pyfaidx</span></code> or <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">pyfaidx</span></code>.</p>
<p>Sorting the PyRanges is likely to improve the speed.
Intervals on the negative strand will be reverse complemented.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that the names in the fasta header and gr must be the same.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.get_fasta.get_transcript_sequence" title="pyranges.get_fasta.get_transcript_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_transcript_sequence</span></code></a></dt><dd><p>obtain mRNA sequences, by joining exons belonging to the same transcript</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">]})</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         5 |         8 |</span>
<span class="go">| chr1         |         0 |         5 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;temp.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">tmp_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;chr1</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">tmp_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ATTACCAT</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="s2">&quot;temp.fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">0      CAT</span>
<span class="go">1    ATTAC</span>
<span class="go">dtype: object</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | seq        |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         5 |         8 | CAT        |</span>
<span class="go">| chr1         |         0 |         5 | ATTAC      |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.get_fasta.get_fasta">
<span class="sig-prename descclassname"><span class="pre">pyranges.get_fasta.</span></span><span class="sig-name descname"><span class="pre">get_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/get_fasta.html#get_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.get_fasta.get_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Deprecated: this function has been moved to Pyranges.get_sequence</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.get_fasta.get_transcript_sequence">
<span class="sig-prename descclassname"><span class="pre">pyranges.get_fasta.</span></span><span class="sig-name descname"><span class="pre">get_transcript_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pyfaidx_fasta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.get_fasta.get_transcript_sequence" title="Permalink to this definition"></a></dt>
<dd><p>Get the sequence of mRNAs, e.g. joining intervals corresponding to exons of the same transcript</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gr</strong> (<a class="reference internal" href="../index.html#pyranges.PyRanges" title="pyranges.PyRanges"><em>PyRanges</em></a>) – Coordinates.</p></li>
<li><p><strong>group_by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – intervals are grouped by this/these ID column(s): these are exons belonging to same transcript</p></li>
<li><p><strong>path</strong> (<em>str</em>) – Path to fasta file. It will be indexed using pyfaidx if an index is not found</p></li>
<li><p><strong>pyfaidx_fasta</strong> (<em>pyfaidx.Fasta</em>) – Alternative method to provide fasta target, as a pyfaidx.Fasta object</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pandas DataFrame with a column for Sequence, plus ID column(s) provided with “group_by”</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function requires the library pyfaidx, it can be installed with
<code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">bioconda</span> <span class="pre">pyfaidx</span></code> or <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">pyfaidx</span></code>.</p>
<p>Sorting the PyRanges is likely to improve the speed.
Intervals on the negative strand will be reverse complemented.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that the names in the fasta header and gr must be the same.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.get_fasta.get_sequence" title="pyranges.get_fasta.get_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequence</span></code></a></dt><dd><p>obtain sequence of single intervals</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">18</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">21</span><span class="p">],</span>
<span class="go">                       &quot;Strand&quot;:[&#39;+&#39;, &#39;-&#39;, &#39;-&#39;],</span>
<span class="go">                       &quot;transcript&quot;: [&#39;t1&#39;, &#39;t2&#39;, &#39;t2&#39;]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       | transcript   |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   | (object)     |</span>
<span class="go">|--------------+-----------+-----------+--------------+--------------|</span>
<span class="go">| chr1         |         0 |         4 | +            | t1           |</span>
<span class="go">| chr1         |         9 |        13 | -            | t2           |</span>
<span class="go">| chr1         |        18 |        21 | -            | t2           |</span>
<span class="go">+--------------+-----------+-----------+--------------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;temp.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">tmp_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;chr1</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">tmp_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;AAACCCTTTGGGAAACCCTTTGGG</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tmp_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_transcript_sequence</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="s2">&quot;temp.fasta&quot;</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="s1">&#39;transcript&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
</pre></div>
</div>
<blockquote>
<div><p>transcript Sequence</p>
</div></blockquote>
<p>0         t1     AAAC
1         t2  AAATCCC</p>
<p>To write to a file in fasta format:
&gt;&gt;&gt; with open(‘outfile.fasta’, ‘w’) as fw:
…     nchars=60
…     for row in seq.itertuples():
…         s = ‘n’.join([ row.Sequence[i:i+nchars] for i in range(0, len(row.Sequence), nchars)])
…         fw.write(f’&gt;{row.transcript}n{s}n’)</p>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../genomicfeatures/index.html" class="btn btn-neutral float-left" title="pyranges.genomicfeatures" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../helpers/index.html" class="btn btn-neutral float-right" title="pyranges.helpers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Endre Bakken Stovner and Marco Mariotti.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>