<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyranges.get_fasta &mdash; pyranges  0.0.124 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            pyranges
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Tutorial of PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html">Introduction to PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#loading-creating-pyranges">Loading/Creating PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#writing-pyranges-to-disk">Writing PyRanges to disk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#inspecting-pyranges">Inspecting PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#accessing-data">Accessing data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#operating-with-data">Operating with data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#overlapping-and-matching-pyranges">Overlapping and matching PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#summary-statistics">Summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#computing-with-pyranges">Computing with PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#working-at-the-transcript-level">Working at the transcript level</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#fetching-external-gene-tracks">Fetching external gene tracks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#rles-run-length-encodings">RLEs: run length encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autoapi/index.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">pyranges</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
          <li class="breadcrumb-item"><a href="../pyranges.html">pyranges</a></li>
      <li class="breadcrumb-item active">pyranges.get_fasta</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pyranges.get_fasta</h1><div class="highlight"><pre>
<span></span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">pyranges</span> <span class="k">as</span> <span class="nn">pr</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<div class="viewcode-block" id="get_fasta"><a class="viewcode-back" href="../../autoapi/pyranges/get_fasta/index.html#pyranges.get_fasta.get_fasta">[docs]</a><span class="k">def</span> <span class="nf">get_fasta</span><span class="p">(</span><span class="n">gr</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pyfaidx_fasta</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get fasta sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gr : PyRanges</span>

<span class="sd">        Coordinates.</span>

<span class="sd">    path : str</span>

<span class="sd">        Path to fasta file. It will be indexed using pyfaidx if an index is not found</span>

<span class="sd">    pyfaidx_fasta : pyfaidx.Fasta</span>

<span class="sd">        Alternative method to provide fasta target, as a pyfaidx.Fasta object</span>


<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Series</span>

<span class="sd">        Sequences, one per interval.</span>

<span class="sd">    Note</span>
<span class="sd">    ----</span>

<span class="sd">    Sorting the PyRanges is likely to improve the speed.</span>
<span class="sd">    Intervals on the negative strand will be reverse complemented.</span>

<span class="sd">    Warning</span>
<span class="sd">    -------</span>

<span class="sd">    Note that the names in the fasta header and gr must be the same.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;, &quot;chr1&quot;],</span>
<span class="sd">    ...                    &quot;Start&quot;: [5, 0], &quot;End&quot;: [8, 5]})</span>

<span class="sd">    &gt;&gt;&gt; gr</span>
<span class="sd">    +--------------+-----------+-----------+</span>
<span class="sd">    | Chromosome   |     Start |       End |</span>
<span class="sd">    | (category)   |   (int32) |   (int32) |</span>
<span class="sd">    |--------------+-----------+-----------|</span>
<span class="sd">    | chr1         |         5 |         8 |</span>
<span class="sd">    | chr1         |         0 |         5 |</span>
<span class="sd">    +--------------+-----------+-----------+</span>
<span class="sd">    Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">    For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">    &gt;&gt;&gt; tmp_handle = open(&quot;temp.fasta&quot;, &quot;w+&quot;)</span>
<span class="sd">    &gt;&gt;&gt; _ = tmp_handle.write(&quot;&gt; chr1\\n&quot;)</span>
<span class="sd">    &gt;&gt;&gt; _ = tmp_handle.write(&quot;ATTACCAT&quot;)</span>
<span class="sd">    &gt;&gt;&gt; tmp_handle.close()</span>

<span class="sd">    &gt;&gt;&gt; seq = pr.get_fasta(gr, &quot;temp.fasta&quot;)</span>

<span class="sd">    &gt;&gt;&gt; seq</span>
<span class="sd">    0      CAT</span>
<span class="sd">    1    ATTAC</span>
<span class="sd">    dtype: object</span>

<span class="sd">    &gt;&gt;&gt; gr.seq = seq</span>
<span class="sd">    &gt;&gt;&gt; gr</span>
<span class="sd">    +--------------+-----------+-----------+------------+</span>
<span class="sd">    | Chromosome   |     Start |       End | seq        |</span>
<span class="sd">    | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">    |--------------+-----------+-----------+------------|</span>
<span class="sd">    | chr1         |         5 |         8 | CAT        |</span>
<span class="sd">    | chr1         |         0 |         5 | ATTAC      |</span>
<span class="sd">    +--------------+-----------+-----------+------------+</span>
<span class="sd">    Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">    For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pyfaidx</span>
    <span class="k">except</span> <span class="ne">ModuleNotFoundError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;pyfaidx must be installed to get fasta sequences. Use `conda install -c bioconda pyfaidx` or `pip install pyfaidx` to install it.&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">pyfaidx_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">pyfaidx_fasta</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">read_ahead</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e5</span><span class="p">))</span>

    <span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">gr</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">tuple</span><span class="p">:</span>  <span class="c1">#input is Stranded</span>
            <span class="n">_fasta</span> <span class="o">=</span> <span class="n">pyfaidx_fasta</span><span class="p">[</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="k">if</span> <span class="n">k</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;-&#39;</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">Start</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span><span class="p">):</span>
                    <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="p">(</span><span class="o">-</span><span class="n">_fasta</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">])</span><span class="o">.</span><span class="n">seq</span> <span class="p">)</span> <span class="c1"># reverse complement</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">Start</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span><span class="p">):</span>
                    <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_fasta</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>        
            <span class="n">_fasta</span> <span class="o">=</span> <span class="n">pyfaidx_fasta</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">Start</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span><span class="p">):</span>
                <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_fasta</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">])</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Endre Bakken Stovner and Marco Mariotti.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>