<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyranges.pyranges &mdash; pyranges  0.0.124 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            pyranges
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Tutorial of PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html">Introduction to PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#loading-creating-pyranges">Loading/Creating PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#writing-pyranges-to-disk">Writing PyRanges to disk</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#inspecting-pyranges">Inspecting PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#accessing-data">Accessing data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#operating-with-data">Operating with data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#overlapping-and-matching-pyranges">Overlapping and matching PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#summary-statistics">Summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#computing-with-pyranges">Computing with PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#working-at-the-transcript-level">Working at the transcript level</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#fetching-external-gene-tracks">Fetching external gene tracks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../how_to_book.html#rles-run-length-encodings">RLEs: run length encodings</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../autoapi/index.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">pyranges</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
          <li class="breadcrumb-item"><a href="../pyranges.html">pyranges</a></li>
      <li class="breadcrumb-item active">pyranges.pyranges</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pyranges.pyranges</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Data structure for genomic intervals and their annotation.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">natsort</span> <span class="kn">import</span> <span class="n">natsorted</span>

<span class="kn">import</span> <span class="nn">pyranges</span> <span class="k">as</span> <span class="nn">pr</span>

<span class="kn">from</span> <span class="nn">pyranges.tostring2</span> <span class="kn">import</span> <span class="n">tostring</span>

<span class="kn">from</span> <span class="nn">pyranges.methods.intersection</span> <span class="kn">import</span> <span class="n">_intersection</span><span class="p">,</span> <span class="n">_overlap</span>
<span class="kn">from</span> <span class="nn">pyranges.multithreaded</span> <span class="kn">import</span> <span class="n">pyrange_apply</span><span class="p">,</span> <span class="n">pyrange_apply_single</span><span class="p">,</span> <span class="n">pyrange_apply_chunks</span><span class="p">,</span> <span class="n">_extend</span><span class="p">,</span> <span class="n">_tes</span><span class="p">,</span> <span class="n">_tss</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;PyRanges&quot;</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Give the kwargs dict default options.&quot;&quot;&quot;</span>

    <span class="n">defaults</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;overlap&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;how&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;invert&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;new_pos&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;suffixes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;_a&quot;</span><span class="p">,</span> <span class="s2">&quot;_b&quot;</span><span class="p">],</span>
        <span class="s2">&quot;suffix&quot;</span><span class="p">:</span> <span class="s2">&quot;_b&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sparse&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;other&quot;</span><span class="p">:</span> <span class="kc">False</span>
        <span class="p">}</span>
    <span class="p">}</span>

    <span class="n">defaults</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">defaults</span>


<div class="viewcode-block" id="PyRanges"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges">[docs]</a><span class="k">class</span> <span class="nc">PyRanges</span><span class="p">():</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;Two-dimensional representation of genomic intervals and their annotations.</span>

<span class="sd">    A PyRanges object must have the columns Chromosome, Start and End. These</span>
<span class="sd">    describe the genomic position and function as implicit row labels. A Strand</span>
<span class="sd">    column is optional and adds strand information to the intervals. Any other</span>
<span class="sd">    columns are allowed and are considered metadata.</span>

<span class="sd">    Operations between PyRanges align intervals based on their position.</span>

<span class="sd">    If a PyRanges is built using the arguments chromosomes, starts, ends and</span>
<span class="sd">    optionally strands, all non-scalars must be of the same length.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pandas.DataFrame or dict of pandas.DataFrame, default None</span>
<span class="sd">        The data to be stored in the PyRanges.</span>

<span class="sd">    chromosomes : array-like or scalar value, default None</span>
<span class="sd">        The chromosome(s) in the PyRanges.</span>

<span class="sd">    starts : array-like, default None</span>
<span class="sd">        The start postions in the PyRanges.</span>

<span class="sd">    ends : array-like, default None</span>
<span class="sd">        The end postions in the PyRanges.</span>

<span class="sd">    strands : array-like or scalar value, default None</span>
<span class="sd">        The strands in the PyRanges.</span>

<span class="sd">    int64 : bool, default False</span>
<span class="sd">        Use np.int64 to represent starts and ends</span>

<span class="sd">    copy_df : bool, default True</span>
<span class="sd">        Copy input pandas.DataFrame</span>

<span class="sd">    See Also</span>
<span class="sd">    --------</span>

<span class="sd">    pyranges.read_bed: read bed-file into PyRanges</span>
<span class="sd">    pyranges.read_bam: read bam-file into PyRanges</span>
<span class="sd">    pyranges.read_gff: read gff-file into PyRanges</span>
<span class="sd">    pyranges.read_gtf: read gtf-file into PyRanges</span>
<span class="sd">    pyranges.from_dict: create PyRanges from dict of columns</span>
<span class="sd">    pyranges.from_string: create PyRanges from multiline string</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>

<span class="sd">    A PyRanges object is represented internally as a dictionary efficiency. The keys are</span>
<span class="sd">    chromosomes or chromosome/strand tuples and the values are pandas DataFrames.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; pr.PyRanges()</span>
<span class="sd">    Empty PyRanges</span>

<span class="sd">    &gt;&gt;&gt; pr.PyRanges(chromosomes=&quot;chr1&quot;, starts=(1, 5), ends=[3, 149],</span>
<span class="sd">    ...             strands=(&quot;+&quot;, &quot;-&quot;), int64=True)</span>
<span class="sd">    +--------------+-----------+-----------+--------------+</span>
<span class="sd">    | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">    | (category)   |   (int64) |   (int64) | (category)   |</span>
<span class="sd">    |--------------+-----------+-----------+--------------|</span>
<span class="sd">    | chr1         |         1 |         3 | +            |</span>
<span class="sd">    | chr1         |         5 |       149 | -            |</span>
<span class="sd">    +--------------+-----------+-----------+--------------+</span>
<span class="sd">    Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">    For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">    &gt;&gt;&gt; df = pd.DataFrame({&quot;Chromosome&quot;: [&quot;chr1&quot;, &quot;chr2&quot;], &quot;Start&quot;: [100, 200],</span>
<span class="sd">    ...                    &quot;End&quot;: [150, 201]})</span>
<span class="sd">    &gt;&gt;&gt; df</span>
<span class="sd">      Chromosome  Start  End</span>
<span class="sd">    0       chr1    100  150</span>
<span class="sd">    1       chr2    200  201</span>
<span class="sd">    &gt;&gt;&gt; pr.PyRanges(df)</span>
<span class="sd">    +--------------+-----------+-----------+</span>
<span class="sd">    | Chromosome   |     Start |       End |</span>
<span class="sd">    | (category)   |   (int32) |   (int32) |</span>
<span class="sd">    |--------------+-----------+-----------|</span>
<span class="sd">    | chr1         |       100 |       150 |</span>
<span class="sd">    | chr2         |       200 |       201 |</span>
<span class="sd">    +--------------+-----------+-----------+</span>
<span class="sd">    Unstranded PyRanges object has 2 rows and 3 columns from 2 chromosomes.</span>
<span class="sd">    For printing, the PyRanges was sorted on Chromosome.</span>


<span class="sd">    &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1], &quot;Strand&quot;: [&quot;+&quot;, &quot;-&quot;], &quot;Start&quot;: [1, 4], &quot;End&quot;: [2, 27],</span>
<span class="sd">    ...                    &quot;TP&quot;: [0, 1], &quot;FP&quot;: [12, 11], &quot;TN&quot;: [10, 9], &quot;FN&quot;: [2, 3]})</span>
<span class="sd">    &gt;&gt;&gt; gr</span>
<span class="sd">    +--------------+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">    |   Chromosome | Strand       |     Start |       End |        TP |        FP |        TN |        FN |</span>
<span class="sd">    |   (category) | (category)   |   (int32) |   (int32) |   (int64) |   (int64) |   (int64) |   (int64) |</span>
<span class="sd">    |--------------+--------------+-----------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="sd">    |            1 | +            |         1 |         2 |         0 |        12 |        10 |         2 |</span>
<span class="sd">    |            1 | -            |         4 |        27 |         1 |        11 |         9 |         3 |</span>
<span class="sd">    +--------------+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">    Stranded PyRanges object has 2 rows and 8 columns from 1 chromosomes.</span>
<span class="sd">    For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">dfs</span> <span class="o">=</span> <span class="kc">None</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Dict mapping chromosomes or chromosome/strand pairs to pandas DataFrames.&quot;&quot;&quot;</span>

    <span class="n">features</span> <span class="o">=</span> <span class="kc">None</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Namespace for genomic-features methods.</span>

<span class="sd">    See Also</span>
<span class="sd">    --------</span>
<span class="sd">    pyranges.genomicfeatures : namespace for feature-functionality</span>
<span class="sd">    pyranges.genomicfeatures.GenomicFeaturesMethods : namespace for feature-functionality</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="kc">None</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Namespace for statistcal methods.</span>

<span class="sd">    See Also</span>
<span class="sd">    --------</span>
<span class="sd">    pyranges.statistics : namespace for statistics</span>
<span class="sd">    pyranges.stats.StatisticsMethods : namespace for statistics</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">chromosomes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">starts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">ends</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">strands</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">int64</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">copy_df</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.init</span> <span class="kn">import</span> <span class="n">_init</span>

        <span class="k">if</span> <span class="n">df</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chromosomes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s2">&quot;Chromosome Start End&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>

        <span class="n">_init</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">,</span> <span class="n">starts</span><span class="p">,</span> <span class="n">ends</span><span class="p">,</span> <span class="n">strands</span><span class="p">,</span> <span class="n">int64</span><span class="p">,</span> <span class="n">copy_df</span><span class="p">)</span>


<div class="viewcode-block" id="PyRanges.__array_ufunc__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__array_ufunc__">[docs]</a>    <span class="k">def</span> <span class="nf">__array_ufunc__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply unary numpy-function.</span>


<span class="sd">        Apply function to all columns which are not index, i.e. Chromosome,</span>
<span class="sd">        Start, End nor Strand.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        Function must produce a vector of equal length.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 2, 3], &quot;Start&quot;: [1, 2, 3],</span>
<span class="sd">        ... &quot;End&quot;: [2, 3, 4], &quot;Score&quot;: [9, 16, 25], &quot;Score2&quot;: [121, 144, 169],</span>
<span class="sd">        ... &quot;Name&quot;: [&quot;n1&quot;, &quot;n2&quot;, &quot;n3&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |     Score |    Score2 | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (int64) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         1 |         2 |         9 |       121 | n1         |</span>
<span class="sd">        |            2 |         2 |         3 |        16 |       144 | n2         |</span>
<span class="sd">        |            3 |         3 |         4 |        25 |       169 | n3         |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; np.sqrt(gr)</span>
<span class="sd">        +--------------+-----------+-----------+-------------+-------------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |       Score |      Score2 | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (float64) |   (float64) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+-------------+-------------+------------|</span>
<span class="sd">        |            1 |         1 |         2 |           3 |          11 | n1         |</span>
<span class="sd">        |            2 |         2 |         3 |           4 |          12 | n2         |</span>
<span class="sd">        |            3 |         3 |         4 |           5 |          13 | n3         |</span>
<span class="sd">        +--------------+-----------+-----------+-------------+-------------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">func</span><span class="p">,</span> <span class="n">call</span><span class="p">,</span> <span class="n">gr</span> <span class="o">=</span> <span class="n">args</span>

        <span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">non_index</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;End&quot;</span><span class="p">,</span> <span class="s2">&quot;Strand&quot;</span><span class="p">]]</span>

        <span class="k">for</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">gr</span><span class="p">:</span>
            <span class="n">subset</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="n">non_index</span><span class="p">]</span><span class="o">.</span><span class="n">select_dtypes</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">number</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span>
            <span class="n">_v</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="n">call</span><span class="p">)(</span><span class="n">df</span><span class="p">[</span><span class="n">subset</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
            <span class="c1"># print(_v)</span>
            <span class="c1"># print(df[_c])</span>
            <span class="n">df</span><span class="p">[</span><span class="n">subset</span><span class="p">]</span> <span class="o">=</span> <span class="n">_v</span>


        <span class="k">return</span> <span class="n">gr</span></div>


        <span class="c1"># self.apply()</span>

<div class="viewcode-block" id="PyRanges.__getattr__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__getattr__">[docs]</a>    <span class="k">def</span> <span class="fm">__getattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return column.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        name : str</span>

<span class="sd">            Column to return</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.Series</span>

<span class="sd">        Example</span>
<span class="sd">        -------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 1], &quot;Start&quot;: [0, 100, 250], &quot;End&quot;: [10, 125, 251]})</span>
<span class="sd">        &gt;&gt;&gt; gr.Start</span>
<span class="sd">        0      0</span>
<span class="sd">        1    100</span>
<span class="sd">        2    250</span>
<span class="sd">        Name: Start, dtype: int32</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.attr</span> <span class="kn">import</span> <span class="n">_getattr</span>

        <span class="k">return</span> <span class="n">_getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.__setattr__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__setattr__">[docs]</a>    <span class="k">def</span> <span class="fm">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">column_name</span><span class="p">,</span> <span class="n">column</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Insert or update column.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        column_name : str</span>

<span class="sd">            Name of column to update or insert.</span>

<span class="sd">        column : list, np.array or pd.Series</span>

<span class="sd">            Data to insert.</span>

<span class="sd">        Example</span>
<span class="sd">        -------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 1], &quot;Start&quot;: [0, 100, 250], &quot;End&quot;: [10, 125, 251]})</span>
<span class="sd">        &gt;&gt;&gt; gr.Start = np.array([1, 1, 2])</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int64) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |        10 |</span>
<span class="sd">        |            1 |         1 |       125 |</span>
<span class="sd">        |            1 |         2 |       251 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.attr</span> <span class="kn">import</span> <span class="n">_setattr</span>

        <span class="k">if</span> <span class="n">column_name</span> <span class="o">==</span> <span class="s2">&quot;columns&quot;</span><span class="p">:</span>
            <span class="n">dfs</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
                <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">column</span>
                <span class="n">dfs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span>
            <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s2">&quot;dfs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dfs</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">_setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">column_name</span><span class="p">,</span> <span class="n">column</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">column_name</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;End&quot;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">dtypes</span><span class="p">[</span><span class="s2">&quot;Start&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dtypes</span><span class="p">[</span><span class="s2">&quot;End&quot;</span><span class="p">]:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning! Start and End columns now have different dtypes: </span><span class="si">{}</span><span class="s2"> and </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">dtypes</span><span class="p">[</span><span class="s2">&quot;Start&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">dtypes</span><span class="p">[</span><span class="s2">&quot;End&quot;</span><span class="p">]))</span></div>

<div class="viewcode-block" id="PyRanges.__getitem__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__getitem__">[docs]</a>    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">val</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Fetch columns or subset on position.</span>

<span class="sd">        If a list is provided, the column(s) in the list is returned. This subsets on columns.</span>

<span class="sd">        If a numpy array is provided, it must be of type bool and the same length as the PyRanges.</span>

<span class="sd">        Otherwise, a subset of the rows is returned with the location info provided.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        val : bool array/Series, tuple, list, str or slice</span>

<span class="sd">            Data to fetch.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.ensembl_gtf()</span>
<span class="sd">        &gt;&gt;&gt; gr.columns</span>
<span class="sd">        Index([&#39;Chromosome&#39;, &#39;Source&#39;, &#39;Feature&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Score&#39;, &#39;Strand&#39;,</span>
<span class="sd">            &#39;Frame&#39;, &#39;gene_biotype&#39;, &#39;gene_id&#39;, &#39;gene_name&#39;, &#39;gene_source&#39;,</span>
<span class="sd">            &#39;gene_version&#39;, &#39;tag&#39;, &#39;transcript_biotype&#39;, &#39;transcript_id&#39;,</span>
<span class="sd">            &#39;transcript_name&#39;, &#39;transcript_source&#39;, &#39;transcript_support_level&#39;,</span>
<span class="sd">            &#39;transcript_version&#39;, &#39;exon_id&#39;, &#39;exon_number&#39;, &#39;exon_version&#39;,</span>
<span class="sd">            &#39;(assigned&#39;, &#39;previous&#39;, &#39;protein_id&#39;, &#39;protein_version&#39;, &#39;ccds_id&#39;],</span>
<span class="sd">            dtype=&#39;object&#39;)</span>

<span class="sd">        &gt;&gt;&gt; gr = gr[[&quot;Source&quot;, &quot;Feature&quot;, &quot;gene_id&quot;]]</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="sd">        | Chromosome   | Source     | Feature      | Start     | End       | Strand       | gene_id         |</span>
<span class="sd">        | (category)   | (object)   | (category)   | (int32)   | (int32)   | (category)   | (object)        |</span>
<span class="sd">        |--------------+------------+--------------+-----------+-----------+--------------+-----------------|</span>
<span class="sd">        | 1            | havana     | gene         | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | havana     | transcript   | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | havana     | exon         | 11868     | 12227     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | havana     | exon         | 12612     | 12721     | +            | ENSG00000223972 |</span>
<span class="sd">        | ...          | ...        | ...          | ...       | ...       | ...          | ...             |</span>
<span class="sd">        | 1            | havana     | gene         | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        | 1            | havana     | transcript   | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        | 1            | havana     | exon         | 1179364   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        | 1            | havana     | exon         | 1173055   | 1176396   | -            | ENSG00000205231 |</span>
<span class="sd">        +--------------+------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 2,446 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        Create boolean Series and use it to subset:</span>

<span class="sd">        &gt;&gt;&gt; s = (gr.Feature == &quot;gene&quot;) | (gr.gene_id == &quot;ENSG00000223972&quot;)</span>
<span class="sd">        &gt;&gt;&gt; gr[s]</span>
<span class="sd">        +--------------+----------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="sd">        | Chromosome   | Source         | Feature      | Start     | End       | Strand       | gene_id         |</span>
<span class="sd">        | (category)   | (object)       | (category)   | (int32)   | (int32)   | (category)   | (object)        |</span>
<span class="sd">        |--------------+----------------+--------------+-----------+-----------+--------------+-----------------|</span>
<span class="sd">        | 1            | havana         | gene         | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | havana         | transcript   | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | havana         | exon         | 11868     | 12227     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | havana         | exon         | 12612     | 12721     | +            | ENSG00000223972 |</span>
<span class="sd">        | ...          | ...            | ...          | ...       | ...       | ...          | ...             |</span>
<span class="sd">        | 1            | havana         | gene         | 1062207   | 1063288   | -            | ENSG00000273443 |</span>
<span class="sd">        | 1            | ensembl_havana | gene         | 1070966   | 1074306   | -            | ENSG00000237330 |</span>
<span class="sd">        | 1            | ensembl_havana | gene         | 1081817   | 1116361   | -            | ENSG00000131591 |</span>
<span class="sd">        | 1            | havana         | gene         | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        +--------------+----------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 95 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; cs = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; cs[10000:100000]</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr2         |     33241 |     33266 | U0         |         0 | +            |</span>
<span class="sd">        | chr2         |     13611 |     13636 | U0         |         0 | -            |</span>
<span class="sd">        | chr2         |     32620 |     32645 | U0         |         0 | -            |</span>
<span class="sd">        | chr3         |     87179 |     87204 | U0         |         0 | +            |</span>
<span class="sd">        | chr4         |     45413 |     45438 | U0         |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 6 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; cs[&quot;chr1&quot;, &quot;-&quot;]</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 100079649 | 100079674 | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 223587418 | 223587443 | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 202450161 | 202450186 | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 156338310 | 156338335 | U0         | 0         | -            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chr1         | 203557775 | 203557800 | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 28114107  | 28114132  | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 21622765  | 21622790  | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 80668132  | 80668157  | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 437 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; cs[&quot;chr5&quot;, &quot;-&quot;, 90000:]</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr5         | 399682    | 399707    | U0         | 0         | -            |</span>
<span class="sd">        | chr5         | 1847502   | 1847527   | U0         | 0         | -            |</span>
<span class="sd">        | chr5         | 5247533   | 5247558   | U0         | 0         | -            |</span>
<span class="sd">        | chr5         | 5300394   | 5300419   | U0         | 0         | -            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chr5         | 178786234 | 178786259 | U0         | 0         | -            |</span>
<span class="sd">        | chr5         | 179268931 | 179268956 | U0         | 0         | -            |</span>
<span class="sd">        | chr5         | 179289594 | 179289619 | U0         | 0         | -            |</span>
<span class="sd">        | chr5         | 180513795 | 180513820 | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 285 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; cs[&quot;chrM&quot;]</span>
<span class="sd">        Empty PyRanges</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.getitem</span> <span class="kn">import</span> <span class="n">_getitem</span>

        <span class="k">return</span> <span class="n">_getitem</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">val</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.__iter__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__iter__">[docs]</a>    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Iterate over the keys and values.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        pyranges.iter : iterate over multiple PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 1], &quot;Start&quot;: [0, 100, 250],</span>
<span class="sd">        ...                   &quot;End&quot;: [10, 125, 251], &quot;Strand&quot;: [&quot;+&quot;, &quot;+&quot;, &quot;-&quot;]})</span>

<span class="sd">        &gt;&gt;&gt; for k, v in gr:</span>
<span class="sd">        ...     print(k)</span>
<span class="sd">        ...     print(v)</span>
<span class="sd">        (&#39;1&#39;, &#39;+&#39;)</span>
<span class="sd">        Chromosome  Start  End Strand</span>
<span class="sd">        0          1      0   10      +</span>
<span class="sd">        1          1    100  125      +</span>
<span class="sd">        (&#39;1&#39;, &#39;-&#39;)</span>
<span class="sd">        Chromosome  Start  End Strand</span>
<span class="sd">        2          1    250  251      -</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">())</span></div>


<div class="viewcode-block" id="PyRanges.__len__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__len__">[docs]</a>    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the number of intervals in the PyRanges.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span></div>


<div class="viewcode-block" id="PyRanges.__str__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__str__">[docs]</a>    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return string representation.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">tostring</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.__repr__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__repr__">[docs]</a>    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return REPL representation.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges._repr_html_"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges._repr_html_">[docs]</a>    <span class="k">def</span> <span class="nf">_repr_html_</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return REPL HTML representation for Jupyter Noteboooks.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">_repr_html_</span><span class="p">()</span></div>

<div class="viewcode-block" id="PyRanges.apply"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.apply">[docs]</a>    <span class="k">def</span> <span class="nf">apply</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply a function to the PyRanges.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        f : function</span>
<span class="sd">            Function to apply on each DataFrame in a PyRanges</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        as_pyranges : bool, default True</span>

<span class="sd">            Whether to return as a PyRanges or dict. If `f` does not return a DataFrame valid for</span>
<span class="sd">            PyRanges, `as_pyranges` must be False.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges or dict</span>
<span class="sd">            Result of applying f to each DataFrame in the PyRanges</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.PyRanges.apply_pair: apply a function to a pair of PyRanges</span>
<span class="sd">        pyranges.PyRanges.apply_chunks: apply a row-based function to a PyRanges in parallel</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        This is the function used internally to carry out almost all unary PyRanges methods.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 2, 2], &quot;Strand&quot;: [&quot;+&quot;, &quot;+&quot;, &quot;-&quot;, &quot;+&quot;],</span>
<span class="sd">        ...                    &quot;Start&quot;: [1, 4, 2, 9], &quot;End&quot;: [2, 27, 13, 10]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+--------------+-----------+-----------|</span>
<span class="sd">        |            1 | +            |         1 |         2 |</span>
<span class="sd">        |            1 | +            |         4 |        27 |</span>
<span class="sd">        |            2 | +            |         9 |        10 |</span>
<span class="sd">        |            2 | -            |         2 |        13 |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 4 columns from 2 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.apply(lambda df: len(df), as_pyranges=False)</span>
<span class="sd">        {(&#39;1&#39;, &#39;+&#39;): 2, (&#39;2&#39;, &#39;+&#39;): 1, (&#39;2&#39;, &#39;-&#39;): 1}</span>

<span class="sd">        &gt;&gt;&gt; gr.apply(lambda df: len(df), as_pyranges=False, strand=False)</span>
<span class="sd">        {&#39;1&#39;: 2, &#39;2&#39;: 2}</span>

<span class="sd">        &gt;&gt;&gt; def add_to_ends(df, **kwargs):</span>
<span class="sd">        ...     df.loc[:, &quot;End&quot;] = kwargs[&quot;slack&quot;] + df.End</span>
<span class="sd">        ...     return df</span>
<span class="sd">        &gt;&gt;&gt; gr.apply(add_to_ends, slack=500)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+--------------+-----------+-----------|</span>
<span class="sd">        |            1 | +            |         1 |       502 |</span>
<span class="sd">        |            1 | +            |         4 |       527 |</span>
<span class="sd">        |            2 | +            |         9 |       510 |</span>
<span class="sd">        |            2 | -            |         2 |       513 |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 4 columns from 2 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">})</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;kwargs&quot;</span><span class="p">,</span> <span class="p">{}))</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">as_pyranges</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.apply_chunks"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.apply_chunks">[docs]</a>    <span class="k">def</span> <span class="nf">apply_chunks</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply a row-based function to arbitrary partitions of the PyRanges.</span>

<span class="sd">        apply_chunks speeds up the application of functions where the result is not affected by</span>
<span class="sd">        applying the function to ordered, non-overlapping splits of the data.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        f : function</span>
<span class="sd">            Row-based or associative function to apply on the partitions.</span>

<span class="sd">        as_pyranges : bool, default False</span>

<span class="sd">            Whether to return as a PyRanges or dict.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. The data is split into nb_cpu partitions.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict of lists</span>
<span class="sd">            Result of applying f to each partition of the DataFrames in the PyRanges.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.PyRanges.apply_pair: apply a function to a pair of PyRanges</span>
<span class="sd">        pyranges.PyRanges.apply_chunks: apply a row-based function to a PyRanges in parallel</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        apply_chunks will only lead to speedups on large datasets or slow-running functions. Using</span>
<span class="sd">        it with nb_cpu=1 is pointless; use apply instead.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 1], &quot;Start&quot;: [2, 3, 5], &quot;End&quot;: [9, 4, 6]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |         2 |         9 |</span>
<span class="sd">        |            1 |         3 |         4 |</span>
<span class="sd">        |            1 |         5 |         6 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.apply_chunks(</span>
<span class="sd">        ... lambda df, **kwargs: list(df.End + kwargs[&quot;add&quot;]), nb_cpu=1, add=1000)</span>
<span class="sd">        {&#39;1&#39;: [[1009, 1004, 1006]]}</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;kwargs&quot;</span><span class="p">,</span> <span class="p">{}))</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_chunks</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">as_pyranges</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="PyRanges.apply_pair"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.apply_pair">[docs]</a>    <span class="k">def</span> <span class="nf">apply_pair</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                   <span class="n">other</span><span class="p">,</span>
                   <span class="n">f</span><span class="p">,</span>
                   <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                   <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Apply a function to a pair of PyRanges.</span>

<span class="sd">        The function is applied to each chromosome or chromosome/strand pair found in at least one</span>
<span class="sd">        of the PyRanges.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        f : function</span>
<span class="sd">            Row-based or associative function to apply on the DataFrames.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        as_pyranges : bool, default False</span>

<span class="sd">            Whether to return as a PyRanges or dict. If `f` does not return a DataFrame valid for</span>
<span class="sd">            PyRanges, `as_pyranges` must be False.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict of lists</span>
<span class="sd">            Result of applying f to each partition of the DataFrames in the PyRanges.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.PyRanges.apply_pair: apply a function to a pair of PyRanges</span>
<span class="sd">        pyranges.PyRanges.apply_chunks: apply a row-based function to a PyRanges in parallel</span>
<span class="sd">        pyranges.iter: iterate over two or more PyRanges</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        This is the function used internally to carry out almost all comparison functions in</span>
<span class="sd">        PyRanges.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; gr2 = pr.data.chipseq_background()</span>

<span class="sd">        &gt;&gt;&gt; gr.apply_pair(gr2, pr.methods.intersection._intersection) # same as gr.intersect(gr2)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 226987603 | 226987617 | U0         |         0 | +            |</span>
<span class="sd">        | chr8         |  38747236 |  38747251 | U0         |         0 | -            |</span>
<span class="sd">        | chr15        |  26105515 |  26105518 | U0         |         0 | +            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1 = pr.data.f1()</span>
<span class="sd">        &gt;&gt;&gt; f1</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f2 = pr.data.f2()</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         2 | a          |         0 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | b          |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.apply_pair(f2, lambda df, df2: (len(df), len(df2)), as_pyranges=False)</span>
<span class="sd">        {(&#39;chr1&#39;, &#39;+&#39;): (2, 2), (&#39;chr1&#39;, &#39;-&#39;): (1, 2)}</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">})</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;kwargs&quot;</span><span class="p">,</span> <span class="p">{}))</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">as_pyranges</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.as_df"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.as_df">[docs]</a>    <span class="k">def</span> <span class="nf">as_df</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return PyRanges as DataFrame.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        DataFrame</span>

<span class="sd">            A DataFrame natural sorted on Chromosome and Strand. The ordering of rows within</span>
<span class="sd">            chromosomes and strands is preserved.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.df : Return PyRanges as DataFrame.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 2, 2], &quot;Start&quot;: [1, 2, 3, 9],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 3, 10, 12], &quot;Gene&quot;: [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;, &quot;D&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Gene       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         1 |         3 | A          |</span>
<span class="sd">        |            1 |         2 |         3 | B          |</span>
<span class="sd">        |            2 |         3 |        10 | C          |</span>
<span class="sd">        |            2 |         9 |        12 | D          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 4 columns from 2 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.as_df()</span>
<span class="sd">          Chromosome  Start  End Gene</span>
<span class="sd">        0          1      1    3    A</span>
<span class="sd">        1          1      2    3    B</span>
<span class="sd">        2          2      3   10    C</span>
<span class="sd">        3          2      9   12    D</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.assign"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.assign">[docs]</a>    <span class="k">def</span> <span class="nf">assign</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add or replace a column.</span>

<span class="sd">        Does not change the original PyRanges.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        col : str</span>

<span class="sd">            Name of column.</span>

<span class="sd">        f : function</span>
<span class="sd">            Function to create new column.</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>
<span class="sd">            A copy of the PyRanges with the column inserted.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1], &quot;Start&quot;: [1, 2], &quot;End&quot;: [3, 5],</span>
<span class="sd">        ... &quot;Name&quot;: [&quot;a&quot;, &quot;b&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         1 |         3 | a          |</span>
<span class="sd">        |            1 |         2 |         5 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.assign(&quot;Blabla&quot;, lambda df: df.Chromosome.astype(str) + &quot;_yadayada&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+------------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Name       | Blabla     |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+------------|</span>
<span class="sd">        |            1 |         1 |         3 | a          | 1_yadayada |</span>
<span class="sd">        |            1 |         2 |         5 | b          | 1_yadayada |</span>
<span class="sd">        +--------------+-----------+-----------+------------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        Note that assigning to an existing name replaces the column:</span>

<span class="sd">        &gt;&gt;&gt; gr.assign(&quot;Name&quot;,</span>
<span class="sd">        ... lambda df, **kwargs: df.Start.astype(str) + kwargs[&quot;sep&quot;] +</span>
<span class="sd">        ... df.Name.str.capitalize(), sep=&quot;_&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         1 |         3 | 1_A        |</span>
<span class="sd">        |            1 |         2 |         5 | 2_B        |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">strand</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">first_result</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
            <span class="n">first_result</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span>
        <span class="p">),</span> <span class="s2">&quot;result of assign function must be Series, but is </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="nb">type</span><span class="p">(</span><span class="n">first_result</span><span class="p">))</span>

        <span class="c1"># do a deepcopy of object</span>
        <span class="n">new_self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">new_self</span><span class="o">.</span><span class="fm">__setattr__</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">result</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_self</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">chromosomes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return chromosomes in natsorted order.&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">natsorted</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">k</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="p">()]))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">natsorted</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="p">()]))</span>

<div class="viewcode-block" id="PyRanges.cluster"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.cluster">[docs]</a>    <span class="k">def</span> <span class="nf">cluster</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">slack</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Give overlapping intervals a common id.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to ignore strand information if PyRanges is stranded.</span>

<span class="sd">        by : str or list, default None</span>

<span class="sd">            Only intervals with an equal value in column(s) `by` are clustered.</span>

<span class="sd">        slack : int, default 0</span>

<span class="sd">            Consider intervals separated by less than `slack` to be in the same cluster. If `slack`</span>
<span class="sd">            is negative, intervals overlapping less than `slack` are not considered to be in the</span>
<span class="sd">            same cluster.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>
<span class="sd">            PyRanges with an ID-column &quot;Cluster&quot; added.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.merge: combine overlapping intervals into one</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 1, 1], &quot;Start&quot;: [1, 2, 3, 9],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 3, 10, 12], &quot;Gene&quot;: [1, 2, 3, 3]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |      Gene |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |         3 |         1 |</span>
<span class="sd">        |            1 |         2 |         3 |         2 |</span>
<span class="sd">        |            1 |         3 |        10 |         3 |</span>
<span class="sd">        |            1 |         9 |        12 |         3 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.cluster()</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |      Gene |   Cluster |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |         3 |         1 |         1 |</span>
<span class="sd">        |            1 |         2 |         3 |         2 |         1 |</span>
<span class="sd">        |            1 |         3 |        10 |         3 |         1 |</span>
<span class="sd">        |            1 |         9 |        12 |         3 |         1 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.cluster(by=&quot;Gene&quot;, count=True)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |      Gene |   Cluster |     Count |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (int32) |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |         3 |         1 |         1 |         1 |</span>
<span class="sd">        |            1 |         2 |         3 |         2 |         2 |         1 |</span>
<span class="sd">        |            1 |         3 |        10 |         3 |         3 |         2 |</span>
<span class="sd">        |            1 |         9 |        12 |         3 |         3 |         2 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        Avoid clustering bookended intervals with slack=-1:</span>

<span class="sd">        &gt;&gt;&gt; gr.cluster(slack=-1)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |      Gene |   Cluster |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |         3 |         1 |         1 |</span>
<span class="sd">        |            1 |         2 |         3 |         2 |         1 |</span>
<span class="sd">        |            1 |         3 |        10 |         3 |         2 |</span>
<span class="sd">        |            1 |         9 |        12 |         3 |         2 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2 = pr.data.ensembl_gtf()[[&quot;Feature&quot;, &quot;Source&quot;]]</span>
<span class="sd">        &gt;&gt;&gt; gr2.cluster(by=[&quot;Feature&quot;, &quot;Source&quot;])</span>
<span class="sd">        +--------------+--------------+---------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   | Feature      | Source        | Start     | End       | Strand       | Cluster   |</span>
<span class="sd">        | (category)   | (category)   | (object)      | (int32)   | (int32)   | (category)   | (int32)   |</span>
<span class="sd">        |--------------+--------------+---------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        | 1            | CDS          | ensembl       | 69090     | 70005     | +            | 1         |</span>
<span class="sd">        | 1            | CDS          | ensembl       | 925941    | 926013    | +            | 2         |</span>
<span class="sd">        | 1            | CDS          | ensembl       | 925941    | 926013    | +            | 2         |</span>
<span class="sd">        | 1            | CDS          | ensembl       | 925941    | 926013    | +            | 2         |</span>
<span class="sd">        | ...          | ...          | ...           | ...       | ...       | ...          | ...       |</span>
<span class="sd">        | 1            | transcript   | havana_tagene | 167128    | 169240    | -            | 1142      |</span>
<span class="sd">        | 1            | transcript   | mirbase       | 17368     | 17436     | -            | 1143      |</span>
<span class="sd">        | 1            | transcript   | mirbase       | 187890    | 187958    | -            | 1144      |</span>
<span class="sd">        | 1            | transcript   | mirbase       | 632324    | 632413    | -            | 1145      |</span>
<span class="sd">        +--------------+--------------+---------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 2,446 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;slack&quot;</span><span class="p">:</span> <span class="n">slack</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">:</span> <span class="n">count</span><span class="p">,</span> <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="n">by</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">_stranded</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span> <span class="ow">and</span> <span class="n">_stranded</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">Strand2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Strand</span>
            <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unstrand</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">by</span><span class="p">:</span>
            <span class="kn">from</span> <span class="nn">pyranges.methods.cluster</span> <span class="kn">import</span> <span class="n">_cluster</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_cluster</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="kn">from</span> <span class="nn">pyranges.methods.cluster</span> <span class="kn">import</span> <span class="n">_cluster_by</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;by&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">by</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_cluster_by</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">gr</span> <span class="o">=</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># each chromosome got overlapping ids (0 to len). Need to make unique!</span>
        <span class="n">new_dfs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">first</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">max_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gr</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">first</span><span class="p">:</span>
                <span class="n">max_id</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">Cluster</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
                <span class="n">new_dfs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>
                <span class="n">first</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">continue</span>

            <span class="n">v</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;Cluster&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">max_id</span>
            <span class="n">max_id</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">Cluster</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
            <span class="n">new_dfs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span> <span class="ow">and</span> <span class="n">_stranded</span><span class="p">:</span>
            <span class="n">new_dfs</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span><span class="p">:</span> <span class="n">d</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Strand2&quot;</span><span class="p">:</span> <span class="s2">&quot;Strand&quot;</span><span class="p">})</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">new_dfs</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">new_dfs</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="PyRanges.copy"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.copy">[docs]</a>    <span class="k">def</span> <span class="nf">copy</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Make a deep copy of the PyRanges.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        See the pandas docs for deep-copying caveats.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">columns</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the column labels of the PyRanges.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pandas.Index</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.chromosomes : return the chromosomes in the PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; f2 = pr.data.f2()</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         2 | a          |         0 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | b          |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f2.columns</span>
<span class="sd">        Index([&#39;Chromosome&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Name&#39;, &#39;Score&#39;, &#39;Strand&#39;], dtype=&#39;object&#39;)</span>

<span class="sd">        &gt;&gt;&gt; f2.columns = f2.columns.str.replace(&quot;Sco|re&quot;, &quot;NYAN&quot;)</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+------------+------------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |   NYANNYAN | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |    (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+------------+--------------|</span>
<span class="sd">        | chr1         |         1 |         2 | a          |          0 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | b          |          0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+------------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
            <span class="k">return</span> <span class="p">[]</span>

        <span class="n">first</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="n">first</span><span class="o">.</span><span class="n">columns</span>

        <span class="k">return</span> <span class="n">columns</span>

<div class="viewcode-block" id="PyRanges.count_overlaps"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.count_overlaps">[docs]</a>    <span class="k">def</span> <span class="nf">count_overlaps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_nonoverlapping</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">overlap_col</span><span class="o">=</span><span class="s2">&quot;NumberOverlaps&quot;</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Count number of overlaps per interval.</span>

<span class="sd">        Count how many intervals in self overlap with those in other.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strandedness : {&quot;same&quot;, &quot;opposite&quot;, None, False}, default None, i.e. auto</span>

<span class="sd">            Whether to perform the operation on the same, opposite or no strand. Use False to</span>
<span class="sd">            ignore the strand. None means use &quot;same&quot; if both PyRanges are stranded, otherwise</span>
<span class="sd">            ignore.</span>

<span class="sd">        keep_nonoverlapping : bool, default True</span>

<span class="sd">            Keep intervals without overlaps.</span>

<span class="sd">        overlap_col : str, default &quot;NumberOverlaps&quot;</span>

<span class="sd">            Name of column with overlap counts.</span>

<span class="sd">        nb_cpu : int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>
<span class="sd">            PyRanges with a column of overlaps added.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.coverage: find coverage of PyRanges</span>
<span class="sd">        pyranges.count_overlaps: count overlaps from multiple PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; f1 = pr.data.f1().drop()</span>
<span class="sd">        &gt;&gt;&gt; f1</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &gt;&gt;&gt; f2 = pr.data.f2().drop()</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         2 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.count_overlaps(f2, overlap_col=&quot;Count&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |     Count |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         0 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         0 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         1 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;keep_nonoverlapping&quot;</span><span class="p">:</span> <span class="n">keep_nonoverlapping</span><span class="p">,</span>
                  <span class="s2">&quot;overlap_col&quot;</span><span class="p">:</span> <span class="n">overlap_col</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.coverage</span> <span class="kn">import</span> <span class="n">_number_overlapping</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_number_overlapping</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.coverage"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.coverage">[docs]</a>    <span class="k">def</span> <span class="nf">coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_nonoverlapping</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">overlap_col</span><span class="o">=</span><span class="s2">&quot;NumberOverlaps&quot;</span><span class="p">,</span> <span class="n">fraction_col</span><span class="o">=</span><span class="s2">&quot;FractionOverlaps&quot;</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Count number of overlaps and their fraction per interval.</span>

<span class="sd">        Count how many intervals in self overlap with those in other.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strandedness : {&quot;same&quot;, &quot;opposite&quot;, None, False}, default None, i.e. auto</span>

<span class="sd">            Whether to perform the operation on the same, opposite or no strand. Use False to</span>
<span class="sd">            ignore the strand. None means use &quot;same&quot; if both PyRanges are stranded, otherwise</span>
<span class="sd">            ignore.</span>

<span class="sd">        keep_nonoverlapping : bool, default True</span>

<span class="sd">            Keep intervals without overlaps.</span>

<span class="sd">        overlap_col : str, default &quot;NumberOverlaps&quot;</span>

<span class="sd">            Name of column with overlap counts.</span>

<span class="sd">        fraction_col : str, default &quot;FractionOverlaps&quot;</span>

<span class="sd">            Name of column with fraction of counts.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>
<span class="sd">            PyRanges with a column of overlaps added.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.count_overlaps: count overlaps from multiple PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; f1 = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 1], &quot;Start&quot;: [3, 8, 5],</span>
<span class="sd">        ...                    &quot;End&quot;: [6,  9, 7]})</span>
<span class="sd">        &gt;&gt;&gt; f1</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |         3 |         6 |</span>
<span class="sd">        |            1 |         8 |         9 |</span>
<span class="sd">        |            1 |         5 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &gt;&gt;&gt; f2 = pr.from_dict({&quot;Chromosome&quot;: [1, 1], &quot;Start&quot;: [1, 6],</span>
<span class="sd">        ...                    &quot;End&quot;: [2, 7]})</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |         2 |</span>
<span class="sd">        |            1 |         6 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; f1.coverage(f2, overlap_col=&quot;C&quot;, fraction_col=&quot;F&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |         C |           F |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (float64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-------------|</span>
<span class="sd">        |            1 |         3 |         6 |         0 |         0   |</span>
<span class="sd">        |            1 |         8 |         9 |         0 |         0   |</span>
<span class="sd">        |            1 |         5 |         7 |         1 |         0.5 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;keep_nonoverlapping&quot;</span><span class="p">:</span> <span class="n">keep_nonoverlapping</span><span class="p">,</span>
                  <span class="s2">&quot;overlap_col&quot;</span><span class="p">:</span> <span class="n">overlap_col</span><span class="p">,</span> <span class="s2">&quot;fraction_col&quot;</span><span class="p">:</span> <span class="n">fraction_col</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_overlaps</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">keep_nonoverlapping</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">overlap_col</span><span class="o">=</span><span class="n">overlap_col</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="n">strandedness</span><span class="p">)</span>

        <span class="n">strand</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;strandedness&quot;</span><span class="p">]</span> <span class="k">else</span> <span class="kc">False</span>
        <span class="n">other</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.coverage</span> <span class="kn">import</span> <span class="n">_coverage</span>

        <span class="n">counts</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_coverage</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">counts</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">df</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return PyRanges as DataFrame.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.as_df : return PyRanges as DataFrame.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">as_df</span><span class="p">()</span>

<div class="viewcode-block" id="PyRanges.drop"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.drop">[docs]</a>    <span class="k">def</span> <span class="nf">drop</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">drop</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">like</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Drop column(s).</span>

<span class="sd">        If no arguments are given, all the columns except Chromosome, Start, End and Strand are</span>
<span class="sd">        dropped.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        drop : str or list, default None</span>

<span class="sd">            Columns to drop.</span>

<span class="sd">        like : str, default None</span>

<span class="sd">            Regex-string matching columns to drop. Matches with Chromosome, Start, End or Strand</span>
<span class="sd">            are ignored.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.unstrand : drop strand information</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1, 1], &quot;Start&quot;: [1, 4], &quot;End&quot;: [5, 6],</span>
<span class="sd">        ...                    &quot;Strand&quot;: [&quot;+&quot;, &quot;-&quot;], &quot;Count&quot;: [1, 2],</span>
<span class="sd">        ...                    &quot;Type&quot;: [&quot;exon&quot;, &quot;exon&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       |     Count | Type       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   |   (int64) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+------------|</span>
<span class="sd">        |            1 |         1 |         5 | +            |         1 | exon       |</span>
<span class="sd">        |            1 |         4 |         6 | -            |         2 | exon       |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.drop()</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        |            1 |         1 |         5 | +            |</span>
<span class="sd">        |            1 |         4 |         6 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        Matches with position-columns are ignored:</span>

<span class="sd">        &gt;&gt;&gt; gr.drop(like=&quot;Chromosome|Strand&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       |     Count | Type       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   |   (int64) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+------------|</span>
<span class="sd">        |            1 |         1 |         5 | +            |         1 | exon       |</span>
<span class="sd">        |            1 |         4 |         6 | -            |         2 | exon       |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.drop(like=&quot;e$&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       |     Count |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        |            1 |         1 |         5 | +            |         1 |</span>
<span class="sd">        |            1 |         4 |         6 | -            |         2 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.drop</span> <span class="kn">import</span> <span class="n">_drop</span>
        <span class="k">return</span> <span class="n">_drop</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">drop</span><span class="p">,</span> <span class="n">like</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.drop_duplicate_positions"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.drop_duplicate_positions">[docs]</a>    <span class="k">def</span> <span class="nf">drop_duplicate_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return PyRanges with duplicate postion rows removed.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to take strand-information into account when considering duplicates.</span>

<span class="sd">        keep : {&quot;first&quot;, &quot;last&quot;, False}</span>

<span class="sd">            Whether to keep first, last or drop all duplicates.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_string(&#39;&#39;&#39;Chromosome Start End Strand Name</span>
<span class="sd">        ... 1 1 2 + A</span>
<span class="sd">        ... 1 1 2 - B</span>
<span class="sd">        ... 1 1 2 + Z&#39;&#39;&#39;)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        |            1 |         1 |         2 | +            | A          |</span>
<span class="sd">        |            1 |         1 |         2 | +            | Z          |</span>
<span class="sd">        |            1 |         1 |         2 | -            | B          |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.drop_duplicate_positions()</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        |            1 |         1 |         2 | +            | A          |</span>
<span class="sd">        |            1 |         1 |         2 | -            | B          |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.drop_duplicate_positions(keep=&quot;last&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        |            1 |         1 |         2 | +            | Z          |</span>
<span class="sd">        |            1 |         1 |         2 | -            | B          |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        Note that the reverse strand is considered to be behind the forward strand:</span>

<span class="sd">        &gt;&gt;&gt; gr.drop_duplicate_positions(keep=&quot;last&quot;, strand=False)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        |            1 |         1 |         2 | -            | B          |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 1 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.drop_duplicate_positions(keep=False, strand=False)</span>
<span class="sd">        Empty PyRanges</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.drop_duplicates</span> <span class="kn">import</span> <span class="n">_drop_duplicate_positions</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;keep&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">keep</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">strand</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>
        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span>
            <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_drop_duplicate_positions</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">dtypes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the dtypes of the PyRanges.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.dtypes</span>
<span class="sd">        Chromosome    category</span>
<span class="sd">        Start            int32</span>
<span class="sd">        End              int32</span>
<span class="sd">        Name            object</span>
<span class="sd">        Score            int64</span>
<span class="sd">        Strand        category</span>
<span class="sd">        dtype: object</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dfs</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">dtypes</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">empty</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Indicate whether PyRanges is empty.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>

    
<div class="viewcode-block" id="PyRanges.extend"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.extend">[docs]</a>    <span class="k">def</span> <span class="nf">extend</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extend the intervals from the ends.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        ext : int or dict of ints with &quot;3&quot; and/or &quot;5&quot; as keys.</span>

<span class="sd">            The number of nucleotides to extend the ends with.</span>
<span class="sd">            If an int is provided, the same extension is applied to both </span>
<span class="sd">            the start and end of intervals, while a dict input allows to control </span>
<span class="sd">            differently the two ends. Note also that 5&#39; and 3&#39; extensions take </span>
<span class="sd">            the strand into account, if the intervals are stranded.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        PyRanges.subsequence : obtain subsequences of intervals</span>
<span class="sd">        PyRanges.spliced_subsequence : obtain subsequences of intervals, providing transcript-level coordinates</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 8, 5], &#39;End&#39;: [6, 9, 7],</span>
<span class="sd">        ...      &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.extend(4)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         0 |        10 | +            |</span>
<span class="sd">        | chr1         |         4 |        13 | +            |</span>
<span class="sd">        | chr1         |         1 |        11 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.extend({&quot;3&quot;: 1})</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         7 | +            |</span>
<span class="sd">        | chr1         |         8 |        10 | +            |</span>
<span class="sd">        | chr1         |         4 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.extend({&quot;3&quot;: 1, &quot;5&quot;: 2})</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         7 | +            |</span>
<span class="sd">        | chr1         |         6 |        10 | +            |</span>
<span class="sd">        | chr1         |         4 |         9 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.extend(-1)</span>
<span class="sd">        Traceback (most recent call last):</span>
<span class="sd">        ...</span>
<span class="sd">        AssertionError: Some intervals are negative or zero length after applying extend!</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ext</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">,</span> <span class="s2">&quot;PyRanges must be stranded to add 5/3-end specific extend.&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">({</span><span class="s2">&quot;ext&quot;</span><span class="p">:</span> <span class="n">ext</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">})</span>

        <span class="n">prg</span> <span class="o">=</span> <span class="n">PyRanges</span><span class="p">(</span>
            <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_extend</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">prg</span></div>
    
    
    

    <span class="c1"># # TODO: use subtract code here instead, easier</span>
    <span class="c1"># def no_overlap(self, other, **kwargs):</span>

    <span class="c1">#     kwargs = fill_kwargs(kwargs)</span>
    <span class="c1">#     kwargs[&quot;invert&quot;] = True</span>
    <span class="c1">#     kwargs[&quot;sparse&quot;] = {&quot;self&quot;: False, &quot;other&quot;: True}</span>

    <span class="c1">#     # if kwargs[&quot;strandedness&quot;] in [&quot;same&quot;, &quot;opposite&quot;]:</span>
    <span class="c1">#     #     kwargs[&quot;strandedness&quot;] = {</span>
    <span class="c1">#     #         &quot;same&quot;: &quot;opposite&quot;,</span>
    <span class="c1">#     #         &quot;opposite&quot;: &quot;same&quot;</span>
    <span class="c1">#     #     }[kwargs[&quot;strandedness&quot;]]</span>
    <span class="c1">#     dfs = pyrange_apply(_overlap, self, other, **kwargs)</span>

    <span class="c1">#     return PyRanges(dfs)</span>


    <span class="c1"># @profile</span>



<div class="viewcode-block" id="PyRanges.five_end"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.five_end">[docs]</a>    <span class="k">def</span> <span class="nf">five_end</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the five prime end of intervals.</span>

<span class="sd">        The five prime end is the start of a forward strand or the end of a reverse strand.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with the five prime ends</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        Requires the PyRanges to be stranded.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.three_end : return the 3&#39; end</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 5], &#39;End&#39;: [9, 7],</span>
<span class="sd">        ...                    &#39;Strand&#39;: [&quot;+&quot;, &quot;-&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         9 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.five_end()</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         4 | +            |</span>
<span class="sd">        | chr1         |         7 |         8 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">,</span> <span class="s2">&quot;Need stranded pyrange to find 5&#39;.&quot;</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span>
            <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_tss</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span></div>

<div class="viewcode-block" id="PyRanges.head"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.head">[docs]</a>    <span class="k">def</span> <span class="nf">head</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the n first rows.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        n : int, default 8</span>

<span class="sd">            Return n rows.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with the n first rows.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.tail : return the last rows</span>
<span class="sd">        PyRanges.sample : return random rows</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.head(3)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         |         0 | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         |         0 | +            |</span>
<span class="sd">        | chr1         | 216711011 | 216711036 | U0         |         0 | +            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">subsetter</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="n">subsetter</span><span class="p">[:</span><span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">subsetter</span><span class="p">]</span></div>


<div class="viewcode-block" id="PyRanges.insert"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.insert">[docs]</a>    <span class="k">def</span> <span class="nf">insert</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add one or more columns to the PyRanges.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : Series, DataFrame or dict</span>
<span class="sd">            Data to insert into the PyRanges. `other` must have the same number of rows as the PyRanges.</span>

<span class="sd">        loc : int, default None, i.e. after last column of PyRanges.</span>
<span class="sd">            Insertion index.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>
<span class="sd">            A copy of the PyRanges with the column(s) inserted starting at `loc`.</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        If a Series, or a dict of Series is used, the Series must have a name.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;L&quot;, &quot;E&quot;, &quot;E&quot;, &quot;T&quot;], &quot;Start&quot;: [1, 1, 2, 3], &quot;End&quot;: [5, 8, 13, 21]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | E            |         1 |         8 |</span>
<span class="sd">        | E            |         2 |        13 |</span>
<span class="sd">        | L            |         1 |         5 |</span>
<span class="sd">        | T            |         3 |        21 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 3 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; s = pd.Series(data = [1, 3, 3, 7], name=&quot;Column&quot;)</span>
<span class="sd">        &gt;&gt;&gt; gr.insert(s)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |    Column |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------|</span>
<span class="sd">        | E            |         1 |         8 |         1 |</span>
<span class="sd">        | E            |         2 |        13 |         3 |</span>
<span class="sd">        | L            |         1 |         5 |         3 |</span>
<span class="sd">        | T            |         3 |        21 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 4 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; df = pd.DataFrame({&quot;NY&quot;: s, &quot;AN&quot;: s})</span>
<span class="sd">        &gt;&gt;&gt; df</span>
<span class="sd">           NY  AN</span>
<span class="sd">        0   1   1</span>
<span class="sd">        1   3   3</span>
<span class="sd">        2   3   3</span>
<span class="sd">        3   7   7</span>

<span class="sd">        Note that the original PyRanges was not affected by previously inserting Column:</span>

<span class="sd">        &gt;&gt;&gt; gr.insert(df, 1)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |        NY |        AN |     Start |       End |</span>
<span class="sd">        | (category)   |   (int64) |   (int64) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        | E            |         1 |         1 |         1 |         8 |</span>
<span class="sd">        | E            |         3 |         3 |         2 |        13 |</span>
<span class="sd">        | L            |         3 |         3 |         1 |         5 |</span>
<span class="sd">        | T            |         7 |         7 |         3 |        21 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; arbitrary_result = gr.apply(</span>
<span class="sd">        ... lambda df: pd.Series(df.Start + df.End, name=&quot;Hi!&quot;), as_pyranges=False)</span>
<span class="sd">        &gt;&gt;&gt; arbitrary_result</span>
<span class="sd">        {&#39;E&#39;: 1     9</span>
<span class="sd">        2    15</span>
<span class="sd">        Name: Hi!, dtype: int32, &#39;L&#39;: 0    6</span>
<span class="sd">        Name: Hi!, dtype: int32, &#39;T&#39;: 3    24</span>
<span class="sd">        Name: Hi!, dtype: int32}</span>

<span class="sd">        &gt;&gt;&gt; gr.insert(arbitrary_result)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |       Hi! |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------|</span>
<span class="sd">        | E            |         1 |         8 |         9 |</span>
<span class="sd">        | E            |         2 |        13 |        15 |</span>
<span class="sd">        | L            |         1 |         5 |         6 |</span>
<span class="sd">        | T            |         3 |        21 |        24 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 4 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">loc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.attr</span> <span class="kn">import</span> <span class="n">_setattr</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)):</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="s2">&quot;Pandas Series or DataFrame must be same length as PyRanges!&quot;</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">other</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Series must have a name!&quot;</span><span class="p">)</span>

                <span class="n">_setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">loc</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">other</span><span class="p">:</span>
                    <span class="n">_setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">other</span><span class="p">[</span><span class="n">c</span><span class="p">],</span> <span class="n">loc</span><span class="p">)</span>
                    <span class="n">loc</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="nb">dict</span><span class="p">)</span> <span class="ow">and</span> <span class="n">other</span><span class="p">:</span>

            <span class="n">first</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">other</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
            <span class="n">is_dataframe</span> <span class="o">=</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">first</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">is_dataframe</span><span class="p">:</span>
                <span class="n">columns</span> <span class="o">=</span> <span class="n">first</span><span class="o">.</span><span class="n">columns</span>

                <span class="n">ds</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
                    <span class="n">ds</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">other</span><span class="o">.</span><span class="n">items</span><span class="p">()})</span>

                <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">columns</span><span class="p">,</span> <span class="n">ds</span><span class="p">):</span>
                    <span class="n">_setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">),</span> <span class="n">d</span><span class="p">,</span> <span class="n">loc</span><span class="p">)</span>
                    <span class="n">loc</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">first</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;Series must have a name!&quot;</span><span class="p">)</span>

                <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">other</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
                <span class="n">_setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">first</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">loc</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span></div>


<div class="viewcode-block" id="PyRanges.intersect"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.intersect">[docs]</a>    <span class="k">def</span> <span class="nf">intersect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">invert</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return overlapping subintervals.</span>

<span class="sd">        Returns the segments of the intervals in self which overlap with those in other.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to intersect.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        how : {None, &quot;first&quot;, &quot;last&quot;, &quot;containment&quot;}, default None, i.e. all</span>

<span class="sd">            What intervals to report. By default reports all overlapping intervals. &quot;containment&quot;</span>
<span class="sd">            reports intervals where the overlapping is contained within it.</span>

<span class="sd">        invert : bool, default False</span>

<span class="sd">            Whether to return the intervals without overlaps.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges with overlapping subintervals.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.set_intersect : set-intersect PyRanges</span>
<span class="sd">        PyRanges.overlap : report overlapping intervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [1, 4, 10],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 9, 11], &quot;ID&quot;: [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | ID         |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |         chr1 |         1 |         3 | a          |</span>
<span class="sd">        |         chr1 |         4 |         9 | b          |</span>
<span class="sd">        |         chr1 |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2 = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [2, 2, 9], &quot;End&quot;: [3, 9, 10]})</span>
<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |         chr1 |         2 |         3 |</span>
<span class="sd">        |         chr1 |         2 |         9 |</span>
<span class="sd">        |         chr1 |         9 |        10 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.intersect(gr2)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | ID         |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |         chr1 |         2 |         3 | a          |</span>
<span class="sd">        |         chr1 |         2 |         3 | a          |</span>
<span class="sd">        |         chr1 |         4 |         9 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.intersect(gr2, how=&quot;first&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | ID         |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |         chr1 |         2 |         3 | a          |</span>
<span class="sd">        |         chr1 |         4 |         9 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.intersect(gr2, how=&quot;containment&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         4 |         9 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;how&quot;</span><span class="p">:</span> <span class="n">how</span><span class="p">,</span> <span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;other&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="k">if</span> <span class="n">invert</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__ix__</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_intersection</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">invert</span><span class="p">:</span>
            <span class="n">found_idxs</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="s2">&quot;__ix__&quot;</span><span class="p">,</span> <span class="p">[])</span>
            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="o">~</span><span class="bp">self</span><span class="o">.</span><span class="n">__ix__</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">found_idxs</span><span class="p">)]</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;__ix__&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="PyRanges.items"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.items">[docs]</a>    <span class="k">def</span> <span class="nf">items</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the pairs of keys and DataFrames.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        dict</span>

<span class="sd">            The dict mapping keys to DataFrames in the PyRanges.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.chromosomes : return the chromosomes</span>
<span class="sd">        PyRanges.keys : return the keys</span>
<span class="sd">        PyRanges.values : return the DataFrames in the PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.f1()</span>
<span class="sd">        &gt;&gt;&gt; gr.items()</span>
<span class="sd">        [((&#39;chr1&#39;, &#39;+&#39;),   Chromosome  Start  End       Name  Score Strand</span>
<span class="sd">        0       chr1      3    6  interval1      0      +</span>
<span class="sd">        2       chr1      8    9  interval3      0      +), ((&#39;chr1&#39;, &#39;-&#39;),   Chromosome  Start  End       Name  Score Strand</span>
<span class="sd">        1       chr1      5    7  interval2      0      -)]</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">natsorted</span><span class="p">([(</span><span class="n">k</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">dfs</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span></div>

<div class="viewcode-block" id="PyRanges.join"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.join">[docs]</a>    <span class="k">def</span> <span class="nf">join</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">report_overlap</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">slack</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_b&quot;</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">apply_strand_suffix</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Join PyRanges on genomic location.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to join.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        how : {None, &quot;left&quot;, &quot;right&quot;}, default None, i.e. &quot;inner&quot;</span>

<span class="sd">            How to handle intervals without overlap. None means only keep overlapping intervals.</span>
<span class="sd">            &quot;left&quot; keeps all intervals in self, &quot;right&quot; keeps all intervals in other.</span>

<span class="sd">        report_overlap : bool, default False</span>

<span class="sd">            Report amount of overlap in base pairs.</span>

<span class="sd">        slack : int, default 0</span>

<span class="sd">            Lengthen intervals in self before joining.</span>

<span class="sd">        suffix : str or tuple, default &quot;_b&quot;</span>

<span class="sd">            Suffix to give overlapping columns in other.</span>

<span class="sd">        apply_strand_suffix : bool, default None</span>

<span class="sd">            If first pyranges is unstranded, but the second is not, the first will be given a strand column.</span>
<span class="sd">            apply_strand_suffix makes the added strand column a regular data column instead by adding a suffix.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges appended with columns of another.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        The chromosome from other will never be reported as it is always the same as in self.</span>

<span class="sd">        As pandas did not have NaN for non-float datatypes until recently, &quot;left&quot; and &quot;right&quot; join</span>
<span class="sd">        give non-overlapping rows the value -1 to avoid promoting columns to object. This will</span>
<span class="sd">        change to NaN in a future version as general NaN becomes stable in pandas.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.new_position : give joined PyRanges new coordinates</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; f1 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 8, 5],</span>
<span class="sd">        ...                    &#39;End&#39;: [6, 9, 7], &#39;Name&#39;: [&#39;interval1&#39;, &#39;interval3&#39;, &#39;interval2&#39;]})</span>
<span class="sd">        &gt;&gt;&gt; f1</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; f2 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...                    &#39;End&#39;: [2, 7], &#39;Name&#39;: [&#39;a&#39;, &#39;b&#39;]})</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         1 |         2 | a          |</span>
<span class="sd">        | chr1         |         6 |         7 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; f1.join(f2)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |   Start_b |     End_b | Name_b     |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         6 |         7 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; f1.join(f2, how=&quot;right&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |   Start_b |     End_b | Name_b     |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         6 |         7 | b          |</span>
<span class="sd">        | chr1         |        -1 |        -1 | -1         |         1 |         2 | a          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        With slack 1, bookended features are joined (see row 1):</span>

<span class="sd">        &gt;&gt;&gt; f1.join(f2, slack=1)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |   Start_b |     End_b | Name_b     |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         6 |         7 | b          |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         6 |         7 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.join</span> <span class="kn">import</span> <span class="n">_write_both</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;how&quot;</span><span class="p">:</span> <span class="n">how</span><span class="p">,</span> <span class="s2">&quot;report_overlap&quot;</span><span class="p">:</span><span class="n">report_overlap</span><span class="p">,</span> <span class="s2">&quot;suffix&quot;</span><span class="p">:</span> <span class="n">suffix</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">,</span> <span class="s2">&quot;apply_strand_suffix&quot;</span><span class="p">:</span> <span class="n">apply_strand_suffix</span><span class="p">}</span>
        <span class="k">if</span> <span class="n">slack</span><span class="p">:</span>
            <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">Start__slack</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Start</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">End__slack</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">End</span>

            <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">slack</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;suffix&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">],</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">suffixes</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">]</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffixes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">suffixes</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">how</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;how&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">how</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="s2">&quot;outer&quot;</span><span class="p">]:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;example_header_other&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">df</span>
        <span class="k">if</span> <span class="n">how</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="s2">&quot;outer&quot;</span><span class="p">]:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;example_header_self&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">df</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_write_both</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">gr</span> <span class="o">=</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">slack</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">gr</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">gr</span><span class="o">.</span><span class="n">Start</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">Start__slack</span>
            <span class="n">gr</span><span class="o">.</span><span class="n">End</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">End__slack</span>
            <span class="n">gr</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">like</span><span class="o">=</span><span class="s2">&quot;(Start|End).*__slack&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">apply_strand_suffix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="kn">import</span> <span class="nn">sys</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;join: Strand data from other will be added as strand data to self.</span><span class="se">\n</span><span class="s2">If this is undesired use the flag apply_strand_suffix=False.</span><span class="se">\n</span><span class="s2">To turn off the warning set apply_strand_suffix to True or False.&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">apply_strand_suffix</span><span class="p">:</span>
                <span class="n">gr</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span> <span class="s2">&quot;Strand&quot;</span> <span class="o">+</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">gr</span></div>

<div class="viewcode-block" id="PyRanges.keys"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.keys">[docs]</a>    <span class="k">def</span> <span class="nf">keys</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the keys.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>

<span class="sd">        Returns the keys (chromosomes or chromosome/strand pairs) as strings or tuples of strings</span>
<span class="sd">        in natsorted order.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.chromosomes : return the chromosomes</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; gr.keys()</span>
<span class="sd">        [(&#39;chr1&#39;, &#39;+&#39;), (&#39;chr1&#39;, &#39;-&#39;), (&#39;chr2&#39;, &#39;+&#39;), (&#39;chr2&#39;, &#39;-&#39;), (&#39;chr3&#39;, &#39;+&#39;), (&#39;chr3&#39;, &#39;-&#39;), (&#39;chr4&#39;, &#39;+&#39;), (&#39;chr4&#39;, &#39;-&#39;), (&#39;chr5&#39;, &#39;+&#39;), (&#39;chr5&#39;, &#39;-&#39;), (&#39;chr6&#39;, &#39;+&#39;), (&#39;chr6&#39;, &#39;-&#39;), (&#39;chr7&#39;, &#39;+&#39;), (&#39;chr7&#39;, &#39;-&#39;), (&#39;chr8&#39;, &#39;+&#39;), (&#39;chr8&#39;, &#39;-&#39;), (&#39;chr9&#39;, &#39;+&#39;), (&#39;chr9&#39;, &#39;-&#39;), (&#39;chr10&#39;, &#39;+&#39;), (&#39;chr10&#39;, &#39;-&#39;), (&#39;chr11&#39;, &#39;+&#39;), (&#39;chr11&#39;, &#39;-&#39;), (&#39;chr12&#39;, &#39;+&#39;), (&#39;chr12&#39;, &#39;-&#39;), (&#39;chr13&#39;, &#39;+&#39;), (&#39;chr13&#39;, &#39;-&#39;), (&#39;chr14&#39;, &#39;+&#39;), (&#39;chr14&#39;, &#39;-&#39;), (&#39;chr15&#39;, &#39;+&#39;), (&#39;chr15&#39;, &#39;-&#39;), (&#39;chr16&#39;, &#39;+&#39;), (&#39;chr16&#39;, &#39;-&#39;), (&#39;chr17&#39;, &#39;+&#39;), (&#39;chr17&#39;, &#39;-&#39;), (&#39;chr18&#39;, &#39;+&#39;), (&#39;chr18&#39;, &#39;-&#39;), (&#39;chr19&#39;, &#39;+&#39;), (&#39;chr19&#39;, &#39;-&#39;), (&#39;chr20&#39;, &#39;+&#39;), (&#39;chr20&#39;, &#39;-&#39;), (&#39;chr21&#39;, &#39;+&#39;), (&#39;chr21&#39;, &#39;-&#39;), (&#39;chr22&#39;, &#39;+&#39;), (&#39;chr22&#39;, &#39;-&#39;), (&#39;chrX&#39;, &#39;+&#39;), (&#39;chrX&#39;, &#39;-&#39;), (&#39;chrY&#39;, &#39;+&#39;), (&#39;chrY&#39;, &#39;-&#39;)]</span>
<span class="sd">        &gt;&gt;&gt; gr.unstrand().keys()</span>
<span class="sd">        [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;, &#39;chr4&#39;, &#39;chr5&#39;, &#39;chr6&#39;, &#39;chr7&#39;, &#39;chr8&#39;, &#39;chr9&#39;, &#39;chr10&#39;, &#39;chr11&#39;, &#39;chr12&#39;, &#39;chr13&#39;, &#39;chr14&#39;, &#39;chr15&#39;, &#39;chr16&#39;, &#39;chr17&#39;, &#39;chr18&#39;, &#39;chr19&#39;, &#39;chr20&#39;, &#39;chr21&#39;, &#39;chr22&#39;, &#39;chrX&#39;, &#39;chrY&#39;]</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="n">natsorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dfs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span></div>

<div class="viewcode-block" id="PyRanges.k_nearest"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.k_nearest">[docs]</a>    <span class="k">def</span> <span class="nf">k_nearest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ties</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_b&quot;</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">apply_strand_suffix</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find k nearest intervals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to find nearest interval in.</span>

<span class="sd">        k : int or list/array/Series of int</span>

<span class="sd">            Number of closest to return. If iterable, must be same length as PyRanges.</span>

<span class="sd">        ties : {None, &quot;first&quot;, &quot;last&quot;, &quot;different&quot;}, default None</span>

<span class="sd">            How to resolve ties, i.e. closest intervals with equal distance. None means that the k nearest intervals are kept.</span>
<span class="sd">            &quot;first&quot; means that the first tie is kept, &quot;last&quot; meanst that the last is kept.</span>
<span class="sd">            &quot;different&quot; means that all nearest intervals with the k unique nearest distances are kept.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are stranded,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        overlap : bool, default True</span>

<span class="sd">            Whether to include overlaps.</span>

<span class="sd">        how : {None, &quot;upstream&quot;, &quot;downstream&quot;}, default None, i.e. both directions</span>

<span class="sd">            Whether to only look for nearest in one direction. Always with respect to the PyRanges</span>
<span class="sd">            it is called on.</span>

<span class="sd">        suffix : str, default &quot;_b&quot;</span>

<span class="sd">            Suffix to give columns with shared name in other.</span>

<span class="sd">        apply_strand_suffix : bool, default None</span>

<span class="sd">            If first pyranges is unstranded, but the second is not, the first will be given a strand column.</span>
<span class="sd">            apply_strand_suffix makes the added strand column a regular data column instead by adding a suffix.</span>


<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges with columns of nearest interval horizontally appended.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        nearest also exists, and is more performant.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.new_position : give joined PyRanges new coordinates</span>
<span class="sd">        PyRanges.nearest : find nearest intervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; f1 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 8, 5],</span>
<span class="sd">        ...                    &#39;End&#39;: [6, 9, 7], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;]})</span>
<span class="sd">        &gt;&gt;&gt; f1</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f2 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...                    &#39;End&#39;: [2, 7], &#39;Strand&#39;: [&#39;+&#39;, &#39;-&#39;]})</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         2 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.k_nearest(f2, k=2)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         6 |         7 | -            |          1 |</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         1 |         2 | +            |         -2 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         6 |         7 | -            |         -2 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         1 |         2 | +            |         -7 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         1 |         2 | +            |          4 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 6 rows and 8 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.k_nearest(f2, how=&quot;upstream&quot;, k=2)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         1 |         2 | +            |         -2 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         6 |         7 | -            |         -2 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         1 |         2 | +            |         -7 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 8 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.k_nearest(f2, k=[1, 2, 1])</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         6 |         7 | -            |          1 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         6 |         7 | -            |         -2 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         1 |         2 | +            |         -7 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 8 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; d1 = {&quot;Chromosome&quot;: [1], &quot;Start&quot;: [5], &quot;End&quot;: [6]}</span>
<span class="sd">        &gt;&gt;&gt; d2 = {&quot;Chromosome&quot;: 1, &quot;Start&quot;: [1] * 2 + [5] * 2 + [9] * 2,</span>
<span class="sd">        ...       &quot;End&quot;: [3] * 2 + [7] * 2 + [11] * 2, &quot;ID&quot;: range(6)}</span>
<span class="sd">        &gt;&gt;&gt; gr, gr2 = pr.from_dict(d1), pr.from_dict(d2)</span>

<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |         5 |         6 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |        ID |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         1 |         3 |         0 |</span>
<span class="sd">        |            1 |         1 |         3 |         1 |</span>
<span class="sd">        |            1 |         5 |         7 |         2 |</span>
<span class="sd">        |            1 |         5 |         7 |         3 |</span>
<span class="sd">        |            1 |         9 |        11 |         4 |</span>
<span class="sd">        |            1 |         9 |        11 |         5 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 6 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.k_nearest(gr2, k=2)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         5 |         6 |         5 |         7 |         2 |          0 |</span>
<span class="sd">        |            1 |         5 |         6 |         5 |         7 |         3 |          0 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.k_nearest(gr2, k=2, ties=&quot;different&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         5 |         6 |         5 |         7 |         2 |          0 |</span>
<span class="sd">        |            1 |         5 |         6 |         5 |         7 |         3 |          0 |</span>
<span class="sd">        |            1 |         5 |         6 |         1 |         3 |         1 |         -3 |</span>
<span class="sd">        |            1 |         5 |         6 |         1 |         3 |         0 |         -3 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.k_nearest(gr2, k=3, ties=&quot;first&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         5 |         6 |         5 |         7 |         2 |          0 |</span>
<span class="sd">        |            1 |         5 |         6 |         1 |         3 |         1 |         -3 |</span>
<span class="sd">        |            1 |         5 |         6 |         9 |        11 |         4 |          4 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.k_nearest(gr2, k=1, overlap=False)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="sd">        |            1 |         5 |         6 |         1 |         3 |         1 |         -3 |</span>
<span class="sd">        |            1 |         5 |         6 |         1 |         3 |         0 |         -3 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.k_nearest</span> <span class="kn">import</span> <span class="n">_nearest</span>
        <span class="kn">from</span> <span class="nn">sorted_nearest</span> <span class="kn">import</span> <span class="n">get_all_ties</span><span class="p">,</span> <span class="n">get_different_ties</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;how&quot;</span><span class="p">:</span> <span class="n">how</span><span class="p">,</span> <span class="s2">&quot;overlap&quot;</span><span class="p">:</span> <span class="n">overlap</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">,</span>
                  <span class="s2">&quot;k&quot;</span><span class="p">:</span> <span class="n">k</span><span class="p">,</span> <span class="s2">&quot;ties&quot;</span><span class="p">:</span> <span class="n">ties</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;stranded&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">overlap</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;overlap&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">ties</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ties&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">try</span><span class="p">:</span> <span class="c1"># if k is a Series</span>
            <span class="n">k</span> <span class="o">=</span> <span class="n">k</span><span class="o">.</span><span class="n">values</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="c1"># how many to nearest to find; might be different for each</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__k__</span> <span class="o">=</span> <span class="n">k</span>
        <span class="c1"># give each their own unique ID</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">__IX__</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_nearest</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">nearest</span> <span class="o">=</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">overlap</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">nearest</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
            <span class="n">overlap_how</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="p">{</span><span class="s2">&quot;first&quot;</span><span class="p">:</span> <span class="s2">&quot;first&quot;</span><span class="p">,</span> <span class="s2">&quot;last&quot;</span><span class="p">:</span> <span class="s2">&quot;last&quot;</span><span class="p">})[</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ties&quot;</span><span class="p">)]</span>
            <span class="n">overlaps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="n">strandedness</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="n">overlap_how</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="n">nb_cpu</span><span class="p">,</span> <span class="n">apply_strand_suffix</span><span class="o">=</span><span class="n">apply_strand_suffix</span><span class="p">)</span>
            <span class="n">overlaps</span><span class="o">.</span><span class="n">Distance</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">overlaps</span><span class="p">,</span> <span class="n">nearest</span><span class="p">])</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">()</span>
        <span class="n">new_result</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">ties</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;first&quot;</span><span class="p">,</span> <span class="s2">&quot;last&quot;</span><span class="p">]:</span>
            <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;__IX__&quot;</span><span class="p">,</span> <span class="s2">&quot;Distance&quot;</span><span class="p">])</span>
                <span class="n">grpby</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;__k__&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">kdf</span> <span class="ow">in</span> <span class="n">grpby</span><span class="p">:</span>
                    <span class="n">grpby2</span> <span class="o">=</span> <span class="n">kdf</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;__IX__&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">_df</span> <span class="o">=</span> <span class="n">grpby2</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
                    <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_df</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">dfs</span><span class="p">:</span>
                    <span class="n">new_result</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">elif</span> <span class="n">ties</span> <span class="o">==</span> <span class="s2">&quot;different&quot;</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">ties</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>

                <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>

                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s2">&quot;__IX__&quot;</span><span class="p">,</span> <span class="s2">&quot;Distance&quot;</span><span class="p">])</span>
                <span class="n">grpby</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;__k__&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">kdf</span> <span class="ow">in</span> <span class="n">grpby</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">ties</span><span class="p">:</span>
                        <span class="n">lx</span> <span class="o">=</span> <span class="n">get_different_ties</span><span class="p">(</span><span class="n">kdf</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">kdf</span><span class="o">.</span><span class="n">__IX__</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">kdf</span><span class="o">.</span><span class="n">Distance</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
                        <span class="n">_df</span> <span class="o">=</span> <span class="n">kdf</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">lx</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">lx</span> <span class="o">=</span> <span class="n">get_all_ties</span><span class="p">(</span><span class="n">kdf</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">kdf</span><span class="o">.</span><span class="n">__IX__</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">kdf</span><span class="o">.</span><span class="n">Distance</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
                        <span class="n">_df</span> <span class="o">=</span> <span class="n">kdf</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">lx</span><span class="p">)</span>
                        <span class="n">_df</span> <span class="o">=</span> <span class="n">_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;__IX__&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
                    <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_df</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">dfs</span><span class="p">:</span>
                    <span class="n">new_result</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">new_result</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="o">.</span><span class="n">__IX__</span><span class="o">.</span><span class="n">is_monotonic</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="s2">&quot;__IX__&quot;</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">like</span><span class="o">=</span><span class="s2">&quot;__IX__|__k__&quot;</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">like</span><span class="o">=</span><span class="s2">&quot;__k__|__IX__&quot;</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">prev_to_neg</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

            <span class="n">strand</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">Strand</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="s2">&quot;Strand&quot;</span> <span class="ow">in</span> <span class="n">df</span> <span class="k">else</span> <span class="s2">&quot;+&quot;</span>

            <span class="n">suffix</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">]</span>

            <span class="n">bools</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;End&quot;</span> <span class="o">+</span> <span class="n">suffix</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">df</span><span class="o">.</span><span class="n">Start</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
                <span class="n">bools</span> <span class="o">=</span> <span class="o">~</span><span class="n">bools</span>

            <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">bools</span><span class="p">,</span> <span class="s2">&quot;Distance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">bools</span><span class="p">,</span> <span class="s2">&quot;Distance&quot;</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">df</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">prev_to_neg</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">])</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">apply_strand_suffix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="kn">import</span> <span class="nn">sys</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;join: Strand data from other will be added as strand data to self.</span><span class="se">\n</span><span class="s2">If this is undesired use the flag apply_strand_suffix=False.</span><span class="se">\n</span><span class="s2">To turn off the warning set apply_strand_suffix to True or False.&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">apply_strand_suffix</span><span class="p">:</span>
                <span class="n">result</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span> <span class="s2">&quot;Strand&quot;</span> <span class="o">+</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">result</span></div>



    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the total length of the intervals.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.lengths : return the intervals lengths</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.f1()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.length</span>
<span class="sd">        6</span>

<span class="sd">        To find the length of the genome covered by the intervals, use merge first:</span>

<span class="sd">        &gt;&gt;&gt; gr.merge(strand=False).length</span>
<span class="sd">        5</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">(</span><span class="n">as_dict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>


<div class="viewcode-block" id="PyRanges.lengths"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.lengths">[docs]</a>    <span class="k">def</span> <span class="nf">lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">as_dict</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the length of each interval.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        as_dict : bool, default False</span>

<span class="sd">            Whether to return lengths as Series or dict of Series per key.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Series or dict of Series with the lengths of each interval.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.lengths : return the intervals lengths</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.f1()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.lengths()</span>
<span class="sd">        0    3</span>
<span class="sd">        1    1</span>
<span class="sd">        2    2</span>
<span class="sd">        dtype: int32</span>

<span class="sd">        To find the length of the genome covered by the intervals, use merge first:</span>

<span class="sd">        &gt;&gt;&gt; gr.Length = gr.lengths()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |    Length |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------+-----------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |         3 |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |         1 |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |         2 |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">as_dict</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
                <span class="k">return</span> <span class="p">{}</span>
            <span class="n">lengths</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">lengths</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span> <span class="o">-</span> <span class="n">df</span><span class="o">.</span><span class="n">Start</span>

            <span class="k">return</span> <span class="n">lengths</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">_lengths</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">_lengths</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">:</span>
                <span class="n">lengths</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span> <span class="o">-</span> <span class="n">df</span><span class="o">.</span><span class="n">Start</span>
                <span class="n">_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lengths</span><span class="p">)</span>

            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">_lengths</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.max_disjoint"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.max_disjoint">[docs]</a>    <span class="k">def</span> <span class="nf">max_disjoint</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">slack</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find the maximal disjoint set of intervals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Find the max disjoint set separately for each strand.</span>

<span class="sd">        slack : int, default 0</span>

<span class="sd">            Consider intervals within a distance of slack to be overlapping.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with maximal disjoint set of intervals.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.data.f1()</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.max_disjoint(strand=False)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;slack&quot;</span><span class="p">:</span> <span class="n">slack</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.max_disjoint</span> <span class="kn">import</span> <span class="n">_max_disjoint</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_max_disjoint</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.merge"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.merge">[docs]</a>    <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">count_col</span><span class="o">=</span><span class="s2">&quot;Count&quot;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">slack</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Merge overlapping intervals into one.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Only merge intervals on same strand.</span>

<span class="sd">        count : bool, default False</span>

<span class="sd">            Count intervals in each superinterval.</span>

<span class="sd">        count_col : str, default &quot;Count&quot;</span>

<span class="sd">            Name of column with counts.</span>

<span class="sd">        by : str or list of str, default None</span>

<span class="sd">            Only merge intervals with equal values in these columns.</span>

<span class="sd">        slack : int, default 0</span>

<span class="sd">            Allow this many nucleotides between each interval to merge.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with superintervals.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        To avoid losing metadata, use cluster instead. If you want to perform a reduction function</span>
<span class="sd">        on the metadata, use pandas groupby.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.cluster : annotate overlapping intervals with common ID</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.ensembl_gtf()[[&quot;Feature&quot;, &quot;gene_name&quot;]]</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="sd">        | 1            | gene         | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | transcript   | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | exon         | 11868     | 12227     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | exon         | 12612     | 12721     | +            | DDX11L1     |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="sd">        | 1            | gene         | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | transcript   | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1179364   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1173055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.merge(count=True, count_col=&quot;Count&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Strand       | Count     |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (category)   | (int32)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        | 1            | 11868     | 14409     | +            | 12        |</span>
<span class="sd">        | 1            | 29553     | 31109     | +            | 11        |</span>
<span class="sd">        | 1            | 52472     | 53312     | +            | 3         |</span>
<span class="sd">        | 1            | 57597     | 64116     | +            | 7         |</span>
<span class="sd">        | ...          | ...       | ...       | ...          | ...       |</span>
<span class="sd">        | 1            | 1062207   | 1063288   | -            | 4         |</span>
<span class="sd">        | 1            | 1070966   | 1074306   | -            | 10        |</span>
<span class="sd">        | 1            | 1081817   | 1116361   | -            | 319       |</span>
<span class="sd">        | 1            | 1173055   | 1179555   | -            | 4         |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 62 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.merge(by=&quot;Feature&quot;, count=True)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+--------------+-----------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Strand       | Feature      | Count     |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (category)   | (category)   | (int32)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+--------------+-----------|</span>
<span class="sd">        | 1            | 65564     | 65573     | +            | CDS          | 1         |</span>
<span class="sd">        | 1            | 69036     | 70005     | +            | CDS          | 2         |</span>
<span class="sd">        | 1            | 924431    | 924948    | +            | CDS          | 1         |</span>
<span class="sd">        | 1            | 925921    | 926013    | +            | CDS          | 11        |</span>
<span class="sd">        | ...          | ...       | ...       | ...          | ...          | ...       |</span>
<span class="sd">        | 1            | 1062207   | 1063288   | -            | transcript   | 1         |</span>
<span class="sd">        | 1            | 1070966   | 1074306   | -            | transcript   | 1         |</span>
<span class="sd">        | 1            | 1081817   | 1116361   | -            | transcript   | 19        |</span>
<span class="sd">        | 1            | 1173055   | 1179555   | -            | transcript   | 1         |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 748 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.merge(by=[&quot;Feature&quot;, &quot;gene_name&quot;], count=True)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+--------------+-------------+-----------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Strand       | Feature      | gene_name   | Count     |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (category)   | (category)   | (object)    | (int32)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+--------------+-------------+-----------|</span>
<span class="sd">        | 1            | 1020172   | 1020373   | +            | CDS          | AGRN        | 1         |</span>
<span class="sd">        | 1            | 1022200   | 1022462   | +            | CDS          | AGRN        | 2         |</span>
<span class="sd">        | 1            | 1034555   | 1034703   | +            | CDS          | AGRN        | 2         |</span>
<span class="sd">        | 1            | 1035276   | 1035324   | +            | CDS          | AGRN        | 4         |</span>
<span class="sd">        | ...          | ...       | ...       | ...          | ...          | ...         | ...       |</span>
<span class="sd">        | 1            | 347981    | 348366    | -            | transcript   | RPL23AP24   | 1         |</span>
<span class="sd">        | 1            | 1173055   | 1179555   | -            | transcript   | TTLL10-AS1  | 1         |</span>
<span class="sd">        | 1            | 14403     | 29570     | -            | transcript   | WASH7P      | 1         |</span>
<span class="sd">        | 1            | 185216    | 195411    | -            | transcript   | WASH9P      | 1         |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+--------------+-------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 807 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">:</span> <span class="n">count</span><span class="p">,</span> <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="n">by</span><span class="p">,</span> <span class="s2">&quot;count_col&quot;</span><span class="p">:</span> <span class="n">count_col</span><span class="p">,</span> <span class="s2">&quot;slack&quot;</span><span class="p">:</span> <span class="n">slack</span><span class="p">}</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;by&quot;</span><span class="p">]:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
            <span class="kn">from</span> <span class="nn">pyranges.methods.merge</span> <span class="kn">import</span> <span class="n">_merge</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_merge</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
            <span class="kn">from</span> <span class="nn">pyranges.methods.merge</span> <span class="kn">import</span> <span class="n">_merge_by</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_merge_by</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.mp"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.mp">[docs]</a>    <span class="k">def</span> <span class="nf">mp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Merge location and print.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">tostring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">merge_position</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span></div>

<div class="viewcode-block" id="PyRanges.mpc"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.mpc">[docs]</a>    <span class="k">def</span> <span class="nf">mpc</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Merge location, print and return self.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">tostring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">merge_position</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span>

        <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="PyRanges.msp"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.msp">[docs]</a>    <span class="k">def</span> <span class="nf">msp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sort on location, merge location info and print.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span>
            <span class="n">tostring</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span>
                <span class="n">merge_position</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span></div>


<div class="viewcode-block" id="PyRanges.mspc"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.mspc">[docs]</a>    <span class="k">def</span> <span class="nf">mspc</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sort on location, merge location, print and return self.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span>
            <span class="n">tostring</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span>
                <span class="n">merge_position</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span>

        <span class="k">return</span> <span class="bp">self</span></div>


<div class="viewcode-block" id="PyRanges.nearest"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.nearest">[docs]</a>    <span class="k">def</span> <span class="nf">nearest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;_b&quot;</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">apply_strand_suffix</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find closest interval.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to find nearest interval in.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        overlap : bool, default True</span>

<span class="sd">            Whether to include overlaps.</span>

<span class="sd">        how : {None, &quot;upstream&quot;, &quot;downstream&quot;}, default None, i.e. both directions</span>

<span class="sd">            Whether to only look for nearest in one direction. Always with respect to the PyRanges</span>
<span class="sd">            it is called on.</span>

<span class="sd">        suffix : str, default &quot;_b&quot;</span>

<span class="sd">            Suffix to give columns with shared name in other.</span>

<span class="sd">        apply_strand_suffix : bool, default None</span>

<span class="sd">            If first pyranges is unstranded, but the second is not, the first will be given the strand column of the second.</span>
<span class="sd">            apply_strand_suffix makes the added strand column a regular data column instead by adding a suffix.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges with columns representing nearest interval horizontally appended.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        A k_nearest also exists, but is less performant.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.new_position : give joined PyRanges new coordinates</span>
<span class="sd">        PyRanges.k_nearest : find k nearest intervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; f1 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 8, 5],</span>
<span class="sd">        ...                    &#39;End&#39;: [6, 9, 7], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;]})</span>
<span class="sd">        &gt;&gt;&gt; f1</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f2 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...                    &#39;End&#39;: [2, 7], &#39;Strand&#39;: [&#39;+&#39;, &#39;-&#39;]})</span>
<span class="sd">        &gt;&gt;&gt; f2</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         2 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.nearest(f2)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         6 |         7 | -            |          1 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         6 |         7 | -            |          2 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; f1.nearest(f2, how=&quot;upstream&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |         1 |         2 | +            |          2 |</span>
<span class="sd">        | chr1         |         8 |         9 | +            |         6 |         7 | -            |          2 |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.nearest</span> <span class="kn">import</span> <span class="n">_nearest</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;how&quot;</span><span class="p">:</span> <span class="n">how</span><span class="p">,</span> <span class="s2">&quot;overlap&quot;</span><span class="p">:</span> <span class="n">overlap</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">,</span> <span class="s2">&quot;suffix&quot;</span><span class="p">:</span> <span class="n">suffix</span><span class="p">,</span> <span class="s2">&quot;apply_strand_suffix&quot;</span><span class="p">:</span> <span class="n">apply_strand_suffix</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;how&quot;</span><span class="p">)</span> <span class="ow">in</span> <span class="s2">&quot;upstream downstream&quot;</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
            <span class="k">assert</span> <span class="n">other</span><span class="o">.</span><span class="n">stranded</span><span class="p">,</span> <span class="s2">&quot;If doing upstream or downstream nearest, other pyranges must be stranded&quot;</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_nearest</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">gr</span> <span class="o">=</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">apply_strand_suffix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="kn">import</span> <span class="nn">sys</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;join: Strand data from other will be added as strand data to self.</span><span class="se">\n</span><span class="s2">If this is undesired use the flag apply_strand_suffix=False.</span><span class="se">\n</span><span class="s2">To turn off the warning set apply_strand_suffix to True or False.&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">apply_strand_suffix</span><span class="p">:</span>
                <span class="n">gr</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span> <span class="s2">&quot;Strand&quot;</span> <span class="o">+</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;suffix&quot;</span><span class="p">])</span>

        <span class="k">return</span> <span class="n">gr</span></div>


<div class="viewcode-block" id="PyRanges.new_position"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.new_position">[docs]</a>    <span class="k">def</span> <span class="nf">new_position</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_pos</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Give new position.</span>

<span class="sd">        The operation join produces a PyRanges with two pairs of start coordinates and two pairs of</span>
<span class="sd">        end coordinates. This operation uses these to give the PyRanges a new position.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        new_pos : {&quot;union&quot;, &quot;intersection&quot;, &quot;swap&quot;}</span>

<span class="sd">           Change of coordinates.</span>

<span class="sd">        suffixes : tuple of str, default (&quot;&quot;, &quot;_b&quot;)</span>

<span class="sd">           Suffixes of coordinate-columns to switch.</span>

<span class="sd">        columns : tuple of str, default None, i.e. auto</span>

<span class="sd">           The name of the coordinate columns. By default uses the two first columns containing</span>
<span class="sd">           &quot;Start&quot; and the two first columns containing &quot;End&quot;.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.join : combine two PyRanges horizontally with SQL-style joins.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with new coordinates.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;],</span>
<span class="sd">        ...                    &#39;Start&#39;: [3, 8, 5], &#39;End&#39;: [6, 9, 7]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         3 |         6 |</span>
<span class="sd">        | chr1         |         8 |         9 |</span>
<span class="sd">        | chr1         |         5 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2 = pr.from_dict({&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...                     &#39;End&#39;: [4, 7]})</span>
<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         1 |         4 |</span>
<span class="sd">        | chr1         |         6 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; j = gr.join(gr2)</span>
<span class="sd">        &gt;&gt;&gt; j</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |   Start_b |     End_b |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        | chr1         |         3 |         6 |         1 |         4 |</span>
<span class="sd">        | chr1         |         5 |         7 |         6 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; j.new_position(&quot;swap&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |   Start_b |     End_b |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        | chr1         |         1 |         4 |         3 |         6 |</span>
<span class="sd">        | chr1         |         6 |         7 |         5 |         7 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; j.new_position(&quot;union&quot;).mp()</span>
<span class="sd">        +--------------------+-----------+-----------+</span>
<span class="sd">        | - Position -       |   Start_b |     End_b |</span>
<span class="sd">        | (Multiple types)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------------+-----------+-----------|</span>
<span class="sd">        | chr1 1-6           |         1 |         4 |</span>
<span class="sd">        | chr1 5-7           |         6 |         7 |</span>
<span class="sd">        +--------------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; j.new_position(&quot;intersection&quot;).mp()</span>
<span class="sd">        +--------------------+-----------+-----------+</span>
<span class="sd">        | - Position -       |   Start_b |     End_b |</span>
<span class="sd">        | (Multiple types)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------------+-----------+-----------|</span>
<span class="sd">        | chr1 1-4           |         1 |         4 |</span>
<span class="sd">        | chr1 6-7           |         6 |         7 |</span>
<span class="sd">        +--------------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; j2 = pr.from_dict({&quot;Chromosome&quot;: [1], &quot;Start&quot;: [3],</span>
<span class="sd">        ...                   &quot;End&quot;: [4], &quot;A&quot;: [1], &quot;B&quot;: [3], &quot;C&quot;: [2], &quot;D&quot;: [5]})</span>
<span class="sd">        &gt;&gt;&gt; j2</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |         A |         B |         C |         D |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (int64) |   (int64) |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         3 |         4 |         1 |         3 |         2 |         5 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; j2.new_position(&quot;intersection&quot;, (&quot;A&quot;, &quot;B&quot;, &quot;C&quot;, &quot;D&quot;))</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |         A |         B |         C |         D |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |   (int64) |   (int64) |   (int64) |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="sd">        |            1 |         2 |         3 |         1 |         3 |         2 |         5 |</span>
<span class="sd">        +--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.new_position</span> <span class="kn">import</span> <span class="n">_new_position</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;new_pos&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_pos</span>

        <span class="k">if</span> <span class="n">columns</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">start1</span><span class="p">,</span> <span class="n">start2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;Start&quot;</span><span class="p">)][:</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">end1</span><span class="p">,</span> <span class="n">end2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;End&quot;</span><span class="p">)][:</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">columns</span> <span class="o">=</span> <span class="p">(</span><span class="n">start1</span><span class="p">,</span> <span class="n">end1</span><span class="p">,</span> <span class="n">start2</span><span class="p">,</span> <span class="n">end2</span><span class="p">)</span>

        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;columns&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">columns</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_new_position</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.overlap"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.overlap">[docs]</a>    <span class="k">def</span> <span class="nf">overlap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">,</span> <span class="n">invert</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return overlapping intervals.</span>

<span class="sd">        Returns the intervals in self which overlap with those in other.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to find overlaps with.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        how : {&quot;first&quot;, &quot;containment&quot;, False, None}, default &quot;first&quot;</span>

<span class="sd">            What intervals to report. By default reports every interval in self with overlap once.</span>
<span class="sd">            &quot;containment&quot; reports all intervals where the overlapping is contained within it.</span>

<span class="sd">        invert : bool, default False</span>

<span class="sd">            Whether to return the intervals without overlaps.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges with overlapping intervals.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.intersect : report overlapping subintervals</span>
<span class="sd">        PyRanges.set_intersect : set-intersect PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [1, 4, 10],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 9, 11], &quot;ID&quot;: [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | ID         |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |         chr1 |         1 |         3 | a          |</span>
<span class="sd">        |         chr1 |         4 |         9 | b          |</span>
<span class="sd">        |         chr1 |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2 = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [2, 2, 9], &quot;End&quot;: [3, 9, 10]})</span>
<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         2 |         3 |</span>
<span class="sd">        | chr1         |         2 |         9 |</span>
<span class="sd">        | chr1         |         9 |        10 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.overlap(gr2)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         1 |         3 | a          |</span>
<span class="sd">        | chr1         |         4 |         9 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.overlap(gr2, how=None)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         1 |         3 | a          |</span>
<span class="sd">        | chr1         |         1 |         3 | a          |</span>
<span class="sd">        | chr1         |         4 |         9 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.overlap(gr2, how=&quot;containment&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         4 |         9 | b          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.overlap(gr2, invert=True)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;other&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;how&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">how</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;invert&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">invert</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="k">if</span> <span class="n">invert</span><span class="p">:</span>
            <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">__ix__</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>

        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_overlap</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">invert</span><span class="p">:</span>
            <span class="n">found_idxs</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="s2">&quot;__ix__&quot;</span><span class="p">,</span> <span class="p">[])</span>
            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="p">[</span><span class="o">~</span><span class="bp">self</span><span class="o">.</span><span class="n">__ix__</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">found_idxs</span><span class="p">)]</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;__ix__&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="PyRanges.pc"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.pc">[docs]</a>    <span class="k">def</span> <span class="nf">pc</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print and return self.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">tostring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span>

        <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="PyRanges.print"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.print">[docs]</a>    <span class="k">def</span> <span class="nf">print</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">merge_position</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print the PyRanges.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        n : int, default 8</span>

<span class="sd">            The number of rows to print.</span>

<span class="sd">        merge_postion : bool, default False</span>

<span class="sd">            Print location in same column to save screen space.</span>

<span class="sd">        sort : bool or str, default False</span>

<span class="sd">            Sort the PyRanges before printing. Will print chromosomsomes or strands interleaved on</span>
<span class="sd">            sort columns.</span>

<span class="sd">        formatting : dict, default None</span>

<span class="sd">            Formatting options per column.</span>

<span class="sd">        chain : False</span>

<span class="sd">            Return the PyRanges. Useful to print intermediate results in call chains.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.pc : print chain</span>
<span class="sd">        PyRanges.sp : sort print</span>
<span class="sd">        PyRanges.mp : merge print</span>
<span class="sd">        PyRanges.spc : sort print chain</span>
<span class="sd">        PyRanges.mpc : merge print chain</span>
<span class="sd">        PyRanges.msp : merge sort print</span>
<span class="sd">        PyRanges.mspc : merge sort print chain</span>
<span class="sd">        PyRanges.rp : raw print dictionary of DataFrames</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 8, 5000],</span>
<span class="sd">        ...      &#39;End&#39;: [6, 9, 7000], &#39;Name&#39;: [&#39;i1&#39;, &#39;i3&#39;, &#39;i2&#39;],</span>
<span class="sd">        ...      &#39;Score&#39;: [1.1, 2.3987, 5.9999995], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |       Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (float64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-------------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | i1         |      1.1    | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | i3         |      2.3987 | +            |</span>
<span class="sd">        | chr1         |      5000 |      7000 | i2         |      6      | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.print(formatting={&quot;Start&quot;: &quot;{:,}&quot;, &quot;Score&quot;: &quot;{:.2f}&quot;})</span>
<span class="sd">        +--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |       Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (float64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-------------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | i1         |         1.1 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | i3         |         2.4 | +            |</span>
<span class="sd">        | chr1         |     5,000 |      7000 | i2         |         6   | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.print(merge_position=True) # gr.mp()</span>
<span class="sd">        +--------------------+------------+-------------+</span>
<span class="sd">        | - Position -       | Name       |       Score |</span>
<span class="sd">        | (Multiple types)   | (object)   |   (float64) |</span>
<span class="sd">        |--------------------+------------+-------------|</span>
<span class="sd">        | chr1 3-6 +         | i1         |      1.1    |</span>
<span class="sd">        | chr1 8-9 +         | i3         |      2.3987 |</span>
<span class="sd">        | chr1 5000-7000 -   | i2         |      6      |</span>
<span class="sd">        +--------------------+------------+-------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; chipseq = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; chipseq</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        To interleave strands in output, use print with `sort=True`:</span>

<span class="sd">        &gt;&gt;&gt; chipseq.print(sort=True, n=20) # chipseq.sp()</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 1325303   | 1325328   | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 1541598   | 1541623   | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 1599121   | 1599146   | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 1820285   | 1820310   | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 2448322   | 2448347   | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 3046141   | 3046166   | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 3437168   | 3437193   | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 3504032   | 3504057   | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 3637087   | 3637112   | U0         | 0         | -            |</span>
<span class="sd">        | chr1         | 3681903   | 3681928   | U0         | 0         | -            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 15548022  | 15548047  | U0         | 0         | +            |</span>
<span class="sd">        | chrY         | 16045242  | 16045267  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 16495497  | 16495522  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 21559181  | 21559206  | U0         | 0         | +            |</span>
<span class="sd">        | chrY         | 21707662  | 21707687  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 21751211  | 21751236  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 21910706  | 21910731  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 22054002  | 22054027  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 22210637  | 22210662  | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome, Start, End and Strand.</span>

<span class="sd">        &gt;&gt;&gt; pr.data.chromsizes().print()</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   | Start     | End       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         | 0         | 249250621 |</span>
<span class="sd">        | chr2         | 0         | 243199373 |</span>
<span class="sd">        | chr3         | 0         | 198022430 |</span>
<span class="sd">        | chr4         | 0         | 191154276 |</span>
<span class="sd">        | ...          | ...       | ...       |</span>
<span class="sd">        | chr22        | 0         | 51304566  |</span>
<span class="sd">        | chrM         | 0         | 16571     |</span>
<span class="sd">        | chrX         | 0         | 155270560 |</span>
<span class="sd">        | chrY         | 0         | 59373566  |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 25 rows and 3 columns from 25 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">s</span> <span class="o">=</span> <span class="n">tostring</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span>
            <span class="n">merge_position</span><span class="o">=</span><span class="n">merge_position</span><span class="p">,</span>
            <span class="n">sort</span><span class="o">=</span><span class="n">sort</span><span class="p">,</span>
            <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="PyRanges.rp"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.rp">[docs]</a>    <span class="k">def</span> <span class="nf">rp</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print dict of DataFrames.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dfs</span><span class="p">)</span></div>



<div class="viewcode-block" id="PyRanges.rpc"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.rpc">[docs]</a>    <span class="k">def</span> <span class="nf">rpc</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print dict of DataFrames and return self.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dfs</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span></div>

<div class="viewcode-block" id="PyRanges.sample"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.sample">[docs]</a>    <span class="k">def</span> <span class="nf">sample</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Subsample arbitrary rows of PyRanges.</span>

<span class="sd">        If n is larger than length of PyRanges, replace must be True.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        n : int, default 8</span>

<span class="sd">            Number of rows to return</span>

<span class="sd">        replace : bool, False</span>

<span class="sd">            Reuse rows.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; np.random.seed(0)</span>
<span class="sd">        &gt;&gt;&gt; gr.sample(n=3)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr2         |  76564764 |  76564789 | U0         |         0 | +            |</span>
<span class="sd">        | chr3         | 185739979 | 185740004 | U0         |         0 | -            |</span>
<span class="sd">        | chr20        |  40373657 |  40373682 | U0         |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.sample(10001)</span>
<span class="sd">        Traceback (most recent call last):</span>
<span class="sd">        ...</span>
<span class="sd">        ValueError: Cannot take a larger sample than population when &#39;replace=False&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">subsetter</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="n">subsetter</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">subsetter</span><span class="p">]</span></div>


<div class="viewcode-block" id="PyRanges.set_intersect"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.set_intersect">[docs]</a>    <span class="k">def</span> <span class="nf">set_intersect</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">new_pos</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return set-theoretical intersection.</span>

<span class="sd">        Like intersect, but both PyRanges are merged first.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to set-intersect.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        how : {None, &quot;first&quot;, &quot;last&quot;, &quot;containment&quot;}, default None, i.e. all</span>

<span class="sd">            What intervals to report. By default reports all overlapping intervals. &quot;containment&quot;</span>
<span class="sd">            reports intervals where the overlapping is contained within it.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges with overlapping subintervals.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.intersect : find overlapping subintervals</span>
<span class="sd">        PyRanges.overlap : report overlapping intervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [1, 4, 10],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 9, 11], &quot;ID&quot;: [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | ID         |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |         chr1 |         1 |         3 | a          |</span>
<span class="sd">        |         chr1 |         4 |         9 | b          |</span>
<span class="sd">        |         chr1 |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2 = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [2, 2, 9], &quot;End&quot;: [3, 9, 10]})</span>
<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |         chr1 |         2 |         3 |</span>
<span class="sd">        |         chr1 |         2 |         9 |</span>
<span class="sd">        |         chr1 |         9 |        10 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.set_intersect(gr2)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         2 |         3 |</span>
<span class="sd">        | chr1         |         4 |         9 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        In this simple unstranded case, this is the same as the below:</span>

<span class="sd">        &gt;&gt;&gt; gr.merge().intersect(gr2.merge())</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         2 |         3 |</span>
<span class="sd">        | chr1         |         4 |         9 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.set_intersect(gr2, how=&quot;containment&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         4 |         9 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">,</span> <span class="s2">&quot;how&quot;</span><span class="p">:</span> <span class="n">how</span><span class="p">,</span> <span class="s2">&quot;nb_cpu&quot;</span><span class="p">:</span> <span class="n">nb_cpu</span><span class="p">,</span> <span class="s2">&quot;new_pos&quot;</span><span class="p">:</span> <span class="n">new_pos</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">strandedness</span> <span class="k">else</span> <span class="kc">False</span>
        <span class="n">self_clusters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>
        <span class="n">other_clusters</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_intersection</span><span class="p">,</span> <span class="n">self_clusters</span><span class="p">,</span> <span class="n">other_clusters</span><span class="p">,</span>
                            <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.set_union"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.set_union">[docs]</a>    <span class="k">def</span> <span class="nf">set_union</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return set-theoretical union.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        other : PyRanges</span>

<span class="sd">            PyRanges to do union with.</span>

<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            A PyRanges with the union of intervals.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.set_intersect : set-theoretical intersection</span>
<span class="sd">        PyRanges.overlap : report overlapping intervals</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [1, 4, 10],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 9, 11], &quot;ID&quot;: [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        |   Chromosome |     Start |       End | ID         |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        |         chr1 |         1 |         3 | a          |</span>
<span class="sd">        |         chr1 |         4 |         9 | b          |</span>
<span class="sd">        |         chr1 |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2 = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [2, 2, 9], &quot;End&quot;: [3, 9, 10]})</span>
<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |         chr1 |         2 |         3 |</span>
<span class="sd">        |         chr1 |         2 |         9 |</span>
<span class="sd">        |         chr1 |         9 |        10 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.set_union(gr2)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         1 |        11 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">empty</span> <span class="ow">and</span> <span class="n">other</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">()</span>

        <span class="n">strand</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">strandedness</span> <span class="k">else</span> <span class="kc">False</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span>
            <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unstrand</span><span class="p">()</span>
            <span class="n">other</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">unstrand</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">strandedness</span> <span class="o">==</span> <span class="s2">&quot;opposite&quot;</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">other</span><span class="p">):</span>
            <span class="n">other</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">other</span><span class="o">.</span><span class="n">Strand</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">Strand</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="s2">&quot;+&quot;</span><span class="p">:</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="s2">&quot;+&quot;</span><span class="p">})</span>


        <span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">],</span> <span class="n">strand</span><span class="p">)</span>

        <span class="n">gr</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gr</span></div>


<div class="viewcode-block" id="PyRanges.sort"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.sort">[docs]</a>    <span class="k">def</span> <span class="nf">sort</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sort by position or columns.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        by : str or list of str, default None</span>

<span class="sd">            Columns to sort by. Default is Start and End.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        Since a PyRanges contains multiple DataFrames, the sorting only happens within dataframes.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            Sorted PyRanges</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.multioverlap : find overlaps with multiple PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.f1()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.split(between=True)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand     |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +          |</span>
<span class="sd">        | chr1         |         6 |         8 | +          |</span>
<span class="sd">        | chr1         |         8 |         9 | +          |</span>
<span class="sd">        | chr1         |         5 |         7 | -          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.split(strand=False)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         3 |         5 |</span>
<span class="sd">        | chr1         |         5 |         6 |</span>
<span class="sd">        | chr1         |         6 |         7 |</span>
<span class="sd">        | chr1         |         8 |         9 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 4 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.split(strand=False, between=True)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         3 |         5 |</span>
<span class="sd">        | chr1         |         5 |         6 |</span>
<span class="sd">        | chr1         |         6 |         7 |</span>
<span class="sd">        | chr1         |         7 |         8 |</span>
<span class="sd">        | chr1         |         8 |         9 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 5 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.sort</span> <span class="kn">import</span> <span class="n">_sort</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
        <span class="k">if</span> <span class="n">by</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;by&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">by</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span>
            <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_sort</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span></div>


<div class="viewcode-block" id="PyRanges.sp"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.sp">[docs]</a>    <span class="k">def</span> <span class="nf">sp</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sort on location and print.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">tostring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span></div>

<div class="viewcode-block" id="PyRanges.spc"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.spc">[docs]</a>    <span class="k">def</span> <span class="nf">spc</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Sort on location, print and return self.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.print : print PyRanges.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">tostring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">formatting</span><span class="o">=</span><span class="n">formatting</span><span class="p">))</span>

        <span class="k">return</span> <span class="bp">self</span></div>


<div class="viewcode-block" id="PyRanges.slack"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.slack">[docs]</a>    <span class="k">def</span> <span class="nf">slack</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">slack</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot; Deprecated: this function has been moved to Pyranges.extend&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">slack</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">spliced_subsequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot; Get subsequences of the intervals, using coordinates mapping to spliced transcripts (without introns)</span>
<span class="sd">        </span>
<span class="sd">        The returned intervals are subregions of self, cut according to specifications.</span>
<span class="sd">        Start and end are relative to the 5&#39; end: 0 means the leftmost nucleotide for + strand</span>
<span class="sd">        intervals, while it means the rightmost one for - strand.</span>
<span class="sd">        This method also allows to manipulate groups of intervals (e.g. exons belonging to same transcripts)</span>
<span class="sd">        through the &#39;by&#39; argument. When using it, start and end refer to the spliced transcript coordinates, </span>
<span class="sd">        meaning that introns are in the count.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        start : int</span>
<span class="sd">            Start of subregion, 0-based and included, counting from the 5&#39; end.</span>
<span class="sd">            Use a negative int to count from the 3&#39;  (e.g. -1 is the last nucleotide)</span>

<span class="sd">        end : int, default None</span>
<span class="sd">            End of subregion, 0-based and excluded, counting from the 5&#39; end. </span>
<span class="sd">            Use a negative int to count from the 3&#39;  (e.g. -1 is the last nucleotide)</span>
<span class="sd">            If None, the existing 3&#39; end is returned.</span>

<span class="sd">        by : list of str, default None</span>
<span class="sd">            intervals are grouped by this/these ID column(s) beforehand, e.g. exons belonging to same transcripts</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>
<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>

<span class="sd">        PyRanges</span>
<span class="sd">            Subregion of self, subsequenced as specified by arguments</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        If the request goes out of bounds (e.g. requesting 100 nts for a 90nt region), only the existing portion is returned</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>
<span class="sd">        subsequence : analogous to this method, but input coordinates refer to the unspliced transcript</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; p  = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 2, 2, 3],</span>
<span class="sd">        ...                   &quot;Strand&quot;: [&quot;+&quot;, &quot;+&quot;, &quot;-&quot;, &quot;-&quot;, &quot;+&quot;],</span>
<span class="sd">        ...                   &quot;Start&quot;: [1, 40, 10, 70, 140],</span>
<span class="sd">        ...                   &quot;End&quot;: [11, 60, 25, 80, 152],</span>
<span class="sd">        ...                   &quot;transcript_id&quot;:[&quot;t1&quot;, &quot;t1&quot;, &quot;t2&quot;, &quot;t2&quot;, &quot;t3&quot;] })</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |         1 |        11 | t1              |</span>
<span class="sd">        |            1 | +            |        40 |        60 | t1              |</span>
<span class="sd">        |            2 | -            |        10 |        25 | t2              |</span>
<span class="sd">        |            2 | -            |        70 |        80 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       152 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        # Get the first 15 nucleotides of *each spliced transcript*, grouping exons by transcript_id:</span>
<span class="sd">        &gt;&gt;&gt; p.spliced_subsequence(0, 15, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int64) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |         1 |        11 | t1              |</span>
<span class="sd">        |            1 | +            |        40 |        45 | t1              |</span>
<span class="sd">        |            2 | -            |        20 |        25 | t2              |</span>
<span class="sd">        |            2 | -            |        70 |        80 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       152 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.        </span>

<span class="sd">        # Get the last 20 nucleotides of each spliced transcript:</span>
<span class="sd">        &gt;&gt;&gt; p.spliced_subsequence(-20, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int64) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |        40 |        60 | t1              |</span>
<span class="sd">        |            2 | -            |        10 |        25 | t2              |</span>
<span class="sd">        |            2 | -            |        70 |        75 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       155 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        # Get region from 25 to 60 of each spliced transcript, or their existing subportion:</span>
<span class="sd">        &gt;&gt;&gt; p.spliced_subsequence(25, 60, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |        55 |        60 | t1              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 1 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        # Get region of each spliced transcript which excludes their first and last 3 nucleotides:</span>
<span class="sd">        &gt;&gt;&gt; p.spliced_subsequence(3, -3, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |         1 |        11 | t1              |</span>
<span class="sd">        |            1 | +            |        40 |        60 | t1              |</span>
<span class="sd">        |            2 | -            |        10 |        25 | t2              |</span>
<span class="sd">        |            2 | -            |        70 |        80 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       155 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="kn">from</span> <span class="nn">pyranges.methods.spliced_subsequence</span> <span class="kn">import</span> <span class="n">_spliced_subseq</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span><span class="o">=</span><span class="kc">True</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="k">else</span> <span class="kc">False</span>
        
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="n">by</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">end</span><span class="p">})</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_spliced_subseq</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
        
<div class="viewcode-block" id="PyRanges.spliced_subsequence"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.spliced_subsequence">[docs]</a>    <span class="k">def</span> <span class="nf">spliced_subsequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot; Get subsequences of the intervals, using coordinates mapping to spliced transcripts (without introns)</span>
<span class="sd">        </span>
<span class="sd">        The returned intervals are subregions of self, cut according to specifications.</span>
<span class="sd">        Start and end are relative to the 5&#39; end: 0 means the leftmost nucleotide for + strand</span>
<span class="sd">        intervals, while it means the rightmost one for - strand.</span>
<span class="sd">        This method also allows to manipulate groups of intervals (e.g. exons belonging to same transcripts)</span>
<span class="sd">        through the &#39;by&#39; argument. When using it, start and end refer to the spliced transcript coordinates, </span>
<span class="sd">        meaning that introns are in the count.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        start : int</span>
<span class="sd">            Start of subregion, 0-based and included, counting from the 5&#39; end.</span>
<span class="sd">            Use a negative int to count from the 3&#39;  (e.g. -1 is the last nucleotide)</span>

<span class="sd">        end : int, default None</span>
<span class="sd">            End of subregion, 0-based and excluded, counting from the 5&#39; end. </span>
<span class="sd">            If None, the existing 3&#39; end is returned.</span>

<span class="sd">        by : list of str, default None</span>
<span class="sd">            intervals are grouped by this/these ID column(s) beforehand, e.g. exons belonging to same transcripts</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>
<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>

<span class="sd">        PyRanges</span>
<span class="sd">            Subregion of self, subsequenced as specified by arguments</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        If the request goes out of bounds (e.g. requesting 100 nts for a 90nt region), only the existing portion is returned</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>
<span class="sd">        subsequence : analogous to this method, but input coordinates refer to the unspliced transcript</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="kn">from</span> <span class="nn">pyranges.methods.spliced_subsequence</span> <span class="kn">import</span> <span class="n">_spliced_subseq</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span><span class="o">=</span><span class="kc">True</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="k">else</span> <span class="kc">False</span>
        
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="n">by</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">end</span><span class="p">})</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_spliced_subseq</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>
            
<div class="viewcode-block" id="PyRanges.split"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.split">[docs]</a>    <span class="k">def</span> <span class="nf">split</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">between</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Split into non-overlapping intervals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to ignore strand information if PyRanges is stranded.</span>

<span class="sd">        between : bool, default False</span>

<span class="sd">            Include lengths between intervals.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with intervals split at overlap points.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.multioverlap : find overlaps with multiple PyRanges</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 5, 5, 11],</span>
<span class="sd">        ...       &#39;End&#39;: [6, 9, 7, 12], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | +            |</span>
<span class="sd">        | chr1         |         5 |         9 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | -            |</span>
<span class="sd">        | chr1         |        11 |        12 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.split()</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand     |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         3 |         5 | +          |</span>
<span class="sd">        | chr1         |         5 |         6 | +          |</span>
<span class="sd">        | chr1         |         6 |         9 | +          |</span>
<span class="sd">        | chr1         |         5 |         7 | -          |</span>
<span class="sd">        | chr1         |        11 |        12 | -          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.split(between=True)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand     |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         3 |         5 | +          |</span>
<span class="sd">        | chr1         |         5 |         6 | +          |</span>
<span class="sd">        | chr1         |         6 |         9 | +          |</span>
<span class="sd">        | chr1         |         5 |         7 | -          |</span>
<span class="sd">        | chr1         |         7 |        11 | -          |</span>
<span class="sd">        | chr1         |        11 |        12 | -          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Stranded PyRanges object has 6 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.split(strand=False)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         3 |         5 |</span>
<span class="sd">        | chr1         |         5 |         6 |</span>
<span class="sd">        | chr1         |         6 |         7 |</span>
<span class="sd">        | chr1         |         7 |         9 |</span>
<span class="sd">        | chr1         |        11 |        12 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 5 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.split(strand=False, between=True)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (object)     |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         3 |         5 |</span>
<span class="sd">        | chr1         |         5 |         6 |</span>
<span class="sd">        | chr1         |         6 |         7 |</span>
<span class="sd">        | chr1         |         7 |         9 |</span>
<span class="sd">        | chr1         |         9 |        11 |</span>
<span class="sd">        | chr1         |        11 |        12 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 6 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">})</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.split</span> <span class="kn">import</span> <span class="n">_split</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_split</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">split</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">between</span><span class="p">:</span>
            <span class="n">strandedness</span> <span class="o">=</span> <span class="s2">&quot;same&quot;</span> <span class="k">if</span> <span class="n">strand</span> <span class="k">else</span> <span class="kc">False</span>
            <span class="n">split</span> <span class="o">=</span> <span class="n">split</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="n">strandedness</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">split</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">stranded</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Whether PyRanges has (valid) strand info.</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        A PyRanges can have invalid values in the Strand-column. It is not considered stranded.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.strands : return the strands</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d =  {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...       &#39;End&#39;: [5, 8], &#39;Strand&#39;: [&#39;+&#39;, &#39;.&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         5 | +            |</span>
<span class="sd">        | chr1         |         6 |         8 | .            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        Considered unstranded due to these Strand values: &#39;.&#39;</span>

<span class="sd">        &gt;&gt;&gt; gr.stranded</span>
<span class="sd">        False</span>

<span class="sd">        &gt;&gt;&gt; &quot;Strand&quot; in gr.columns</span>
<span class="sd">        True</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">keys</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">keys</span><span class="p">):</span>
            <span class="c1"># so that stranded ops work with empty dataframes</span>
            <span class="k">return</span> <span class="kc">True</span>

        <span class="n">key</span> <span class="o">=</span> <span class="n">keys</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">return</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">strands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return strands.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        If the strand-column contains an invalid value, [] is returned.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.stranded : whether has valid strand info</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; d =  {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...       &#39;End&#39;: [5, 8], &#39;Strand&#39;: [&#39;+&#39;, &#39;.&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         5 | +            |</span>
<span class="sd">        | chr1         |         6 |         8 | .            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        Considered unstranded due to these Strand values: &#39;.&#39;</span>

<span class="sd">        &gt;&gt;&gt; gr.strands</span>
<span class="sd">        []</span>

<span class="sd">        &gt;&gt;&gt; gr.Strand.drop_duplicates().to_list()</span>
<span class="sd">        [&#39;+&#39;, &#39;.&#39;]</span>

<span class="sd">        &gt;&gt;&gt; gr.Strand = [&quot;+&quot;, &quot;-&quot;]</span>
<span class="sd">        &gt;&gt;&gt; gr.strands</span>
<span class="sd">        [&#39;+&#39;, &#39;-&#39;]</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[]</span>

        <span class="k">return</span> <span class="n">natsorted</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">k</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="p">()]))</span>


<div class="viewcode-block" id="PyRanges.subset"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.subset">[docs]</a>    <span class="k">def</span> <span class="nf">subset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a subset of the rows.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        f : function</span>
<span class="sd">            Function which returns boolean Series equal to length of df.</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        nb_cpu : int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        PyRanges can also be subsetted directly with a boolean Series. This function is slightly</span>
<span class="sd">        faster, but more cumbersome.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges subset on rows.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.f1()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.subset(lambda df: df.Start &gt; 4)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        Also possible:</span>

<span class="sd">        &gt;&gt;&gt; gr[gr.Start &gt; 4]</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="sd">        | chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span>
            <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unstrand</span><span class="p">()</span>


        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">})</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">()</span>

        <span class="n">first_result</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="k">assert</span> <span class="n">first_result</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="nb">bool</span><span class="p">,</span> <span class="s2">&quot;result of subset function must be bool, but is </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">first_result</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">result</span><span class="p">]</span></div>

<div class="viewcode-block" id="PyRanges.subsequence"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.subsequence">[docs]</a>    <span class="k">def</span> <span class="nf">subsequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot; Get subsequences of the intervals.</span>
<span class="sd">        </span>
<span class="sd">        The returned intervals are subregions of self, cut according to specifications.</span>
<span class="sd">        Start and end are relative to the 5&#39; end: 0 means the leftmost nucleotide for + strand</span>
<span class="sd">        intervals, while it means the rightmost one for - strand.</span>
<span class="sd">        This method also allows to manipulate groups of intervals (e.g. exons belonging to same transcripts)</span>
<span class="sd">        through the &#39;by&#39; argument. When using it, start and end refer to the unspliced transcript coordinates, </span>
<span class="sd">        meaning that introns are included in the count.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        start : int</span>
<span class="sd">            Start of subregion, 0-based and included, counting from the 5&#39; end.</span>
<span class="sd">            Use a negative int to count from the 3&#39;  (e.g. -1 is the last nucleotide)</span>

<span class="sd">        end : int, default None</span>
<span class="sd">            End of subregion, 0-based and excluded, counting from the 5&#39; end. </span>
<span class="sd">            Use a negative int to count from the 3&#39;  (e.g. -1 is the last nucleotide)</span>

<span class="sd">            If None, the existing 3&#39; end is returned.</span>

<span class="sd">        by : list of str, default None</span>
<span class="sd">            intervals are grouped by this/these ID column(s) beforehand, e.g. exons belonging to same transcripts</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>
<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>

<span class="sd">        PyRanges</span>
<span class="sd">            Subregion of self, subsequenced as specified by arguments</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        If the request goes out of bounds (e.g. requesting 100 nts for a 90nt region), only the existing portion is returned</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>
<span class="sd">        spliced_subsequence : analogous to this method, but intronic regions are not counted, so that input coordinates refer to the spliced transcript        </span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>
<span class="sd">        &gt;&gt;&gt; p  = pr.from_dict({&quot;Chromosome&quot;: [1, 1, 2, 2, 3],</span>
<span class="sd">        ...                   &quot;Strand&quot;: [&quot;+&quot;, &quot;+&quot;, &quot;-&quot;, &quot;-&quot;, &quot;+&quot;],</span>
<span class="sd">        ...                   &quot;Start&quot;: [1, 40, 2, 30, 140],</span>
<span class="sd">        ...                   &quot;End&quot;: [20, 60, 13, 45, 155],</span>
<span class="sd">        ...                   &quot;transcript_id&quot;:[&quot;t1&quot;, &quot;t1&quot;, &quot;t2&quot;, &quot;t2&quot;, &quot;t3&quot;] })</span>
<span class="sd">        &gt;&gt;&gt; p</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |         1 |        20 | t1              |</span>
<span class="sd">        |            1 | +            |        40 |        60 | t1              |</span>
<span class="sd">        |            2 | -            |         2 |        13 | t2              |</span>
<span class="sd">        |            2 | -            |        30 |        45 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       155 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        # Get the first 10 nucleotides (at the 5&#39;) of *each interval* (each line of the dataframe):</span>
<span class="sd">        &gt;&gt;&gt; p.subsequence(0, 10)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |         1 |        20 | t1              |</span>
<span class="sd">        |            1 | +            |        40 |        60 | t1              |</span>
<span class="sd">        |            2 | -            |         2 |        13 | t2              |</span>
<span class="sd">        |            2 | -            |        30 |        45 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       155 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        # Get the first 10 nucleotides of *each transcript*, grouping exons by transcript_id:</span>
<span class="sd">        &gt;&gt;&gt; p.subsequence(0, 10, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |         1 |        11 | t1              |</span>
<span class="sd">        |            2 | -            |        35 |        45 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       150 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        # Get the last 20 nucleotides of each transcript:</span>
<span class="sd">        &gt;&gt;&gt; p.subsequence(-20, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |        40 |        60 | t1              |</span>
<span class="sd">        |            2 | -            |        30 |        39 | t2              |</span>
<span class="sd">        |            2 | -            |         2 |        13 | t2              |</span>
<span class="sd">        |            3 | +            |       140 |       150 | t3              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 5 columns from 3 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        # Get region from 30 to 330 of each transcript, or their existing subportion:</span>
<span class="sd">        &gt;&gt;&gt; p.subsequence(30, 300, by=&#39;transcript_id&#39;)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        |   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="sd">        |   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+-----------------|</span>
<span class="sd">        |            1 | +            |        51 |        60 | t1              |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 1 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">pyranges.methods.subsequence</span> <span class="kn">import</span> <span class="n">_subseq</span>

        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;by&quot;</span><span class="p">:</span> <span class="n">by</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">end</span><span class="p">})</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_subseq</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.subtract"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.subtract">[docs]</a>    <span class="k">def</span> <span class="nf">subtract</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Subtract intervals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        strandedness : {None, &quot;same&quot;, &quot;opposite&quot;, False}, default None, i.e. auto</span>

<span class="sd">            Whether to compare PyRanges on the same strand, the opposite or ignore strand</span>
<span class="sd">            information. The default, None, means use &quot;same&quot; if both PyRanges are strande,</span>
<span class="sd">            otherwise ignore the strand information.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        pyranges.PyRanges.overlap : use with invert=True to return all intervals without overlap</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [1, 4, 10],</span>
<span class="sd">        ...                    &quot;End&quot;: [3, 9, 11], &quot;ID&quot;: [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;]})</span>
<span class="sd">        &gt;&gt;&gt; gr2 = pr.from_dict({&quot;Chromosome&quot;: [&quot;chr1&quot;] * 3, &quot;Start&quot;: [2, 2, 9], &quot;End&quot;: [3, 9, 10]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         1 |         3 | a          |</span>
<span class="sd">        | chr1         |         4 |         9 | b          |</span>
<span class="sd">        | chr1         |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr2</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         2 |         3 |</span>
<span class="sd">        | chr1         |         2 |         9 |</span>
<span class="sd">        | chr1         |         9 |        10 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.subtract(gr2)</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | ID         |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------|</span>
<span class="sd">        | chr1         |         1 |         2 | a          |</span>
<span class="sd">        | chr1         |        10 |        11 | c          |</span>
<span class="sd">        +--------------+-----------+-----------+------------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.subtraction</span> <span class="kn">import</span> <span class="n">_subtraction</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strandedness&quot;</span><span class="p">:</span> <span class="n">strandedness</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;other&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">strand</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="n">strandedness</span> <span class="k">else</span> <span class="kc">False</span>
        <span class="n">other_clusters</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_overlaps</span><span class="p">(</span><span class="n">other_clusters</span><span class="p">,</span> <span class="n">strandedness</span><span class="o">=</span><span class="n">strandedness</span><span class="p">,</span> <span class="n">overlap_col</span><span class="o">=</span><span class="s2">&quot;__num__&quot;</span><span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">pyrange_apply</span><span class="p">(</span><span class="n">_subtraction</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">other_clusters</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="bp">self</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;__num__&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">result</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;__num__&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.summary"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.summary">[docs]</a>    <span class="k">def</span> <span class="nf">summary</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">to_stdout</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">return_df</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return info.</span>

<span class="sd">        Count refers to the number of intervals, the rest to the lengths.</span>

<span class="sd">        The column &quot;pyrange&quot; describes the data as is. &quot;coverage_forward&quot; and &quot;coverage_reverse&quot;</span>
<span class="sd">        describe the data after strand-specific merging of overlapping intervals.</span>
<span class="sd">        &quot;coverage_unstranded&quot; describes the data after merging, without considering the strands.</span>

<span class="sd">        The row &quot;count&quot; is the number of intervals and &quot;sum&quot; is their total length. The rest describe the lengths of the</span>
<span class="sd">        intervals.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        to_stdout : bool, default True</span>

<span class="sd">            Print summary.</span>

<span class="sd">        return_df : bool, default False</span>

<span class="sd">            Return df with summary.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            None or DataFrame with summary.</span>


<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.ensembl_gtf()[[&quot;Feature&quot;, &quot;gene_id&quot;]]</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_id         |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)        |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-----------------|</span>
<span class="sd">        | 1            | gene         | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | transcript   | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | exon         | 11868     | 12227     | +            | ENSG00000223972 |</span>
<span class="sd">        | 1            | exon         | 12612     | 12721     | +            | ENSG00000223972 |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...             |</span>
<span class="sd">        | 1            | gene         | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        | 1            | transcript   | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        | 1            | exon         | 1179364   | 1179555   | -            | ENSG00000205231 |</span>
<span class="sd">        | 1            | exon         | 1173055   | 1176396   | -            | ENSG00000205231 |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="sd">        Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.summary()</span>
<span class="sd">        +-------+------------------+--------------------+--------------------+-----------------------+</span>
<span class="sd">        |       |          pyrange |   coverage_forward |   coverage_reverse |   coverage_unstranded |</span>
<span class="sd">        |-------+------------------+--------------------+--------------------+-----------------------|</span>
<span class="sd">        | count |   2446           |               39   |               23   |                  32   |</span>
<span class="sd">        | mean  |   2291.92        |             7058.1 |            30078.6 |               27704.2 |</span>
<span class="sd">        | std   |  11906.9         |            10322.3 |            59467.7 |               67026.9 |</span>
<span class="sd">        | min   |      1           |               83   |              154   |                  83   |</span>
<span class="sd">        | 25%   |     90           |             1051   |             1204   |                1155   |</span>
<span class="sd">        | 50%   |    138           |             2541   |             6500   |                6343   |</span>
<span class="sd">        | 75%   |    382.25        |             7168   |            23778   |               20650.8 |</span>
<span class="sd">        | max   | 241726           |            43065   |           241726   |              291164   |</span>
<span class="sd">        | sum   |      5.60603e+06 |           275266   |           691807   |              886534   |</span>
<span class="sd">        +-------+------------------+--------------------+--------------------+-----------------------+</span>

<span class="sd">        &gt;&gt;&gt; gr.summary(return_df=True, to_stdout=False)</span>
<span class="sd">                    pyrange  coverage_forward  coverage_reverse  coverage_unstranded</span>
<span class="sd">        count  2.446000e+03         39.000000         23.000000            32.000000</span>
<span class="sd">        mean   2.291918e+03       7058.102564      30078.565217         27704.187500</span>
<span class="sd">        std    1.190685e+04      10322.309347      59467.695265         67026.868647</span>
<span class="sd">        min    1.000000e+00         83.000000        154.000000            83.000000</span>
<span class="sd">        25%    9.000000e+01       1051.000000       1204.000000          1155.000000</span>
<span class="sd">        50%    1.380000e+02       2541.000000       6500.000000          6343.000000</span>
<span class="sd">        75%    3.822500e+02       7168.000000      23778.000000         20650.750000</span>
<span class="sd">        max    2.417260e+05      43065.000000     241726.000000        291164.000000</span>
<span class="sd">        sum    5.606031e+06     275266.000000     691807.000000        886534.000000</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.summary</span> <span class="kn">import</span> <span class="n">_summary</span>

        <span class="k">return</span> <span class="n">_summary</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">to_stdout</span><span class="p">,</span> <span class="n">return_df</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.tail"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.tail">[docs]</a>    <span class="k">def</span> <span class="nf">tail</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">8</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the n last rows.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        n : int, default 8</span>

<span class="sd">            Return n rows.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            PyRanges with the n last rows.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.head : return the first rows</span>
<span class="sd">        PyRanges.sample : return random rows</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.tail(3)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chrY         |  13517892 |  13517917 | U0         |         0 | -            |</span>
<span class="sd">        | chrY         |   8010951 |   8010976 | U0         |         0 | -            |</span>
<span class="sd">        | chrY         |   7405376 |   7405401 | U0         |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">subsetter</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="n">subsetter</span><span class="p">[(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="n">n</span><span class="p">):]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="n">subsetter</span><span class="p">]</span></div>


<div class="viewcode-block" id="PyRanges.tile"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.tile">[docs]</a>    <span class="k">def</span> <span class="nf">tile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tile_size</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return overlapping genomic tiles.</span>

<span class="sd">        The genome is divided into bookended tiles of length `tile_size` and one is returned per</span>
<span class="sd">        overlapping interval.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        tile_size : int</span>
<span class="sd">            Length of the tiles.</span>

<span class="sd">        overlap : bool, default False</span>

<span class="sd">            Add column of nucleotide overlap to each tile.</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            Tiled PyRanges.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.PyRanges.window : divide intervals into windows</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.ensembl_gtf()[[&quot;Feature&quot;, &quot;gene_name&quot;]]</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="sd">        | 1            | gene         | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | transcript   | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | exon         | 11868     | 12227     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | exon         | 12612     | 12721     | +            | DDX11L1     |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="sd">        | 1            | gene         | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | transcript   | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1179364   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1173055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.tile(200)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="sd">        | 1            | gene         | 11800     | 12000     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | gene         | 12000     | 12200     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | gene         | 12200     | 12400     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | gene         | 12400     | 12600     | +            | DDX11L1     |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="sd">        | 1            | exon         | 1175600   | 1175800   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1175800   | 1176000   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1176000   | 1176200   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1176200   | 1176400   | -            | TTLL10-AS1  |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        Stranded PyRanges object has 30,538 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.tile(100, overlap=True)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+---------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_name   | TileOverlap   |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    | (int32)       |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-------------+---------------|</span>
<span class="sd">        | 1            | gene         | 11800     | 11900     | +            | DDX11L1     | 32            |</span>
<span class="sd">        | 1            | gene         | 11900     | 12000     | +            | DDX11L1     | 100           |</span>
<span class="sd">        | 1            | gene         | 12000     | 12100     | +            | DDX11L1     | 100           |</span>
<span class="sd">        | 1            | gene         | 12100     | 12200     | +            | DDX11L1     | 100           |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...         | ...           |</span>
<span class="sd">        | 1            | exon         | 1176000   | 1176100   | -            | TTLL10-AS1  | 100           |</span>
<span class="sd">        | 1            | exon         | 1176100   | 1176200   | -            | TTLL10-AS1  | 100           |</span>
<span class="sd">        | 1            | exon         | 1176200   | 1176300   | -            | TTLL10-AS1  | 100           |</span>
<span class="sd">        | 1            | exon         | 1176300   | 1176400   | -            | TTLL10-AS1  | 96            |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+---------------+</span>
<span class="sd">        Stranded PyRanges object has 58,516 rows and 7 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.windows</span> <span class="kn">import</span> <span class="n">_tiles</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;overlap&quot;</span><span class="p">:</span> <span class="n">overlap</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;tile_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tile_size</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_tiles</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></div>

<div class="viewcode-block" id="PyRanges.to_example"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_example">[docs]</a>    <span class="k">def</span> <span class="nf">to_example</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return as dict.</span>

<span class="sd">        Used for easily creating examples for copy and pasting.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        n : int, default 10</span>
<span class="sd">            Number of rows. Half is taken from the start, the other half from the end.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.from_dict : create PyRanges from dict</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.chipseq()</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="sd">        | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="sd">        | chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="sd">        | ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="sd">        | chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="sd">        | chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; d = gr.to_example(n=4)</span>
<span class="sd">        &gt;&gt;&gt; d</span>
<span class="sd">        {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chrY&#39;, &#39;chrY&#39;], &#39;Start&#39;: [212609534, 169887529, 8010951, 7405376], &#39;End&#39;: [212609559, 169887554, 8010976, 7405401], &#39;Name&#39;: [&#39;U0&#39;, &#39;U0&#39;, &#39;U0&#39;, &#39;U0&#39;], &#39;Score&#39;: [0, 0, 0, 0], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; pr.from_dict(d)</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="sd">        | chr1         | 212609534 | 212609559 | U0         |         0 | +            |</span>
<span class="sd">        | chr1         | 169887529 | 169887554 | U0         |         0 | +            |</span>
<span class="sd">        | chrY         |   8010951 |   8010976 | U0         |         0 | -            |</span>
<span class="sd">        | chrY         |   7405376 |   7405401 | U0         |         0 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 4 rows and 6 columns from 2 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">nrows_half</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="n">first</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">nrows_half</span><span class="p">)</span>
            <span class="n">last</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="n">nrows_half</span><span class="p">)</span>
            <span class="n">example</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">first</span><span class="p">,</span> <span class="n">last</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">example</span> <span class="o">=</span> <span class="bp">self</span>

        <span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">example</span><span class="p">,</span> <span class="n">c</span><span class="p">))</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">example</span><span class="o">.</span><span class="n">columns</span><span class="p">}</span>

        <span class="k">return</span> <span class="n">d</span></div>

<div class="viewcode-block" id="PyRanges.three_end"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.three_end">[docs]</a>    <span class="k">def</span> <span class="nf">three_end</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the 3&#39;-end.</span>

<span class="sd">        The 3&#39;-end is the start of intervals on the reverse strand and the end of intervals on the</span>
<span class="sd">        forward strand.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>
<span class="sd">            PyRanges with the 3&#39;.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        PyRanges.five_end : return the five prime end</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d =  {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...       &#39;End&#39;: [5, 8], &#39;Strand&#39;: [&#39;+&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         5 | +            |</span>
<span class="sd">        | chr1         |         6 |         8 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.three_end()</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         5 |         6 | +            |</span>
<span class="sd">        | chr1         |         6 |         7 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">,</span> <span class="s2">&quot;Need stranded pyrange to find 3&#39;.&quot;</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="n">fill_kwargs</span><span class="p">({</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span>
            <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_tes</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span></div>


<span class="c1">#     def to_bam(self, path=None, header=None, chromosome_sizes=None, chain=False):</span>

<span class="c1">#         r&quot;&quot;&quot;Write to bam.</span>

<span class="c1">#         Parameters</span>
<span class="c1">#         ----------</span>
<span class="c1">#         path : str, default None</span>
<span class="c1">#             Where to write. If None, returns string representation.</span>

<span class="c1">#         keep : bool, default True</span>

<span class="c1">#             Whether to keep all columns, not just Chromosome, Start, End,</span>
<span class="c1">#             Name, Score, Strand when writing.</span>

<span class="c1">#         compression : str, compression type to use, by default infer based on extension.</span>
<span class="c1">#             See pandas.DataFree.to_csv for more info.</span>

<span class="c1">#         header : dict</span>

<span class="c1">#             Header to use in the bamfile. See the pysam docs for how it should look.</span>
<span class="c1">#             Or use the header attribute from another pyasam.AlignmentFile.</span>

<span class="c1">#         chromosome_sizes : PyRanges or dict</span>

<span class="c1">#             If dict: map of chromosome names to chromosome length.</span>

<span class="c1">#         chain : bool, default False</span>
<span class="c1">#             Whether to return the PyRanges after writing.</span>

<span class="c1">#         Note</span>
<span class="c1">#         ----</span>

<span class="c1">#         The following pyranges columns are used when writing:</span>

<span class="c1">#         Chromosome, Start, End, Strand, MapQ, Flag, QueryStart, QueryEnd, Name, Cigar, Quality</span>

<span class="c1">#         Examples</span>
<span class="c1">#         --------</span>

<span class="c1">#         &gt;&gt;&gt; header = {&quot;SQ&quot;: [{&quot;SN&quot;: 1, &quot;LN&quot;: 249250621}]}</span>

<span class="c1">#         &gt;&gt;&gt; c = &#39;&#39;&#39;Name	Flag Chromosome Start End MapQ Cigar QuerySequence Quality</span>
<span class="c1"># read1	115	1	142618765 142618790	255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:65536	ZL:i:25</span>
<span class="c1"># read2	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:214748	ZL:i:25</span>
<span class="c1"># read3	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:2147484	ZL:i:25</span>
<span class="c1"># read4	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:2147483647	ZL:i:25</span>
<span class="c1"># read5	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:-65536	ZL:i:25</span>
<span class="c1"># read6	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:-214748	ZL:i:25</span>
<span class="c1"># read7	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:-2147484	ZL:i:25</span>
<span class="c1"># read8	115	1	142618765 142618790  255	25M	CGACCCACTCCGCCATTTTCATCCG	IIGIIIHIGIIFIIIIIIIGIGIII	NM:i:0ZP:i:-2147483647	ZL:i:25&#39;&#39;&#39;</span>

<span class="c1">#         &gt;&gt;&gt;</span>
<span class="c1">#         &quot;&quot;&quot;</span>




<div class="viewcode-block" id="PyRanges.to_bed"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_bed">[docs]</a>    <span class="k">def</span> <span class="nf">to_bed</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;infer&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Write to bed.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        path : str, default None</span>
<span class="sd">            Where to write. If None, returns string representation.</span>

<span class="sd">        keep : bool, default True</span>

<span class="sd">            Whether to keep all columns, not just Chromosome, Start, End,</span>
<span class="sd">            Name, Score, Strand when writing.</span>

<span class="sd">        compression : str, compression type to use, by default infer based on extension.</span>
<span class="sd">            See pandas.DataFree.to_csv for more info.</span>

<span class="sd">        chain : bool, default False</span>
<span class="sd">            Whether to return the PyRanges after writing.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d =  {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...       &#39;End&#39;: [5, 8], &#39;Strand&#39;: [&#39;+&#39;, &#39;-&#39;], &quot;Gene&quot;: [1, 2]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |      Gene |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        | chr1         |         1 |         5 | +            |         1 |</span>
<span class="sd">        | chr1         |         6 |         8 | -            |         2 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; print(gr.to_bed())</span>
<span class="sd">        chr1	1	5	.	.	+	1</span>
<span class="sd">        chr1	6	8	.	.	-	2</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        Does not include noncanonical bed-column `Gene`:</span>

<span class="sd">        &gt;&gt;&gt; print(gr.to_bed(keep=False))</span>
<span class="sd">        chr1	1	5	.	.	+</span>
<span class="sd">        chr1	6	8	.	.	-</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; gr.to_bed(&quot;test.bed&quot;, chain=True)</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |      Gene |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        | chr1         |         1 |         5 | +            |         1 |</span>
<span class="sd">        | chr1         |         6 |         8 | -            |         2 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; open(&quot;test.bed&quot;).readlines()</span>
<span class="sd">        [&#39;chr1\t1\t5\t.\t.\t+\t1\n&#39;, &#39;chr1\t6\t8\t.\t.\t-\t2\n&#39;]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">pyranges.out</span> <span class="kn">import</span> <span class="n">_to_bed</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">_to_bed</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="n">keep</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">path</span> <span class="ow">and</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="PyRanges.to_bigwig"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_bigwig">[docs]</a>    <span class="k">def</span> <span class="nf">to_bigwig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chromosome_sizes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rpm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">divide</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">value_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">dryrun</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Write regular or value coverage to bigwig.</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        To create one bigwig per strand, subset the PyRanges first.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        path : str</span>

<span class="sd">            Where to write bigwig.</span>

<span class="sd">        chromosome_sizes : PyRanges or dict</span>

<span class="sd">            If dict: map of chromosome names to chromosome length.</span>

<span class="sd">        rpm : True</span>

<span class="sd">            Whether to normalize data by dividing by total number of intervals and multiplying by</span>
<span class="sd">            1e6.</span>

<span class="sd">        divide : bool, default False</span>

<span class="sd">            (Only useful with value_col) Divide value coverage by regular coverage and take log2.</span>

<span class="sd">        value_col : str, default None</span>

<span class="sd">            Name of column to compute coverage of.</span>

<span class="sd">        dryrun : bool, default False</span>

<span class="sd">            Return data that would be written without writing bigwigs.</span>

<span class="sd">        chain : bool, default False</span>
<span class="sd">            Whether to return the PyRanges after writing.</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        Requires pybigwig to be installed.</span>

<span class="sd">        If you require more control over the normalization process, use pyranges.to_bigwig()</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        pyranges.to_bigwig : write pandas DataFrame to bigwig.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d =  {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 4, 6],</span>
<span class="sd">        ...       &#39;End&#39;: [7, 8, 10], &#39;Strand&#39;: [&#39;+&#39;, &#39;-&#39;, &#39;-&#39;],</span>
<span class="sd">        ...       &#39;Value&#39;: [10, 20, 30]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |     Value |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="sd">        |--------------+-----------+-----------+--------------+-----------|</span>
<span class="sd">        | chr1         |         1 |         7 | +            |        10 |</span>
<span class="sd">        | chr1         |         4 |         8 | -            |        20 |</span>
<span class="sd">        | chr1         |         6 |        10 | -            |        30 |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+-----------+</span>
<span class="sd">        Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.to_bigwig(dryrun=True, rpm=False)</span>
<span class="sd">        +--------------+-----------+-----------+-------------+</span>
<span class="sd">        | Chromosome   |     Start |       End |       Score |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (float64) |</span>
<span class="sd">        |--------------+-----------+-----------+-------------|</span>
<span class="sd">        | chr1         |         1 |         4 |           1 |</span>
<span class="sd">        | chr1         |         4 |         6 |           2 |</span>
<span class="sd">        | chr1         |         6 |         7 |           3 |</span>
<span class="sd">        | chr1         |         7 |         8 |           2 |</span>
<span class="sd">        | chr1         |         8 |        10 |           1 |</span>
<span class="sd">        +--------------+-----------+-----------+-------------+</span>
<span class="sd">        Unstranded PyRanges object has 5 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.to_bigwig(dryrun=True, rpm=False, value_col=&quot;Value&quot;)</span>
<span class="sd">        +--------------+-----------+-----------+-------------+</span>
<span class="sd">        | Chromosome   |     Start |       End |       Score |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (float64) |</span>
<span class="sd">        |--------------+-----------+-----------+-------------|</span>
<span class="sd">        | chr1         |         1 |         4 |          10 |</span>
<span class="sd">        | chr1         |         4 |         6 |          30 |</span>
<span class="sd">        | chr1         |         6 |         7 |          60 |</span>
<span class="sd">        | chr1         |         7 |         8 |          50 |</span>
<span class="sd">        | chr1         |         8 |        10 |          30 |</span>
<span class="sd">        +--------------+-----------+-----------+-------------+</span>
<span class="sd">        Unstranded PyRanges object has 5 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.to_bigwig(dryrun=True, rpm=False, value_col=&quot;Value&quot;, divide=True)</span>
<span class="sd">        +--------------+-----------+-----------+-------------+</span>
<span class="sd">        | Chromosome   |     Start |       End |       Score |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |   (float64) |</span>
<span class="sd">        |--------------+-----------+-----------+-------------|</span>
<span class="sd">        | chr1         |         0 |         1 |   nan       |</span>
<span class="sd">        | chr1         |         1 |         4 |     3.32193 |</span>
<span class="sd">        | chr1         |         4 |         6 |     3.90689 |</span>
<span class="sd">        | chr1         |         6 |         7 |     4.32193 |</span>
<span class="sd">        | chr1         |         7 |         8 |     4.64386 |</span>
<span class="sd">        | chr1         |         8 |        10 |     4.90689 |</span>
<span class="sd">        +--------------+-----------+-----------+-------------+</span>
<span class="sd">        Unstranded PyRanges object has 6 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.out</span> <span class="kn">import</span> <span class="n">_to_bigwig</span>

        <span class="k">if</span> <span class="n">chromosome_sizes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">chromosome_sizes</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">()</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">_to_bigwig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">chromosome_sizes</span><span class="p">,</span> <span class="n">rpm</span><span class="p">,</span> <span class="n">divide</span><span class="p">,</span> <span class="n">value_col</span><span class="p">,</span> <span class="n">dryrun</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">dryrun</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span>

        <span class="k">if</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span></div>

<div class="viewcode-block" id="PyRanges.to_csv"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_csv">[docs]</a>    <span class="k">def</span> <span class="nf">to_csv</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;infer&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Write to comma- or other value-separated file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        path : str, default None, i.e. return string representation.</span>

<span class="sd">            Where to write file.</span>

<span class="sd">        sep : str, default &quot;,&quot;</span>

<span class="sd">            String of length 1. Field delimiter for the output file.</span>

<span class="sd">        header : bool, default True</span>

<span class="sd">            Write out the column names.</span>

<span class="sd">        compression : {‘infer’, ‘gzip’, ‘bz2’, ‘zip’, ‘xz’, None}, default &quot;infer&quot;</span>

<span class="sd">            Which compression to use. Uses file extension to infer by default.</span>

<span class="sd">        chain: bool, default False</span>

<span class="sd">            Whether to return the PyRanges after writing.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&quot;Chromosome&quot;: [1] * 3, &quot;Start&quot;: [1, 3, 5], &quot;End&quot;: [4, 6, 9], &quot;Feature&quot;: [&quot;gene&quot;, &quot;exon&quot;, &quot;exon&quot;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; print(gr.to_csv(sep=&quot;\t&quot;))</span>
<span class="sd">        Chromosome	Start	End	Feature</span>
<span class="sd">        1	1	4	gene</span>
<span class="sd">        1	3	6	exon</span>
<span class="sd">        1	5	9	exon</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.out</span> <span class="kn">import</span> <span class="n">_to_csv</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">_to_csv</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="n">sep</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">header</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">path</span> <span class="ow">and</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="PyRanges.to_gff3"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_gff3">[docs]</a>    <span class="k">def</span> <span class="nf">to_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;infer&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Write to General Feature Format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        path : str, default None, i.e. return string representation.</span>

<span class="sd">            Where to write file.</span>

<span class="sd">        compression : {‘infer’, ‘gzip’, ‘bz2’, ‘zip’, ‘xz’, None}, default &quot;infer&quot;</span>

<span class="sd">            Which compression to use. Uses file extension to infer by default.</span>

<span class="sd">        chain: bool, default False</span>

<span class="sd">            Whether to return the PyRanges after writing.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        GTF uses a different naming-convention for columns than PyRanges.</span>
<span class="sd">        This is the mapping between column names:</span>

<span class="sd">        ``{&quot;seqname&quot;: &quot;Chromosome&quot;, &quot;source&quot;: &quot;Source&quot;, &quot;type&quot;: &quot;Feature&quot;, &quot;start&quot;: &quot;Start&quot;, &quot;end&quot;: &quot;End&quot;, &quot;score&quot;: &quot;Score&quot;, &quot;strand&quot;: &quot;Strand&quot;, &quot;phase&quot;: &quot;Frame&quot;, &quot;attributes&quot;: &quot;Attribute&quot;}``</span>

<span class="sd">        All other columns are appended as a field in the attribute string.</span>

<span class="sd">        Nonexisting columns will be added with a &#39;.&#39; to represent the missing values.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        pyranges.read_gff3 : read GFF3 files</span>
<span class="sd">        pyranges.to_gtf : write to GTF format</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&quot;Chromosome&quot;: [1] * 3, &quot;Start&quot;: [1, 3, 5], &quot;End&quot;: [4, 6, 9], &quot;Feature&quot;: [&quot;gene&quot;, &quot;exon&quot;, &quot;exon&quot;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; print(gr.to_gff3())</span>
<span class="sd">        1	.	gene	2	4	.	.	.</span>
<span class="sd">        1	.	exon	4	6	.	.	.</span>
<span class="sd">        1	.	exon	6	9	.	.	.</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; gr.Gene = [1, 2, 3]</span>
<span class="sd">        &gt;&gt;&gt; gr.function = [&quot;a b&quot;, &quot;c&quot;, &quot;def&quot;]</span>
<span class="sd">        &gt;&gt;&gt; print(gr.to_gff3())</span>
<span class="sd">        1	.	gene	2	4	.	.	.	Gene=1;function=a b</span>
<span class="sd">        1	.	exon	4	6	.	.	.	Gene=2;function=c</span>
<span class="sd">        1	.	exon	6	9	.	.	.	Gene=3;function=def</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.out</span> <span class="kn">import</span> <span class="n">_to_gff3</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">_to_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">path</span> <span class="ow">and</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="PyRanges.to_gtf"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_gtf">[docs]</a>    <span class="k">def</span> <span class="nf">to_gtf</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;infer&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Write to Gene Transfer Format.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        path : str, default None, i.e. return string representation.</span>

<span class="sd">            Where to write file.</span>

<span class="sd">        compression : {‘infer’, ‘gzip’, ‘bz2’, ‘zip’, ‘xz’, None}, default &quot;infer&quot;</span>

<span class="sd">            Which compression to use. Uses file extension to infer by default.</span>

<span class="sd">        chain: bool, default False</span>

<span class="sd">            Whether to return the PyRanges after writing.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>

<span class="sd">        GTF uses a different naming-convention for columns than PyRanges.</span>
<span class="sd">        This is the mapping between column names:</span>

<span class="sd">        ``{&quot;seqname&quot;: &quot;Chromosome&quot;, &quot;source&quot;: &quot;Source&quot;, &quot;feature&quot;: &quot;Feature&quot;, &quot;start&quot;: &quot;Start&quot;, &quot;end&quot;: &quot;End&quot;, &quot;score&quot;: &quot;Score&quot;, &quot;strand&quot;: &quot;Strand&quot;, &quot;frame&quot;: &quot;Frame&quot;, &quot;attribute&quot;: &quot;Attribute&quot;}``</span>

<span class="sd">        All other columns are appended as a field in the attribute string.</span>

<span class="sd">        Nonexisting columns will be added with a &#39;.&#39; to represent the missing values.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        pyranges.read_gtf : read GTF files</span>
<span class="sd">        pyranges.to_gff3 : write to GFF3 format</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&quot;Chromosome&quot;: [1] * 3, &quot;Start&quot;: [1, 3, 5], &quot;End&quot;: [4, 6, 9], &quot;Feature&quot;: [&quot;gene&quot;, &quot;exon&quot;, &quot;exon&quot;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; print(gr.to_gtf())</span>
<span class="sd">        1	.	gene	2	4	.	.	.</span>
<span class="sd">        1	.	exon	4	6	.	.	.</span>
<span class="sd">        1	.	exon	6	9	.	.	.</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; gr.name = [&quot;Tim&quot;, &quot;Eric&quot;, &quot;Endre&quot;]</span>
<span class="sd">        &gt;&gt;&gt; gr.prices = [&quot;Cheap&quot;, &quot;Premium&quot;, &quot;Fine European&quot;]</span>
<span class="sd">        &gt;&gt;&gt; print(gr.to_gtf())</span>
<span class="sd">        1	.	gene	2	4	.	.	.	name &quot;Tim&quot;; prices &quot;Cheap&quot;;</span>
<span class="sd">        1	.	exon	4	6	.	.	.	name &quot;Eric&quot;; prices &quot;Premium&quot;;</span>
<span class="sd">        1	.	exon	6	9	.	.	.	name &quot;Endre&quot;; prices &quot;Fine European&quot;;</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.out</span> <span class="kn">import</span> <span class="n">_to_gtf</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">_to_gtf</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">path</span> <span class="ow">and</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="PyRanges.to_rle"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.to_rle">[docs]</a>    <span class="k">def</span> <span class="nf">to_rle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rpm</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return as RleDict.</span>

<span class="sd">        Create collection of Rles representing the coverage or other numerical value.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        value_col : str, default None</span>
<span class="sd">            Numerical column to create RleDict from.</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>
<span class="sd">            Whether to treat strands serparately.</span>

<span class="sd">        rpm : bool, default False</span>
<span class="sd">            Normalize by multiplying with `1e6/(number_intervals)`.</span>

<span class="sd">        nb_cpu : int, default 1</span>
<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        pyrle.RleDict</span>

<span class="sd">            Rle with coverage or other info from the PyRanges.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d = {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [3, 8, 5],</span>
<span class="sd">        ...      &#39;End&#39;: [6, 9, 7], &#39;Score&#39;: [0.1, 5, 3.14], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr.to_rle()</span>
<span class="sd">        chr1 +</span>
<span class="sd">        --</span>
<span class="sd">        +--------+-----+-----+-----+-----+</span>
<span class="sd">        | Runs   | 3   | 3   | 2   | 1   |</span>
<span class="sd">        |--------+-----+-----+-----+-----|</span>
<span class="sd">        | Values | 0.0 | 1.0 | 0.0 | 1.0 |</span>
<span class="sd">        +--------+-----+-----+-----+-----+</span>
<span class="sd">        Rle of length 9 containing 4 elements (avg. length 2.25)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        chr1 -</span>
<span class="sd">        --</span>
<span class="sd">        +--------+-----+-----+</span>
<span class="sd">        | Runs   | 5   | 2   |</span>
<span class="sd">        |--------+-----+-----|</span>
<span class="sd">        | Values | 0.0 | 1.0 |</span>
<span class="sd">        +--------+-----+-----+</span>
<span class="sd">        Rle of length 7 containing 2 elements (avg. length 3.5)</span>
<span class="sd">        RleDict object with 2 chromosomes/strand pairs.</span>

<span class="sd">        &gt;&gt;&gt; gr.to_rle(value_col=&quot;Score&quot;)</span>
<span class="sd">        chr1 +</span>
<span class="sd">        --</span>
<span class="sd">        +--------+-----+-----+-----+-----+</span>
<span class="sd">        | Runs   | 3   | 3   | 2   | 1   |</span>
<span class="sd">        |--------+-----+-----+-----+-----|</span>
<span class="sd">        | Values | 0.0 | 0.1 | 0.0 | 5.0 |</span>
<span class="sd">        +--------+-----+-----+-----+-----+</span>
<span class="sd">        Rle of length 9 containing 4 elements (avg. length 2.25)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        chr1 -</span>
<span class="sd">        --</span>
<span class="sd">        +--------+-----+------+</span>
<span class="sd">        | Runs   | 5   | 2    |</span>
<span class="sd">        |--------+-----+------|</span>
<span class="sd">        | Values | 0.0 | 3.14 |</span>
<span class="sd">        +--------+-----+------+</span>
<span class="sd">        Rle of length 7 containing 2 elements (avg. length 3.5)</span>
<span class="sd">        RleDict object with 2 chromosomes/strand pairs.</span>

<span class="sd">        &gt;&gt;&gt; gr.to_rle(value_col=&quot;Score&quot;, strand=False)</span>
<span class="sd">        chr1</span>
<span class="sd">        +--------+-----+-----+------+------+-----+-----+</span>
<span class="sd">        | Runs   | 3   | 2   | 1    | 1    | 1   | 1   |</span>
<span class="sd">        |--------+-----+-----+------+------+-----+-----|</span>
<span class="sd">        | Values | 0.0 | 0.1 | 3.24 | 3.14 | 0.0 | 5.0 |</span>
<span class="sd">        +--------+-----+-----+------+------+-----+-----+</span>
<span class="sd">        Rle of length 9 containing 6 elements (avg. length 1.5)</span>
<span class="sd">        Unstranded RleDict object with 1 chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.to_rle(rpm=True)</span>
<span class="sd">        chr1 +</span>
<span class="sd">        --</span>
<span class="sd">        +--------+-----+-------------------+-----+-------------------+</span>
<span class="sd">        | Runs   | 3   | 3                 | 2   | 1                 |</span>
<span class="sd">        |--------+-----+-------------------+-----+-------------------|</span>
<span class="sd">        | Values | 0.0 | 333333.3333333333 | 0.0 | 333333.3333333333 |</span>
<span class="sd">        +--------+-----+-------------------+-----+-------------------+</span>
<span class="sd">        Rle of length 9 containing 4 elements (avg. length 2.25)</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        chr1 -</span>
<span class="sd">        --</span>
<span class="sd">        +--------+-----+-------------------+</span>
<span class="sd">        | Runs   | 5   | 2                 |</span>
<span class="sd">        |--------+-----+-------------------|</span>
<span class="sd">        | Values | 0.0 | 333333.3333333333 |</span>
<span class="sd">        +--------+-----+-------------------+</span>
<span class="sd">        Rle of length 7 containing 2 elements (avg. length 3.5)</span>
<span class="sd">        RleDict object with 2 chromosomes/strand pairs.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.to_rle</span> <span class="kn">import</span> <span class="n">_to_rle</span>

        <span class="k">return</span> <span class="n">_to_rle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value_col</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span> <span class="n">rpm</span><span class="o">=</span><span class="n">rpm</span><span class="p">,</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="n">nb_cpu</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.unstrand"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.unstrand">[docs]</a>    <span class="k">def</span> <span class="nf">unstrand</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove strand.</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>

<span class="sd">        Removes Strand column even if PyRanges is not stranded.</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>

<span class="sd">        PyRanges.stranded : whether PyRanges contains valid strand info.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; d =  {&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;], &#39;Start&#39;: [1, 6],</span>
<span class="sd">        ...       &#39;End&#39;: [5, 8], &#39;Strand&#39;: [&#39;+&#39;, &#39;-&#39;]}</span>
<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict(d)</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        | Chromosome   |     Start |       End | Strand       |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="sd">        |--------------+-----------+-----------+--------------|</span>
<span class="sd">        | chr1         |         1 |         5 | +            |</span>
<span class="sd">        | chr1         |         6 |         8 | -            |</span>
<span class="sd">        +--------------+-----------+-----------+--------------+</span>
<span class="sd">        Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.unstrand()</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        | Chromosome   |     Start |       End |</span>
<span class="sd">        | (category)   |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        | chr1         |         1 |         5 |</span>
<span class="sd">        | chr1         |         6 |         8 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span> <span class="ow">and</span> <span class="s2">&quot;Strand&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;Strand&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">],</span> <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]])</span>

        <span class="n">gr</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span>
                                                                       <span class="kc">True</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">dfs</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.values"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.values">[docs]</a>    <span class="k">def</span> <span class="nf">values</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the underlying DataFrames.&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="p">[</span><span class="n">df</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">df</span><span class="o">.</span><span class="n">empty</span><span class="p">]</span></div>

<div class="viewcode-block" id="PyRanges.window"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.window">[docs]</a>    <span class="k">def</span> <span class="nf">window</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return overlapping genomic windows.</span>

<span class="sd">        Windows of length `window_size` are returned.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        window_size : int</span>
<span class="sd">            Length of the windows.</span>

<span class="sd">        strand : bool, default None, i.e. auto</span>

<span class="sd">            Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use</span>
<span class="sd">            chromosome/strand pairs if the PyRanges is stranded.</span>

<span class="sd">        nb_cpu: int, default 1</span>

<span class="sd">            How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.</span>
<span class="sd">            Will only lead to speedups on large datasets.</span>

<span class="sd">        **kwargs</span>
<span class="sd">            Additional keyword arguments to pass as keyword arguments to `f`</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        PyRanges</span>

<span class="sd">            Tiled PyRanges.</span>

<span class="sd">        See also</span>
<span class="sd">        --------</span>

<span class="sd">        pyranges.PyRanges.tile : divide intervals into adjacent tiles.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.from_dict({&quot;Chromosome&quot;: [1], &quot;Start&quot;: [895], &quot;End&quot;: [1259]})</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |       895 |      1259 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr.window(200)</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        |   Chromosome |     Start |       End |</span>
<span class="sd">        |   (category) |   (int32) |   (int32) |</span>
<span class="sd">        |--------------+-----------+-----------|</span>
<span class="sd">        |            1 |       895 |      1095 |</span>
<span class="sd">        |            1 |      1095 |      1259 |</span>
<span class="sd">        +--------------+-----------+-----------+</span>
<span class="sd">        Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome.</span>

<span class="sd">        &gt;&gt;&gt; gr = pr.data.ensembl_gtf()[[&quot;Feature&quot;, &quot;gene_name&quot;]]</span>
<span class="sd">        &gt;&gt;&gt; gr</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="sd">        | 1            | gene         | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | transcript   | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | exon         | 11868     | 12227     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | exon         | 12612     | 12721     | +            | DDX11L1     |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="sd">        | 1            | gene         | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | transcript   | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1179364   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1173055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>

<span class="sd">        &gt;&gt;&gt; gr.window(1000)</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        | Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="sd">        | (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="sd">        |--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="sd">        | 1            | gene         | 11868     | 12868     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | gene         | 12868     | 13868     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | gene         | 13868     | 14409     | +            | DDX11L1     |</span>
<span class="sd">        | 1            | transcript   | 11868     | 12868     | +            | DDX11L1     |</span>
<span class="sd">        | ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="sd">        | 1            | exon         | 1173055   | 1174055   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1174055   | 1175055   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1175055   | 1176055   | -            | TTLL10-AS1  |</span>
<span class="sd">        | 1            | exon         | 1176055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="sd">        +--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="sd">        Stranded PyRanges object has 7,516 rows and 6 columns from 1 chromosomes.</span>
<span class="sd">        For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">pyranges.methods.windows</span> <span class="kn">import</span> <span class="n">_windows</span>

        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stranded</span>

        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sparse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;self&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
        <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;window_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">window_size</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pyrange_apply_single</span><span class="p">(</span><span class="n">_windows</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></div>


<div class="viewcode-block" id="PyRanges.__getstate__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__getstate__">[docs]</a>    <span class="k">def</span> <span class="nf">__getstate__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">dfs</span></div>

<div class="viewcode-block" id="PyRanges.__setstate__"><a class="viewcode-back" href="../../autoapi/pyranges/pyranges/index.html#pyranges.PyRanges.__setstate__">[docs]</a>    <span class="k">def</span> <span class="nf">__setstate__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">d</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">[</span><span class="s2">&quot;dfs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span></div></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Endre Bakken Stovner and Marco Mariotti.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>