<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyranges.readers &mdash; pyranges  0.0.124 documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="pyranges.statistics" href="../statistics/index.html" />
    <link rel="prev" title="pyranges.pyranges" href="../pyranges/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            pyranges
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial of PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../how_to_book.html">How-to-book</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../data/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../genomicfeatures/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.genomicfeatures</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../get_fasta/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.get_fasta</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../helpers/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.helpers</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../multioverlap/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.multioverlap</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../multithreaded/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.multithreaded</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../pyranges/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.pyranges</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.readers</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../statistics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.statistics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../stats/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.stats</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../version/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.version</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">pyranges</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.readers</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/autoapi/pyranges/readers/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-pyranges.readers">
<span id="pyranges-readers"></span><h1><a class="reference internal" href="#module-pyranges.readers" title="pyranges.readers"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.readers</span></code></a><a class="headerlink" href="#module-pyranges.readers" title="Permalink to this heading"></a></h1>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.readers.read_bed" title="pyranges.readers.read_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_bed</span></code></a>(f[, as_df, nrows])</p></td>
<td><p>Return bed file as PyRanges.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.readers.read_bam" title="pyranges.readers.read_bam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_bam</span></code></a>(f[, sparse, as_df, mapq, required_flag, ...])</p></td>
<td><p>Return bam file as PyRanges.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.readers._fetch_gene_transcript_exon_id" title="pyranges.readers._fetch_gene_transcript_exon_id"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_fetch_gene_transcript_exon_id</span></code></a>(attribute[, annotation])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.readers.skiprows" title="pyranges.readers.skiprows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">skiprows</span></code></a>(f)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.readers.read_gtf" title="pyranges.readers.read_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_gtf</span></code></a>(f[, full, as_df, nrows, duplicate_attr])</p></td>
<td><p>Read files in the Gene Transfer Format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.readers.read_gtf_full" title="pyranges.readers.read_gtf_full"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_gtf_full</span></code></a>(f[, as_df, nrows, skiprows, duplicate_attr])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.readers.to_rows" title="pyranges.readers.to_rows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rows</span></code></a>(anno)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.readers.to_rows_keep_duplicates" title="pyranges.readers.to_rows_keep_duplicates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rows_keep_duplicates</span></code></a>(anno)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.readers.read_gtf_restricted" title="pyranges.readers.read_gtf_restricted"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_gtf_restricted</span></code></a>(f, skiprows[, as_df, nrows])</p></td>
<td><p>seqname - name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix. Important note: the seqname must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.readers.to_rows_gff3" title="pyranges.readers.to_rows_gff3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rows_gff3</span></code></a>(anno)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.readers.read_gff3" title="pyranges.readers.read_gff3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_gff3</span></code></a>(f[, full, annotation, as_df, nrows])</p></td>
<td><p>Read files in the General Feature Format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyranges.readers.read_bigwig" title="pyranges.readers.read_bigwig"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_bigwig</span></code></a>(f[, as_df])</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_bed">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_bed" title="Permalink to this definition"></a></dt>
<dd><p>Return bed file as PyRanges.</p>
<p>This is a reader for files that follow the bed format. They can have from
3-12 columns which will be named like so:</p>
<p>Chromosome Start End Name Score Strand ThickStart ThickEnd ItemRGB
BlockCount BlockSizes BlockStarts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to bed file</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>nrows</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of rows to return.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If you just want to create a PyRanges from a tab-delimited bed-like file,
use <cite>pr.PyRanges(pandas.read_table(f))</cite> instead.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_example_path</span><span class="p">(</span><span class="s2">&quot;aorta.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |      9939 |     10138 | H3K27me3   |         7 | +            |</span>
<span class="go">| chr1         |      9953 |     10152 | H3K27me3   |         5 | +            |</span>
<span class="go">| chr1         |      9916 |     10115 | H3K27me3   |         5 | -            |</span>
<span class="go">| chr1         |      9951 |     10150 | H3K27me3   |         8 | -            |</span>
<span class="go">| chr1         |      9978 |     10177 | H3K27me3   |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bed</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">as_df</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">  Chromosome  Start    End      Name  Score Strand</span>
<span class="go">0       chr1   9916  10115  H3K27me3      5      -</span>
<span class="go">1       chr1   9939  10138  H3K27me3      7      +</span>
<span class="go">2       chr1   9951  10150  H3K27me3      8      -</span>
<span class="go">3       chr1   9953  10152  H3K27me3      5      +</span>
<span class="go">4       chr1   9978  10177  H3K27me3      7      -</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_bam">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_bam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">required_flag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_flag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1540</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_bam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_bam" title="Permalink to this definition"></a></dt>
<dd><p>Return bam file as PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to bam file</p></li>
<li><p><strong>sparse</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to return only.</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>mapq</strong> (<em>int</em><em>, </em><em>default 0</em>) – Minimum mapping quality score.</p></li>
<li><p><strong>required_flag</strong> (<em>int</em><em>, </em><em>default 0</em>) – Flags which must be present for the interval to be read.</p></li>
<li><p><strong>filter_flag</strong> (<em>int</em><em>, </em><em>default 1540</em>) – Ignore reads with these flags. Default 1540, which means that either
the read is unmapped, the read failed vendor or platfrom quality
checks, or the read is a PCR or optical duplicate.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This functionality requires the library <cite>bamread</cite>. It can be installed with
<cite>pip install bamread</cite> or <cite>conda install -c bioconda bamread</cite>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_example_path</span><span class="p">(</span><span class="s2">&quot;control.bam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | Flag       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (uint16)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         | 887771    | 887796    | +            | 16         |</span>
<span class="go">| chr1         | 994660    | 994685    | +            | 16         |</span>
<span class="go">| chr1         | 1770383   | 1770408   | +            | 16         |</span>
<span class="go">| chr1         | 1995141   | 1995166   | +            | 16         |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...        |</span>
<span class="go">| chrY         | 57402214  | 57402239  | +            | 16         |</span>
<span class="go">| chrY         | 10643526  | 10643551  | -            | 0          |</span>
<span class="go">| chrY         | 11776321  | 11776346  | -            | 0          |</span>
<span class="go">| chrY         | 20557165  | 20557190  | -            | 0          |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 5 columns from 25 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">sparse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">as_df</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">     Chromosome    Start      End Strand  Flag  QueryStart  QueryEnd Name Cigar Quality</span>
<span class="go">0          chr1   887771   887796      +    16           0        25   U0   25M    None</span>
<span class="go">1          chr1   994660   994685      +    16           0        25   U0   25M    None</span>
<span class="go">2          chr1  1041102  1041127      -     0           0        25   U0   25M    None</span>
<span class="go">3          chr1  1770383  1770408      +    16           0        25   U0   25M    None</span>
<span class="go">4          chr1  1995141  1995166      +    16           0        25   U0   25M    None</span>
<span class="go">...         ...      ...      ...    ...   ...         ...       ...  ...   ...     ...</span>
<span class="go">9995       chrM     3654     3679      -     0           0        25   U0   25M    None</span>
<span class="go">9996       chrM     3900     3925      +    16           0        25   U0   25M    None</span>
<span class="go">9997       chrM    13006    13031      +    16           0        25   U0   25M    None</span>
<span class="go">9998       chrM    14257    14282      -     0           0        25   U0   25M    None</span>
<span class="go">9999       chrM    14257    14282      -     0           0        25   U0   25M    None</span>

<span class="go">[10000 rows x 10 columns]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers._fetch_gene_transcript_exon_id">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">_fetch_gene_transcript_exon_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attribute</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#_fetch_gene_transcript_exon_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers._fetch_gene_transcript_exon_id" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.skiprows">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">skiprows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#skiprows"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.skiprows" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_gtf">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_gtf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">duplicate_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_gtf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_gtf" title="Permalink to this definition"></a></dt>
<dd><p>Read files in the Gene Transfer Format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to GTF file.</p></li>
<li><p><strong>full</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to read and interpret the annotation column.</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>nrows</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of rows to read. Default None, i.e. all.</p></li>
<li><p><strong>duplicate_attr</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to handle (potential) duplicate attributes or just keep last one.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The GTF format encodes both Start and End as 1-based included.
PyRanges (and also the DF returned by this function, if as_df=True), instead
encodes intervals as 0-based, Start included and End excluded.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_gff3" title="pyranges.read_gff3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_gff3</span></code></a></dt><dd><p>read files in the General Feature Format</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">get_example_path</span><span class="p">(</span><span class="s2">&quot;ensembl.gtf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">read_gtf</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># +--------------+------------+--------------+-----------+-----------+------------+--------------+------------+-----------------+----------------+-------+</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | Chromosome   | Source     | Feature      | Start     | End       | Score      | Strand       | Frame      | gene_id         | gene_version   | +18   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | (category)   | (object)   | (category)   | (int32)   | (int32)   | (object)   | (category)   | (object)   | (object)        | (object)       | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># |--------------+------------+--------------+-----------+-----------+------------+--------------+------------+-----------------+----------------+-------|</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | gene         | 11868     | 14409     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | transcript   | 11868     | 14409     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | exon         | 11868     | 12227     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | havana     | exon         | 12612     | 12721     | .          | +            | .          | ENSG00000223972 | 5              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | ...          | ...        | ...          | ...       | ...       | ...        | ...          | ...        | ...             | ...            | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | transcript   | 120724    | 133723    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | exon         | 133373    | 133723    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | exon         | 129054    | 129223    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># | 1            | ensembl    | exon         | 120873    | 120932    | .          | -            | .          | ENSG00000238009 | 6              | ...   |</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># +--------------+------------+--------------+-----------+-----------+------------+--------------+------------+-----------------+----------------+-------+</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Stranded PyRanges object has 95 rows and 28 columns from 1 chromosomes.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># 18 hidden columns: gene_name, gene_source, gene_biotype, transcript_id, transcript_version, transcript_name, transcript_source, transcript_biotype, tag, transcript_support_level, ... (+ 8 more.)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_gtf_full">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_gtf_full</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skiprows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">duplicate_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_gtf_full"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_gtf_full" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.to_rows">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">to_rows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anno</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#to_rows"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.to_rows" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.to_rows_keep_duplicates">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">to_rows_keep_duplicates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anno</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#to_rows_keep_duplicates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.to_rows_keep_duplicates" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_gtf_restricted">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_gtf_restricted</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skiprows</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_gtf_restricted"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_gtf_restricted" title="Permalink to this definition"></a></dt>
<dd><p>seqname - name of the chromosome or scaffold; chromosome names can be given with or without the ‘chr’ prefix. Important note: the seqname must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.
# source - name of the program that generated this feature, or the data source (database or project name)
feature - feature type name, e.g. Gene, Variation, Similarity
start - Start position of the feature, with sequence numbering starting at 1.
end - End position of the feature, with sequence numbering starting at 1.
score - A floating point value.
strand - defined as + (forward) or - (reverse).
# frame - One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on..
attribute - A semicolon-separated list of tag-value pairs, providing additional information about each feature.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.to_rows_gff3">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">to_rows_gff3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">anno</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#to_rows_gff3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.to_rows_gff3" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_gff3">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_gff3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_gff3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_gff3" title="Permalink to this definition"></a></dt>
<dd><p>Read files in the General Feature Format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>str</em>) – Path to GFF file.</p></li>
<li><p><strong>full</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to read and interpret the annotation column.</p></li>
<li><p><strong>as_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as pandas DataFrame instead of PyRanges.</p></li>
<li><p><strong>nrows</strong> (<em>int</em><em>, </em><em>default None</em>) – Number of rows to read. Default None, i.e. all.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The gff3 format encodes both Start and End as 1-based included.
PyRanges (and also the DF returned by this function, if as_df=True), instead
encodes intervals as 0-based, Start included and End excluded.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_gtf" title="pyranges.read_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_gtf</span></code></a></dt><dd><p>read files in the Gene Transfer Format</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyranges.readers.read_bigwig">
<span class="sig-prename descclassname"><span class="pre">pyranges.readers.</span></span><span class="sig-name descname"><span class="pre">read_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/readers.html#read_bigwig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.readers.read_bigwig" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../pyranges/index.html" class="btn btn-neutral float-left" title="pyranges.pyranges" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../statistics/index.html" class="btn btn-neutral float-right" title="pyranges.statistics" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Endre Bakken Stovner and Marco Mariotti.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>