<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pyranges.pyranges &mdash; pyranges  0.0.124 documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="pyranges.readers" href="../readers/index.html" />
    <link rel="prev" title="pyranges.multithreaded" href="../multithreaded/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            pyranges
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial of PyRanges</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../how_to_book.html">How-to-book</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../data/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.data</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../genomicfeatures/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.genomicfeatures</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../get_fasta/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.get_fasta</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../helpers/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.helpers</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../multioverlap/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.multioverlap</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../multithreaded/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.multithreaded</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.pyranges</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../readers/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.readers</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../statistics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.statistics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../stats/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.stats</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../version/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.version</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">pyranges</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.pyranges</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/autoapi/pyranges/pyranges/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-pyranges.pyranges">
<span id="pyranges-pyranges"></span><h1><a class="reference internal" href="#module-pyranges.pyranges" title="pyranges.pyranges"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pyranges.pyranges</span></code></a><a class="headerlink" href="#module-pyranges.pyranges" title="Permalink to this heading"></a></h1>
<p>Data structure for genomic intervals and their annotation.</p>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="classes">
<h3>Classes<a class="headerlink" href="#classes" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges</span></code></a></p></td>
<td><p>Two-dimensional representation of genomic intervals and their annotations.</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyranges.pyranges.</span></span><span class="sig-name descname"><span class="pre">PyRanges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">starts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strands</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int64</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges" title="Permalink to this definition"></a></dt>
<dd><p>Two-dimensional representation of genomic intervals and their annotations.</p>
<p>A PyRanges object must have the columns Chromosome, Start and End. These
describe the genomic position and function as implicit row labels. A Strand
column is optional and adds strand information to the intervals. Any other
columns are allowed and are considered metadata.</p>
<p>Operations between PyRanges align intervals based on their position.</p>
<p>If a PyRanges is built using the arguments chromosomes, starts, ends and
optionally strands, all non-scalars must be of the same length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em><em> or </em><em>dict</em><em> of </em><em>pandas.DataFrame</em><em>, </em><em>default None</em>) – The data to be stored in the PyRanges.</p></li>
<li><p><strong>chromosomes</strong> (<em>array-like</em><em> or </em><em>scalar value</em><em>, </em><em>default None</em>) – The chromosome(s) in the PyRanges.</p></li>
<li><p><strong>starts</strong> (<em>array-like</em><em>, </em><em>default None</em>) – The start postions in the PyRanges.</p></li>
<li><p><strong>ends</strong> (<em>array-like</em><em>, </em><em>default None</em>) – The end postions in the PyRanges.</p></li>
<li><p><strong>strands</strong> (<em>array-like</em><em> or </em><em>scalar value</em><em>, </em><em>default None</em>) – The strands in the PyRanges.</p></li>
<li><p><strong>int64</strong> (<em>bool</em><em>, </em><em>default False</em>) – Use np.int64 to represent starts and ends</p></li>
<li><p><strong>copy_df</strong> (<em>bool</em><em>, </em><em>default True</em>) – Copy input pandas.DataFrame</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_bed" title="pyranges.read_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_bed</span></code></a></dt><dd><p>read bed-file into PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.read_bam" title="pyranges.read_bam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_bam</span></code></a></dt><dd><p>read bam-file into PyRanges</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_gff</span></code></dt><dd><p>read gff-file into PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.read_gtf" title="pyranges.read_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_gtf</span></code></a></dt><dd><p>read gtf-file into PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.from_dict" title="pyranges.from_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.from_dict</span></code></a></dt><dd><p>create PyRanges from dict of columns</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.from_string" title="pyranges.from_string"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.from_string</span></code></a></dt><dd><p>create PyRanges from multiline string</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>A PyRanges object is represented internally as a dictionary efficiency. The keys are
chromosomes or chromosome/strand tuples and the values are pandas DataFrames.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">()</span>
<span class="go">Empty PyRanges</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">chromosomes</span><span class="o">=</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="n">starts</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">ends</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">149</span><span class="p">],</span>
<span class="gp">... </span>            <span class="n">strands</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">),</span> <span class="n">int64</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int64) |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         3 | +            |</span>
<span class="go">| chr1         |         5 |       149 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">150</span><span class="p">,</span> <span class="mi">201</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span>
<span class="go">  Chromosome  Start  End</span>
<span class="go">0       chr1    100  150</span>
<span class="go">1       chr2    200  201</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |       100 |       150 |</span>
<span class="go">| chr2         |       200 |       201 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">27</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;TP&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;FP&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">12</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="s2">&quot;TN&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;FN&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End |        TP |        FP |        TN |        FN |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) |   (int64) |   (int64) |   (int64) |   (int64) |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="go">|            1 | +            |         1 |         2 |         0 |        12 |        10 |         2 |</span>
<span class="go">|            1 | -            |         4 |        27 |         1 |        11 |         9 |         3 |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.chromosomes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosomes</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>Return chromosomes in natsorted order.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.columns">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">columns</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.columns" title="Permalink to this definition"></a></dt>
<dd><p>Return the column labels of the PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.Index</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.chromosomes" title="pyranges.pyranges.PyRanges.chromosomes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.chromosomes</span></code></a></dt><dd><p>return the chromosomes in the PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f2</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         2 | a          |         0 | +            |</span>
<span class="go">| chr1         |         6 |         7 | b          |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">columns</span>
<span class="go">Index([&#39;Chromosome&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Name&#39;, &#39;Score&#39;, &#39;Strand&#39;], dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Sco|re&quot;</span><span class="p">,</span> <span class="s2">&quot;NYAN&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+------------+------------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |   NYANNYAN | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |    (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+------------+--------------|</span>
<span class="go">| chr1         |         1 |         2 | a          |          0 | +            |</span>
<span class="go">| chr1         |         6 |         7 | b          |          0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+------------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.df">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">df</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.df" title="Permalink to this definition"></a></dt>
<dd><p>Return PyRanges as DataFrame.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.as_df" title="pyranges.pyranges.PyRanges.as_df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.as_df</span></code></a></dt><dd><p>return PyRanges as DataFrame.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.dtypes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dtypes</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.dtypes" title="Permalink to this definition"></a></dt>
<dd><p>Return the dtypes of the PyRanges.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="go">| chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">dtypes</span>
<span class="go">Chromosome    category</span>
<span class="go">Start            int32</span>
<span class="go">End              int32</span>
<span class="go">Name            object</span>
<span class="go">Score            int64</span>
<span class="go">Strand        category</span>
<span class="go">dtype: object</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.empty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">empty</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.empty" title="Permalink to this definition"></a></dt>
<dd><p>Indicate whether PyRanges is empty.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.length" title="Permalink to this definition"></a></dt>
<dd><p>Return the total length of the intervals.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.lengths" title="pyranges.pyranges.PyRanges.lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.lengths</span></code></a></dt><dd><p>return the intervals lengths</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">length</span>
<span class="go">6</span>
</pre></div>
</div>
<p>To find the length of the genome covered by the intervals, use merge first:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">length</span>
<span class="go">5</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.stranded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stranded</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.stranded" title="Permalink to this definition"></a></dt>
<dd><p>Whether PyRanges has (valid) strand info.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A PyRanges can have invalid values in the Strand-column. It is not considered stranded.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.strands" title="pyranges.pyranges.PyRanges.strands"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.strands</span></code></a></dt><dd><p>return the strands</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span>  <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;.&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         5 | +            |</span>
<span class="go">| chr1         |         6 |         8 | .            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
<span class="go">Considered unstranded due to these Strand values: &#39;.&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">stranded</span>
<span class="go">False</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;Strand&quot;</span> <span class="ow">in</span> <span class="n">gr</span><span class="o">.</span><span class="n">columns</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.strands">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strands</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.strands" title="Permalink to this definition"></a></dt>
<dd><p>Return strands.</p>
<p class="rubric">Notes</p>
<p>If the strand-column contains an invalid value, [] is returned.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.stranded" title="pyranges.pyranges.PyRanges.stranded"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.stranded</span></code></a></dt><dd><p>whether has valid strand info</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span>  <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;.&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         5 | +            |</span>
<span class="go">| chr1         |         6 |         8 | .            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
<span class="go">Considered unstranded due to these Strand values: &#39;.&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">strands</span>
<span class="go">[]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">Strand</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
<span class="go">[&#39;+&#39;, &#39;.&#39;]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">Strand</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">strands</span>
<span class="go">[&#39;+&#39;, &#39;-&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.dfs">
<span class="sig-name descname"><span class="pre">dfs</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.dfs" title="Permalink to this definition"></a></dt>
<dd><p>Dict mapping chromosomes or chromosome/strand pairs to pandas DataFrames.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.features">
<span class="sig-name descname"><span class="pre">features</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.features" title="Permalink to this definition"></a></dt>
<dd><p>Namespace for genomic-features methods.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../genomicfeatures/index.html#module-pyranges.genomicfeatures" title="pyranges.genomicfeatures"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.genomicfeatures</span></code></a></dt><dd><p>namespace for feature-functionality</p>
</dd>
<dt><a class="reference internal" href="../genomicfeatures/index.html#pyranges.genomicfeatures.GenomicFeaturesMethods" title="pyranges.genomicfeatures.GenomicFeaturesMethods"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.genomicfeatures.GenomicFeaturesMethods</span></code></a></dt><dd><p>namespace for feature-functionality</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.stats">
<span class="sig-name descname"><span class="pre">stats</span></span><a class="headerlink" href="#pyranges.pyranges.PyRanges.stats" title="Permalink to this definition"></a></dt>
<dd><p>Namespace for statistcal methods.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../statistics/index.html#module-pyranges.statistics" title="pyranges.statistics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.statistics</span></code></a></dt><dd><p>namespace for statistics</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.stats.StatisticsMethods</span></code></dt><dd><p>namespace for statistics</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__array_ufunc__">
<span class="sig-name descname"><span class="pre">__array_ufunc__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__array_ufunc__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__array_ufunc__" title="Permalink to this definition"></a></dt>
<dd><p>Apply unary numpy-function.</p>
<p>Apply function to all columns which are not index, i.e. Chromosome,
Start, End nor Strand.</p>
<p class="rubric">Notes</p>
<p>Function must produce a vector of equal length.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span><span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;Score&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">25</span><span class="p">],</span> <span class="s2">&quot;Score2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">121</span><span class="p">,</span> <span class="mi">144</span><span class="p">,</span> <span class="mi">169</span><span class="p">],</span>
<span class="gp">... </span><span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;n1&quot;</span><span class="p">,</span> <span class="s2">&quot;n2&quot;</span><span class="p">,</span> <span class="s2">&quot;n3&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |     Score |    Score2 | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (int64) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+------------|</span>
<span class="go">|            1 |         1 |         2 |         9 |       121 | n1         |</span>
<span class="go">|            2 |         2 |         3 |        16 |       144 | n2         |</span>
<span class="go">|            3 |         3 |         4 |        25 |       169 | n3         |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">gr</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-------------+-------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |       Score |      Score2 | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (float64) |   (float64) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+-------------+-------------+------------|</span>
<span class="go">|            1 |         1 |         2 |           3 |          11 | n1         |</span>
<span class="go">|            2 |         2 |         3 |           4 |          12 | n2         |</span>
<span class="go">|            3 |         3 |         4 |           5 |          13 | n3         |</span>
<span class="go">+--------------+-----------+-----------+-------------+-------------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__getattr__">
<span class="sig-name descname"><span class="pre">__getattr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__getattr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__getattr__" title="Permalink to this definition"></a></dt>
<dd><p>Return column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – Column to return</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.Series</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">250</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">125</span><span class="p">,</span> <span class="mi">251</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">Start</span>
<span class="go">0      0</span>
<span class="go">1    100</span>
<span class="go">2    250</span>
<span class="go">Name: Start, dtype: int32</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__setattr__">
<span class="sig-name descname"><span class="pre">__setattr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__setattr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__setattr__" title="Permalink to this definition"></a></dt>
<dd><p>Insert or update column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column_name</strong> (<em>str</em>) – Name of column to update or insert.</p></li>
<li><p><strong>column</strong> (<em>list</em><em>, </em><em>np.array</em><em> or </em><em>pd.Series</em>) – Data to insert.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">250</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">125</span><span class="p">,</span> <span class="mi">251</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">Start</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int64) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         1 |        10 |</span>
<span class="go">|            1 |         1 |       125 |</span>
<span class="go">|            1 |         2 |       251 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">val</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__getitem__" title="Permalink to this definition"></a></dt>
<dd><p>Fetch columns or subset on position.</p>
<p>If a list is provided, the column(s) in the list is returned. This subsets on columns.</p>
<p>If a numpy array is provided, it must be of type bool and the same length as the PyRanges.</p>
<p>Otherwise, a subset of the rows is returned with the location info provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>val</strong> (<em>bool array/Series</em><em>, </em><em>tuple</em><em>, </em><em>list</em><em>, </em><em>str</em><em> or </em><em>slice</em>) – Data to fetch.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">columns</span>
<span class="go">Index([&#39;Chromosome&#39;, &#39;Source&#39;, &#39;Feature&#39;, &#39;Start&#39;, &#39;End&#39;, &#39;Score&#39;, &#39;Strand&#39;,</span>
<span class="go">    &#39;Frame&#39;, &#39;gene_biotype&#39;, &#39;gene_id&#39;, &#39;gene_name&#39;, &#39;gene_source&#39;,</span>
<span class="go">    &#39;gene_version&#39;, &#39;tag&#39;, &#39;transcript_biotype&#39;, &#39;transcript_id&#39;,</span>
<span class="go">    &#39;transcript_name&#39;, &#39;transcript_source&#39;, &#39;transcript_support_level&#39;,</span>
<span class="go">    &#39;transcript_version&#39;, &#39;exon_id&#39;, &#39;exon_number&#39;, &#39;exon_version&#39;,</span>
<span class="go">    &#39;(assigned&#39;, &#39;previous&#39;, &#39;protein_id&#39;, &#39;protein_version&#39;, &#39;ccds_id&#39;],</span>
<span class="go">    dtype=&#39;object&#39;)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">gr</span><span class="p">[[</span><span class="s2">&quot;Source&quot;</span><span class="p">,</span> <span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="go">| Chromosome   | Source     | Feature      | Start     | End       | Strand       | gene_id         |</span>
<span class="go">| (category)   | (object)   | (category)   | (int32)   | (int32)   | (category)   | (object)        |</span>
<span class="go">|--------------+------------+--------------+-----------+-----------+--------------+-----------------|</span>
<span class="go">| 1            | havana     | gene         | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | havana     | transcript   | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | havana     | exon         | 11868     | 12227     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | havana     | exon         | 12612     | 12721     | +            | ENSG00000223972 |</span>
<span class="go">| ...          | ...        | ...          | ...       | ...       | ...          | ...             |</span>
<span class="go">| 1            | havana     | gene         | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">| 1            | havana     | transcript   | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">| 1            | havana     | exon         | 1179364   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">| 1            | havana     | exon         | 1173055   | 1176396   | -            | ENSG00000205231 |</span>
<span class="go">+--------------+------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<p>Create boolean Series and use it to subset:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">Feature</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">gene_id</span> <span class="o">==</span> <span class="s2">&quot;ENSG00000223972&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="p">[</span><span class="n">s</span><span class="p">]</span>
<span class="go">+--------------+----------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="go">| Chromosome   | Source         | Feature      | Start     | End       | Strand       | gene_id         |</span>
<span class="go">| (category)   | (object)       | (category)   | (int32)   | (int32)   | (category)   | (object)        |</span>
<span class="go">|--------------+----------------+--------------+-----------+-----------+--------------+-----------------|</span>
<span class="go">| 1            | havana         | gene         | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | havana         | transcript   | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | havana         | exon         | 11868     | 12227     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | havana         | exon         | 12612     | 12721     | +            | ENSG00000223972 |</span>
<span class="go">| ...          | ...            | ...          | ...       | ...       | ...          | ...             |</span>
<span class="go">| 1            | havana         | gene         | 1062207   | 1063288   | -            | ENSG00000273443 |</span>
<span class="go">| 1            | ensembl_havana | gene         | 1070966   | 1074306   | -            | ENSG00000237330 |</span>
<span class="go">| 1            | ensembl_havana | gene         | 1081817   | 1116361   | -            | ENSG00000131591 |</span>
<span class="go">| 1            | havana         | gene         | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">+--------------+----------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="go">Stranded PyRanges object has 95 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cs</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cs</span><span class="p">[</span><span class="mi">10000</span><span class="p">:</span><span class="mi">100000</span><span class="p">]</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr2         |     33241 |     33266 | U0         |         0 | +            |</span>
<span class="go">| chr2         |     13611 |     13636 | U0         |         0 | -            |</span>
<span class="go">| chr2         |     32620 |     32645 | U0         |         0 | -            |</span>
<span class="go">| chr3         |     87179 |     87204 | U0         |         0 | +            |</span>
<span class="go">| chr4         |     45413 |     45438 | U0         |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 6 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cs</span><span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 100079649 | 100079674 | U0         | 0         | -            |</span>
<span class="go">| chr1         | 223587418 | 223587443 | U0         | 0         | -            |</span>
<span class="go">| chr1         | 202450161 | 202450186 | U0         | 0         | -            |</span>
<span class="go">| chr1         | 156338310 | 156338335 | U0         | 0         | -            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chr1         | 203557775 | 203557800 | U0         | 0         | -            |</span>
<span class="go">| chr1         | 28114107  | 28114132  | U0         | 0         | -            |</span>
<span class="go">| chr1         | 21622765  | 21622790  | U0         | 0         | -            |</span>
<span class="go">| chr1         | 80668132  | 80668157  | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 437 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cs</span><span class="p">[</span><span class="s2">&quot;chr5&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="mi">90000</span><span class="p">:]</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr5         | 399682    | 399707    | U0         | 0         | -            |</span>
<span class="go">| chr5         | 1847502   | 1847527   | U0         | 0         | -            |</span>
<span class="go">| chr5         | 5247533   | 5247558   | U0         | 0         | -            |</span>
<span class="go">| chr5         | 5300394   | 5300419   | U0         | 0         | -            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chr5         | 178786234 | 178786259 | U0         | 0         | -            |</span>
<span class="go">| chr5         | 179268931 | 179268956 | U0         | 0         | -            |</span>
<span class="go">| chr5         | 179289594 | 179289619 | U0         | 0         | -            |</span>
<span class="go">| chr5         | 180513795 | 180513820 | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 285 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cs</span><span class="p">[</span><span class="s2">&quot;chrM&quot;</span><span class="p">]</span>
<span class="go">Empty PyRanges</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__iter__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__iter__" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over the keys and values.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.iter</span></code></dt><dd><p>iterate over multiple PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">250</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">125</span><span class="p">,</span> <span class="mi">251</span><span class="p">],</span> <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]})</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gr</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
<span class="go">(&#39;1&#39;, &#39;+&#39;)</span>
<span class="go">Chromosome  Start  End Strand</span>
<span class="go">0          1      0   10      +</span>
<span class="go">1          1    100  125      +</span>
<span class="go">(&#39;1&#39;, &#39;-&#39;)</span>
<span class="go">Chromosome  Start  End Strand</span>
<span class="go">2          1    250  251      -</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__len__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__len__" title="Permalink to this definition"></a></dt>
<dd><p>Return the number of intervals in the PyRanges.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__str__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return string representation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__repr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return REPL representation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges._repr_html_">
<span class="sig-name descname"><span class="pre">_repr_html_</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges._repr_html_"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges._repr_html_" title="Permalink to this definition"></a></dt>
<dd><p>Return REPL HTML representation for Jupyter Noteboooks.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.apply">
<span class="sig-name descname"><span class="pre">apply</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_pyranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.apply"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.apply" title="Permalink to this definition"></a></dt>
<dd><p>Apply a function to the PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>function</em>) – Function to apply on each DataFrame in a PyRanges</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use
chromosome/strand pairs if the PyRanges is stranded.</p></li>
<li><p><strong>as_pyranges</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to return as a PyRanges or dict. If <cite>f</cite> does not return a DataFrame valid for
PyRanges, <cite>as_pyranges</cite> must be False.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Result of applying f to each DataFrame in the PyRanges</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a> or dict</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.apply_pair" title="pyranges.PyRanges.apply_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.apply_pair</span></code></a></dt><dd><p>apply a function to a pair of PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.apply_chunks" title="pyranges.PyRanges.apply_chunks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.apply_chunks</span></code></a></dt><dd><p>apply a row-based function to a PyRanges in parallel</p>
</dd>
</dl>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is the function used internally to carry out almost all unary PyRanges methods.</p>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">27</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+--------------+-----------+-----------|</span>
<span class="go">|            1 | +            |         1 |         2 |</span>
<span class="go">|            1 | +            |         4 |        27 |</span>
<span class="go">|            2 | +            |         9 |        10 |</span>
<span class="go">|            2 | -            |         2 |        13 |</span>
<span class="go">+--------------+--------------+-----------+-----------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 4 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">),</span> <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">{(&#39;1&#39;, &#39;+&#39;): 2, (&#39;2&#39;, &#39;+&#39;): 1, (&#39;2&#39;, &#39;-&#39;): 1}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">),</span> <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">{&#39;1&#39;: 2, &#39;2&#39;: 2}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">add_to_ends</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;End&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;slack&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">df</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">add_to_ends</span><span class="p">,</span> <span class="n">slack</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+--------------+-----------+-----------|</span>
<span class="go">|            1 | +            |         1 |       502 |</span>
<span class="go">|            1 | +            |         4 |       527 |</span>
<span class="go">|            2 | +            |         9 |       510 |</span>
<span class="go">|            2 | -            |         2 |       513 |</span>
<span class="go">+--------------+--------------+-----------+-----------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 4 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.apply_chunks">
<span class="sig-name descname"><span class="pre">apply_chunks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_pyranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.apply_chunks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.apply_chunks" title="Permalink to this definition"></a></dt>
<dd><p>Apply a row-based function to arbitrary partitions of the PyRanges.</p>
<p>apply_chunks speeds up the application of functions where the result is not affected by
applying the function to ordered, non-overlapping splits of the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>function</em>) – Row-based or associative function to apply on the partitions.</p></li>
<li><p><strong>as_pyranges</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as a PyRanges or dict.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. The data is split into nb_cpu partitions.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Result of applying f to each partition of the DataFrames in the PyRanges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict of lists</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.apply_pair" title="pyranges.PyRanges.apply_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.apply_pair</span></code></a></dt><dd><p>apply a function to a pair of PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.apply_chunks" title="pyranges.PyRanges.apply_chunks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.apply_chunks</span></code></a></dt><dd><p>apply a row-based function to a PyRanges in parallel</p>
</dd>
</dl>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>apply_chunks will only lead to speedups on large datasets or slow-running functions. Using
it with nb_cpu=1 is pointless; use apply instead.</p>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         2 |         9 |</span>
<span class="go">|            1 |         3 |         4 |</span>
<span class="go">|            1 |         5 |         6 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">apply_chunks</span><span class="p">(</span>
<span class="gp">... </span><span class="k">lambda</span> <span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">End</span> <span class="o">+</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;add&quot;</span><span class="p">]),</span> <span class="n">nb_cpu</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">add</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="go">{&#39;1&#39;: [[1009, 1004, 1006]]}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.apply_pair">
<span class="sig-name descname"><span class="pre">apply_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_pyranges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.apply_pair"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.apply_pair" title="Permalink to this definition"></a></dt>
<dd><p>Apply a function to a pair of PyRanges.</p>
<p>The function is applied to each chromosome or chromosome/strand pair found in at least one
of the PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>function</em>) – Row-based or associative function to apply on the DataFrames.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>as_pyranges</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return as a PyRanges or dict. If <cite>f</cite> does not return a DataFrame valid for
PyRanges, <cite>as_pyranges</cite> must be False.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Result of applying f to each partition of the DataFrames in the PyRanges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict of lists</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.apply_pair" title="pyranges.PyRanges.apply_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.apply_pair</span></code></a></dt><dd><p>apply a function to a pair of PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.apply_chunks" title="pyranges.PyRanges.apply_chunks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.apply_chunks</span></code></a></dt><dd><p>apply a row-based function to a PyRanges in parallel</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.iter</span></code></dt><dd><p>iterate over two or more PyRanges</p>
</dd>
</dl>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is the function used internally to carry out almost all comparison functions in
PyRanges.</p>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq_background</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">apply_pair</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">pr</span><span class="o">.</span><span class="n">methods</span><span class="o">.</span><span class="n">intersection</span><span class="o">.</span><span class="n">_intersection</span><span class="p">)</span> <span class="c1"># same as gr.intersect(gr2)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 226987603 | 226987617 | U0         |         0 | +            |</span>
<span class="go">| chr8         |  38747236 |  38747251 | U0         |         0 | -            |</span>
<span class="go">| chr15        |  26105515 |  26105518 | U0         |         0 | +            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f2</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         2 | a          |         0 | +            |</span>
<span class="go">| chr1         |         6 |         7 | b          |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">apply_pair</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">df</span><span class="p">,</span> <span class="n">df2</span><span class="p">:</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">df2</span><span class="p">)),</span> <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">{(&#39;chr1&#39;, &#39;+&#39;): (2, 2), (&#39;chr1&#39;, &#39;-&#39;): (1, 2)}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.as_df">
<span class="sig-name descname"><span class="pre">as_df</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.as_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.as_df" title="Permalink to this definition"></a></dt>
<dd><p>Return PyRanges as DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A DataFrame natural sorted on Chromosome and Strand. The ordering of rows within
chromosomes and strands is preserved.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.df" title="pyranges.pyranges.PyRanges.df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.df</span></code></a></dt><dd><p>Return PyRanges as DataFrame.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span> <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Gene       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|            1 |         1 |         3 | A          |</span>
<span class="go">|            1 |         2 |         3 | B          |</span>
<span class="go">|            2 |         3 |        10 | C          |</span>
<span class="go">|            2 |         9 |        12 | D          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 4 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">as_df</span><span class="p">()</span>
<span class="go">  Chromosome  Start  End Gene</span>
<span class="go">0          1      1    3    A</span>
<span class="go">1          1      2    3    B</span>
<span class="go">2          2      3   10    C</span>
<span class="go">3          2      9   12    D</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.assign">
<span class="sig-name descname"><span class="pre">assign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.assign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.assign" title="Permalink to this definition"></a></dt>
<dd><p>Add or replace a column.</p>
<p>Does not change the original PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col</strong> (<em>str</em>) – Name of column.</p></li>
<li><p><strong>f</strong> (<em>function</em>) – Function to create new column.</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use
chromosome/strand pairs if the PyRanges is stranded.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the PyRanges with the column inserted.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span><span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|            1 |         1 |         3 | a          |</span>
<span class="go">|            1 |         2 |         5 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="s2">&quot;Blabla&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">Chromosome</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_yadayada&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Name       | Blabla     |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------+------------|</span>
<span class="go">|            1 |         1 |         3 | a          | 1_yadayada |</span>
<span class="go">|            1 |         2 |         5 | b          | 1_yadayada |</span>
<span class="go">+--------------+-----------+-----------+------------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<p>Note that assigning to an existing name replaces the column:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="k">lambda</span> <span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">Start</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;sep&quot;</span><span class="p">]</span> <span class="o">+</span>
<span class="gp">... </span><span class="n">df</span><span class="o">.</span><span class="n">Name</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">capitalize</span><span class="p">(),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|            1 |         1 |         3 | 1_A        |</span>
<span class="go">|            1 |         2 |         5 | 2_B        |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.boundaries">
<span class="sig-name descname"><span class="pre">boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group_by</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.pyranges.PyRanges.boundaries" title="Permalink to this definition"></a></dt>
<dd><p>Return the boundaries of groups of intervals (e.g. transcripts)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>group_by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Name(s) of column(s) to group intervals</p></li>
<li><p><strong>agg</strong> (<em>dict</em><em> or </em><em>None</em>) – Defines how to aggregate metadata columns. Provided as
dictionary of column names -&gt; functions, function names or list of such,
as accepted by the Pandas.DataFrame.agg method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>One interval per group, with the min(Start) and max(End) of the group</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">110</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">40</span><span class="p">,</span> <span class="mi">68</span><span class="p">,</span> <span class="mi">130</span><span class="p">],</span> <span class="s2">&quot;transcript_id&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;tr1&quot;</span><span class="p">,</span> <span class="s2">&quot;tr1&quot;</span><span class="p">,</span> <span class="s2">&quot;tr2&quot;</span><span class="p">],</span> <span class="s2">&quot;meta&quot;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">]</span> <span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">length</span><span class="o">=</span><span class="n">gr</span><span class="o">.</span><span class="n">lengths</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+-----------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End | transcript_id   |      meta |    length |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)        |   (int64) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------------+-----------+-----------|</span>
<span class="go">|            1 |         1 |        40 | tr1             |         1 |        39 |</span>
<span class="go">|            1 |        60 |        68 | tr1             |         2 |         8 |</span>
<span class="go">|            1 |       110 |       130 | tr2             |         3 |        20 |</span>
<span class="go">+--------------+-----------+-----------+-----------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">boundaries</span><span class="p">(</span><span class="s2">&quot;transcript_id&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)        |</span>
<span class="go">|--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 |         1 |        68 | tr1             |</span>
<span class="go">|            1 |       110 |       130 | tr2             |</span>
<span class="go">+--------------+-----------+-----------+-----------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">boundaries</span><span class="p">(</span><span class="s2">&quot;transcript_id&quot;</span><span class="p">,</span> <span class="n">agg</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;length&quot;</span><span class="p">:</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span> <span class="s2">&quot;meta&quot;</span><span class="p">:</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">})</span>
<span class="go">+--------------+-----------+-----------+-----------------+------------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End | transcript_id   | meta       |    length |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)        | (object)   |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------------+------------+-----------|</span>
<span class="go">|            1 |         1 |        68 | tr1             | a,b        |        47 |</span>
<span class="go">|            1 |       110 |       130 | tr2             | c          |        20 |</span>
<span class="go">+--------------+-----------+-----------+-----------------+------------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.calculate_frame">
<span class="sig-name descname"><span class="pre">calculate_frame</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">by</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.pyranges.PyRanges.calculate_frame" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the frame of each genomic interval, assuming all are coding sequences (CDS), and add it as column inplace.</p>
<p>After this, the input Pyranges will contain an added “Frame” column, which determines the base of the CDS that is the first base of a codon.
Resulting values are in range between 0 and 2 included. 0 indicates that the first base of the CDS is the first base of a codon,
1 indicates the second base and 2 indicates the third base of the CDS.
While the 5’-most interval of each transcript has always 0 frame, the following ones may have any of these values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em>) – Column(s) to group by the intervals: coding exons belonging to the same transcript have the same values in this/these column(s).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The “Frame” column is added inplace.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span><span class="s2">&quot;+&quot;</span><span class="p">,</span><span class="s2">&quot;+&quot;</span><span class="p">,</span><span class="s2">&quot;-&quot;</span><span class="p">,</span><span class="s2">&quot;-&quot;</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">31</span><span class="p">,</span><span class="mi">52</span><span class="p">,</span><span class="mi">101</span><span class="p">,</span><span class="mi">201</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span><span class="mi">45</span><span class="p">,</span><span class="mi">90</span><span class="p">,</span><span class="mi">130</span><span class="p">,</span><span class="mi">218</span><span class="p">],</span>
<span class="gp">... </span>                 <span class="s2">&quot;transcript_id&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;t1&quot;</span><span class="p">,</span><span class="s2">&quot;t1&quot;</span><span class="p">,</span><span class="s2">&quot;t1&quot;</span><span class="p">,</span><span class="s2">&quot;t2&quot;</span><span class="p">,</span><span class="s2">&quot;t2&quot;</span><span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |         1 |        10 | t1              |</span>
<span class="go">|            1 | +            |        31 |        45 | t1              |</span>
<span class="go">|            1 | +            |        52 |        90 | t1              |</span>
<span class="go">|            2 | -            |       101 |       130 | t2              |</span>
<span class="go">|            2 | -            |       201 |       218 | t2              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 5 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">calculate_frame</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+-----------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |     Frame |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) | (object)        |   (int32) |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------+-----------|</span>
<span class="go">|            1 | +            |         1 |        10 | t1              |         0 |</span>
<span class="go">|            1 | +            |        31 |        45 | t1              |         0 |</span>
<span class="go">|            1 | +            |        52 |        90 | t1              |         2 |</span>
<span class="go">|            2 | -            |       101 |       130 | t2              |         2 |</span>
<span class="go">|            2 | -            |       201 |       218 | t2              |         0 |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+-----------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 6 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.cluster">
<span class="sig-name descname"><span class="pre">cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.cluster" title="Permalink to this definition"></a></dt>
<dd><p>Give overlapping intervals a common id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to ignore strand information if PyRanges is stranded.</p></li>
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>default None</em>) – Only intervals with an equal value in column(s) <cite>by</cite> are clustered.</p></li>
<li><p><strong>slack</strong> (<em>int</em><em>, </em><em>default 0</em>) – Consider intervals separated by less than <cite>slack</cite> to be in the same cluster. If <cite>slack</cite>
is negative, intervals overlapping less than <cite>slack</cite> are not considered to be in the
same cluster.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with an ID-column “Cluster” added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Bookended intervals (i.e. the End of a PyRanges interval is the Start of
another one) are by default considered to overlap.
Avoid this with slack=-1.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.merge" title="pyranges.pyranges.PyRanges.merge"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.merge</span></code></a></dt><dd><p>combine overlapping intervals into one</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span> <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |      Gene |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         3 |         1 |</span>
<span class="go">|            1 |         2 |         3 |         2 |</span>
<span class="go">|            1 |         3 |        10 |         3 |</span>
<span class="go">|            1 |         9 |        12 |         3 |</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">cluster</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |      Gene |   Cluster |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         3 |         1 |         1 |</span>
<span class="go">|            1 |         2 |         3 |         2 |         1 |</span>
<span class="go">|            1 |         3 |        10 |         3 |         1 |</span>
<span class="go">|            1 |         9 |        12 |         3 |         1 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |      Gene |   Cluster |     Count |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (int32) |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         3 |         1 |         1 |         1 |</span>
<span class="go">|            1 |         2 |         3 |         2 |         2 |         1 |</span>
<span class="go">|            1 |         3 |        10 |         3 |         3 |         2 |</span>
<span class="go">|            1 |         9 |        12 |         3 |         3 |         2 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<p>Avoid clustering bookended intervals with slack=-1:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">slack</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |      Gene |   Cluster |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         3 |         1 |         1 |</span>
<span class="go">|            1 |         2 |         3 |         2 |         1 |</span>
<span class="go">|            1 |         3 |        10 |         3 |         2 |</span>
<span class="go">|            1 |         9 |        12 |         3 |         2 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()[[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;Source&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span><span class="o">.</span><span class="n">cluster</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;Source&quot;</span><span class="p">])</span>
<span class="go">+--------------+--------------+---------------+-----------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   | Feature      | Source        | Start     | End       | Strand       | Cluster   |</span>
<span class="go">| (category)   | (category)   | (object)      | (int32)   | (int32)   | (category)   | (int32)   |</span>
<span class="go">|--------------+--------------+---------------+-----------+-----------+--------------+-----------|</span>
<span class="go">| 1            | CDS          | ensembl       | 69090     | 70005     | +            | 1         |</span>
<span class="go">| 1            | CDS          | ensembl       | 925941    | 926013    | +            | 2         |</span>
<span class="go">| 1            | CDS          | ensembl       | 925941    | 926013    | +            | 2         |</span>
<span class="go">| 1            | CDS          | ensembl       | 925941    | 926013    | +            | 2         |</span>
<span class="go">| ...          | ...          | ...           | ...       | ...       | ...          | ...       |</span>
<span class="go">| 1            | transcript   | havana_tagene | 167128    | 169240    | -            | 1142      |</span>
<span class="go">| 1            | transcript   | mirbase       | 17368     | 17436     | -            | 1143      |</span>
<span class="go">| 1            | transcript   | mirbase       | 187890    | 187958    | -            | 1144      |</span>
<span class="go">| 1            | transcript   | mirbase       | 632324    | 632413    | -            | 1145      |</span>
<span class="go">+--------------+--------------+---------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.copy" title="Permalink to this definition"></a></dt>
<dd><p>Make a deep copy of the PyRanges.</p>
<p class="rubric">Notes</p>
<p>See the pandas docs for deep-copying caveats.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.count_overlaps">
<span class="sig-name descname"><span class="pre">count_overlaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_nonoverlapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NumberOverlaps'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.count_overlaps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.count_overlaps" title="Permalink to this definition"></a></dt>
<dd><p>Count number of overlaps per interval.</p>
<p>Count how many intervals in self overlap with those in other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strandedness</strong> (<em>{&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>None</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to perform the operation on the same, opposite or no strand. Use False to
ignore the strand. None means use “same” if both PyRanges are stranded, otherwise
ignore.</p></li>
<li><p><strong>keep_nonoverlapping</strong> (<em>bool</em><em>, </em><em>default True</em>) – Keep intervals without overlaps.</p></li>
<li><p><strong>overlap_col</strong> (<em>str</em><em>, </em><em>default &quot;NumberOverlaps&quot;</em>) – Name of column with overlap counts.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with a column of overlaps added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.coverage" title="pyranges.pyranges.PyRanges.coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.coverage</span></code></a></dt><dd><p>find coverage of PyRanges</p>
</dd>
<dt><a class="reference internal" href="../index.html#pyranges.count_overlaps" title="pyranges.count_overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.count_overlaps</span></code></a></dt><dd><p>count overlaps from multiple PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span><span class="o">.</span><span class="n">drop</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |</span>
<span class="go">| chr1         |         8 |         9 | +            |</span>
<span class="go">| chr1         |         5 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f2</span><span class="p">()</span><span class="o">.</span><span class="n">drop</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         2 | +            |</span>
<span class="go">| chr1         |         6 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">count_overlaps</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">overlap_col</span><span class="o">=</span><span class="s2">&quot;Count&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |     Count |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------|</span>
<span class="go">| chr1         |         3 |         6 | +            |         0 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         0 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         1 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.coverage">
<span class="sig-name descname"><span class="pre">coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_nonoverlapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NumberOverlaps'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FractionOverlaps'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.coverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.coverage" title="Permalink to this definition"></a></dt>
<dd><p>Count number of overlaps and their fraction per interval.</p>
<p>Count how many intervals in self overlap with those in other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strandedness</strong> (<em>{&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>None</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to perform the operation on the same, opposite or no strand. Use False to
ignore the strand. None means use “same” if both PyRanges are stranded, otherwise
ignore.</p></li>
<li><p><strong>keep_nonoverlapping</strong> (<em>bool</em><em>, </em><em>default True</em>) – Keep intervals without overlaps.</p></li>
<li><p><strong>overlap_col</strong> (<em>str</em><em>, </em><em>default &quot;NumberOverlaps&quot;</em>) – Name of column with overlap counts.</p></li>
<li><p><strong>fraction_col</strong> (<em>str</em><em>, </em><em>default &quot;FractionOverlaps&quot;</em>) – Name of column with fraction of counts.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with a column of overlaps added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.count_overlaps" title="pyranges.count_overlaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.count_overlaps</span></code></a></dt><dd><p>count overlaps from multiple PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span>  <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         3 |         6 |</span>
<span class="go">|            1 |         8 |         9 |</span>
<span class="go">|            1 |         5 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         2 |</span>
<span class="go">|            1 |         6 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">coverage</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">overlap_col</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">fraction_col</span><span class="o">=</span><span class="s2">&quot;F&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-------------+</span>
<span class="go">|   Chromosome |     Start |       End |         C |           F |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (float64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-------------|</span>
<span class="go">|            1 |         3 |         6 |         0 |         0   |</span>
<span class="go">|            1 |         8 |         9 |         0 |         0   |</span>
<span class="go">|            1 |         5 |         7 |         1 |         0.5 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.drop">
<span class="sig-name descname"><span class="pre">drop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">drop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">like</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.drop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.drop" title="Permalink to this definition"></a></dt>
<dd><p>Drop column(s).</p>
<p>If no arguments are given, all the columns except Chromosome, Start, End and Strand are
dropped.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>drop</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>default None</em>) – Columns to drop.</p></li>
<li><p><strong>like</strong> (<em>str</em><em>, </em><em>default None</em>) – Regex-string matching columns to drop. Matches with Chromosome, Start, End or Strand
are ignored.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.unstrand" title="pyranges.pyranges.PyRanges.unstrand"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.unstrand</span></code></a></dt><dd><p>drop strand information</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">],</span> <span class="s2">&quot;Count&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       |     Count | Type       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   |   (int64) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+------------|</span>
<span class="go">|            1 |         1 |         5 | +            |         1 | exon       |</span>
<span class="go">|            1 |         4 |         6 | -            |         2 | exon       |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">|            1 |         1 |         5 | +            |</span>
<span class="go">|            1 |         4 |         6 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<p>Matches with position-columns are ignored:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">like</span><span class="o">=</span><span class="s2">&quot;Chromosome|Strand&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       |     Count | Type       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   |   (int64) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+------------|</span>
<span class="go">|            1 |         1 |         5 | +            |         1 | exon       |</span>
<span class="go">|            1 |         4 |         6 | -            |         2 | exon       |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">like</span><span class="o">=</span><span class="s2">&quot;e$&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       |     Count |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------|</span>
<span class="go">|            1 |         1 |         5 | +            |         1 |</span>
<span class="go">|            1 |         4 |         6 | -            |         2 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.drop_duplicate_positions">
<span class="sig-name descname"><span class="pre">drop_duplicate_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'first'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.drop_duplicate_positions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.drop_duplicate_positions" title="Permalink to this definition"></a></dt>
<dd><p>Return PyRanges with duplicate postion rows removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to take strand-information into account when considering duplicates.</p></li>
<li><p><strong>keep</strong> (<em>{&quot;first&quot;</em><em>, </em><em>&quot;last&quot;</em><em>, </em><em>False}</em>) – Whether to keep first, last or drop all duplicates.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;&#39;&#39;Chromosome Start End Strand Name</span>
<span class="gp">... </span><span class="s1">1 1 2 + A</span>
<span class="gp">... </span><span class="s1">1 1 2 - B</span>
<span class="gp">... </span><span class="s1">1 1 2 + Z&#39;&#39;&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------|</span>
<span class="go">|            1 |         1 |         2 | +            | A          |</span>
<span class="go">|            1 |         1 |         2 | +            | Z          |</span>
<span class="go">|            1 |         1 |         2 | -            | B          |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop_duplicate_positions</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------|</span>
<span class="go">|            1 |         1 |         2 | +            | A          |</span>
<span class="go">|            1 |         1 |         2 | -            | B          |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop_duplicate_positions</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;last&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------|</span>
<span class="go">|            1 |         1 |         2 | +            | Z          |</span>
<span class="go">|            1 |         1 |         2 | -            | B          |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<p>Note that the reverse strand is considered to be behind the forward strand:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop_duplicate_positions</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;last&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Strand       | Name       |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (category)   | (object)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+------------|</span>
<span class="go">|            1 |         1 |         2 | -            | B          |</span>
<span class="go">+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 1 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">drop_duplicate_positions</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Empty PyRanges</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ext</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.extend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.extend" title="Permalink to this definition"></a></dt>
<dd><p>Extend the intervals from the ends.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ext</strong> (<em>int</em><em> or </em><em>dict</em><em> of </em><em>ints with &quot;3&quot; and/or &quot;5&quot; as keys.</em>) – The number of nucleotides to extend the ends with.
If an int is provided, the same extension is applied to both
the start and end of intervals, while a dict input allows to control
differently the two ends. Note also that 5’ and 3’ extensions take
the strand into account, if the intervals are stranded.</p></li>
<li><p><strong>group_by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>default: None</em>) – group intervals by these column name(s), so that the extension is applied
only to the left-most and/or right-most interval.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.subsequence" title="pyranges.pyranges.PyRanges.subsequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.subsequence</span></code></a></dt><dd><p>obtain subsequences of intervals</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.spliced_subsequence" title="pyranges.pyranges.PyRanges.spliced_subsequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.spliced_subsequence</span></code></a></dt><dd><p>obtain subsequences of intervals, providing transcript-level coordinates</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span>
<span class="gp">... </span>     <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |</span>
<span class="go">| chr1         |         8 |         9 | +            |</span>
<span class="go">| chr1         |         5 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         0 |        10 | +            |</span>
<span class="go">| chr1         |         4 |        13 | +            |</span>
<span class="go">| chr1         |         1 |        11 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">extend</span><span class="p">({</span><span class="s2">&quot;3&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">})</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         7 | +            |</span>
<span class="go">| chr1         |         8 |        10 | +            |</span>
<span class="go">| chr1         |         4 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">extend</span><span class="p">({</span><span class="s2">&quot;3&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         7 | +            |</span>
<span class="go">| chr1         |         6 |        10 | +            |</span>
<span class="go">| chr1         |         4 |         9 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">AssertionError</span>: <span class="n">Some intervals are negative or zero length after applying extend!</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.five_end">
<span class="sig-name descname"><span class="pre">five_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.five_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.five_end" title="Permalink to this definition"></a></dt>
<dd><p>Return the five prime end of intervals.</p>
<p>The five prime end is the start of a forward strand or the end of a reverse strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>PyRanges with the five prime ends</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Requires the PyRanges to be stranded.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.three_end" title="pyranges.pyranges.PyRanges.three_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.three_end</span></code></a></dt><dd><p>return the 3’ end</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         9 | +            |</span>
<span class="go">| chr1         |         5 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">five_end</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         4 | +            |</span>
<span class="go">| chr1         |         7 |         8 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.head">
<span class="sig-name descname"><span class="pre">head</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.head"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.head" title="Permalink to this definition"></a></dt>
<dd><p>Return the n first rows.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 8</em>) – Return n rows.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with the n first rows.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.tail" title="pyranges.pyranges.PyRanges.tail"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.tail</span></code></a></dt><dd><p>return the last rows</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.sample" title="pyranges.pyranges.PyRanges.sample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.sample</span></code></a></dt><dd><p>return random rows</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="go">| chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         |         0 | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         |         0 | +            |</span>
<span class="go">| chr1         | 216711011 | 216711036 | U0         |         0 | +            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.insert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.insert" title="Permalink to this definition"></a></dt>
<dd><p>Add one or more columns to the PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>Series</em><em>, </em><em>DataFrame</em><em> or </em><em>dict</em>) – Data to insert into the PyRanges. <cite>other</cite> must have the same number of rows as the PyRanges.</p></li>
<li><p><strong>loc</strong> (<em>int</em><em>, </em><em>default None</em><em>, </em><em>i.e. after last column</em><em> of </em><em>PyRanges.</em>) – Insertion index.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A copy of the PyRanges with the column(s) inserted starting at <cite>loc</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a Series, or a dict of Series is used, the Series must have a name.</p>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;L&quot;</span><span class="p">,</span> <span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">21</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| E            |         1 |         8 |</span>
<span class="go">| E            |         2 |        13 |</span>
<span class="go">| L            |         1 |         5 |</span>
<span class="go">| T            |         3 |        21 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 3 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Column&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |    Column |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------|</span>
<span class="go">| E            |         1 |         8 |         1 |</span>
<span class="go">| E            |         2 |        13 |         3 |</span>
<span class="go">| L            |         1 |         5 |         3 |</span>
<span class="go">| T            |         3 |        21 |         7 |</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 4 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;NY&quot;</span><span class="p">:</span> <span class="n">s</span><span class="p">,</span> <span class="s2">&quot;AN&quot;</span><span class="p">:</span> <span class="n">s</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span>
<span class="go">   NY  AN</span>
<span class="go">0   1   1</span>
<span class="go">1   3   3</span>
<span class="go">2   3   3</span>
<span class="go">3   7   7</span>
</pre></div>
</div>
<p>Note that the original PyRanges was not affected by previously inserting Column:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">| Chromosome   |        NY |        AN |     Start |       End |</span>
<span class="go">| (category)   |   (int64) |   (int64) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------|</span>
<span class="go">| E            |         1 |         1 |         1 |         8 |</span>
<span class="go">| E            |         3 |         3 |         2 |        13 |</span>
<span class="go">| L            |         3 |         3 |         1 |         5 |</span>
<span class="go">| T            |         7 |         7 |         3 |        21 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 5 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">arbitrary_result</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
<span class="gp">... </span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">Start</span> <span class="o">+</span> <span class="n">df</span><span class="o">.</span><span class="n">End</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Hi!&quot;</span><span class="p">),</span> <span class="n">as_pyranges</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">arbitrary_result</span>
<span class="go">{&#39;E&#39;: 1     9</span>
<span class="go">2    15</span>
<span class="go">Name: Hi!, dtype: int32, &#39;L&#39;: 0    6</span>
<span class="go">Name: Hi!, dtype: int32, &#39;T&#39;: 3    24</span>
<span class="go">Name: Hi!, dtype: int32}</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">arbitrary_result</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |       Hi! |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------|</span>
<span class="go">| E            |         1 |         8 |         9 |</span>
<span class="go">| E            |         2 |        13 |        15 |</span>
<span class="go">| L            |         1 |         5 |         6 |</span>
<span class="go">| T            |         3 |        21 |        24 |</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 4 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.intersect">
<span class="sig-name descname"><span class="pre">intersect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.intersect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.intersect" title="Permalink to this definition"></a></dt>
<dd><p>Return overlapping subintervals.</p>
<p>Returns the segments of the intervals in self which overlap with those in other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to intersect.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>how</strong> (<em>{None</em><em>, </em><em>&quot;first&quot;</em><em>, </em><em>&quot;last&quot;</em><em>, </em><em>&quot;containment&quot;}</em><em>, </em><em>default None</em><em>, </em><em>i.e. all</em>) – What intervals to report. By default reports all overlapping intervals. “containment”
reports intervals where the overlapping is contained within it.</p></li>
<li><p><strong>invert</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the intervals without overlaps.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PyRanges with overlapping subintervals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.set_intersect" title="pyranges.pyranges.PyRanges.set_intersect"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.set_intersect</span></code></a></dt><dd><p>set-intersect PyRanges</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.overlap" title="pyranges.pyranges.PyRanges.overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.overlap</span></code></a></dt><dd><p>report overlapping intervals</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | ID         |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|         chr1 |         1 |         3 | a          |</span>
<span class="go">|         chr1 |         4 |         9 | b          |</span>
<span class="go">|         chr1 |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|         chr1 |         2 |         3 |</span>
<span class="go">|         chr1 |         2 |         9 |</span>
<span class="go">|         chr1 |         9 |        10 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | ID         |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|         chr1 |         2 |         3 | a          |</span>
<span class="go">|         chr1 |         2 |         3 | a          |</span>
<span class="go">|         chr1 |         4 |         9 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | ID         |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|         chr1 |         2 |         3 | a          |</span>
<span class="go">|         chr1 |         4 |         9 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;containment&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         4 |         9 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.items">
<span class="sig-name descname"><span class="pre">items</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.items"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.items" title="Permalink to this definition"></a></dt>
<dd><p>Return the pairs of keys and DataFrames.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The dict mapping keys to DataFrames in the PyRanges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.chromosomes" title="pyranges.pyranges.PyRanges.chromosomes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.chromosomes</span></code></a></dt><dd><p>return the chromosomes</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.keys" title="pyranges.pyranges.PyRanges.keys"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.keys</span></code></a></dt><dd><p>return the keys</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.values" title="pyranges.pyranges.PyRanges.values"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.values</span></code></a></dt><dd><p>return the DataFrames in the PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="go">[((&#39;chr1&#39;, &#39;+&#39;),   Chromosome  Start  End       Name  Score Strand</span>
<span class="go">0       chr1      3    6  interval1      0      +</span>
<span class="go">2       chr1      8    9  interval3      0      +), ((&#39;chr1&#39;, &#39;-&#39;),   Chromosome  Start  End       Name  Score Strand</span>
<span class="go">1       chr1      5    7  interval2      0      -)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.join">
<span class="sig-name descname"><span class="pre">join</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_b'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply_strand_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.join"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.join" title="Permalink to this definition"></a></dt>
<dd><p>Join PyRanges on genomic location.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to join.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>how</strong> (<em>{None</em><em>, </em><em>&quot;left&quot;</em><em>, </em><em>&quot;right&quot;}</em><em>, </em><em>default None</em><em>, </em><em>i.e. &quot;inner&quot;</em>) – How to handle intervals without overlap. None means only keep overlapping intervals.
“left” keeps all intervals in self, “right” keeps all intervals in other.</p></li>
<li><p><strong>report_overlap</strong> (<em>bool</em><em>, </em><em>default False</em>) – Report amount of overlap in base pairs.</p></li>
<li><p><strong>slack</strong> (<em>int</em><em>, </em><em>default 0</em>) – Lengthen intervals in self before joining.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>default &quot;_b&quot;</em>) – Suffix to give overlapping columns in other.</p></li>
<li><p><strong>apply_strand_suffix</strong> (<em>bool</em><em>, </em><em>default None</em>) – If first pyranges is unstranded, but the second is not, the first will be given a strand column.
apply_strand_suffix makes the added strand column a regular data column instead by adding a suffix.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PyRanges appended with columns of another.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The chromosome from other will never be reported as it is always the same as in self.</p>
<p>As pandas did not have NaN for non-float datatypes until recently, “left” and “right” join
give non-overlapping rows the value -1 to avoid promoting columns to object. This will
change to NaN in a future version as general NaN becomes stable in pandas.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.new_position" title="pyranges.pyranges.PyRanges.new_position"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.new_position</span></code></a></dt><dd><p>give joined PyRanges new coordinates</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;interval1&#39;</span><span class="p">,</span> <span class="s1">&#39;interval3&#39;</span><span class="p">,</span> <span class="s1">&#39;interval2&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         1 |         2 | a          |</span>
<span class="go">| chr1         |         6 |         7 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |   Start_b |     End_b | Name_b     |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         6 |         7 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |   Start_b |     End_b | Name_b     |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         6 |         7 | b          |</span>
<span class="go">| chr1         |        -1 |        -1 | -1         |         1 |         2 | a          |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<p>With slack 1, bookended features are joined (see row 1):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">slack</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |   Start_b |     End_b | Name_b     |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         6 |         7 | b          |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         6 |         7 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.keys">
<span class="sig-name descname"><span class="pre">keys</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.keys"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.keys" title="Permalink to this definition"></a></dt>
<dd><p>Return the keys.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>Returns the keys (chromosomes or chromosome/strand pairs) as strings or tuples of strings</em></p></li>
<li><p><em>in natsorted order.</em></p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.chromosomes" title="pyranges.pyranges.PyRanges.chromosomes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.chromosomes</span></code></a></dt><dd><p>return the chromosomes</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[(&#39;chr1&#39;, &#39;+&#39;), (&#39;chr1&#39;, &#39;-&#39;), (&#39;chr2&#39;, &#39;+&#39;), (&#39;chr2&#39;, &#39;-&#39;), (&#39;chr3&#39;, &#39;+&#39;), (&#39;chr3&#39;, &#39;-&#39;), (&#39;chr4&#39;, &#39;+&#39;), (&#39;chr4&#39;, &#39;-&#39;), (&#39;chr5&#39;, &#39;+&#39;), (&#39;chr5&#39;, &#39;-&#39;), (&#39;chr6&#39;, &#39;+&#39;), (&#39;chr6&#39;, &#39;-&#39;), (&#39;chr7&#39;, &#39;+&#39;), (&#39;chr7&#39;, &#39;-&#39;), (&#39;chr8&#39;, &#39;+&#39;), (&#39;chr8&#39;, &#39;-&#39;), (&#39;chr9&#39;, &#39;+&#39;), (&#39;chr9&#39;, &#39;-&#39;), (&#39;chr10&#39;, &#39;+&#39;), (&#39;chr10&#39;, &#39;-&#39;), (&#39;chr11&#39;, &#39;+&#39;), (&#39;chr11&#39;, &#39;-&#39;), (&#39;chr12&#39;, &#39;+&#39;), (&#39;chr12&#39;, &#39;-&#39;), (&#39;chr13&#39;, &#39;+&#39;), (&#39;chr13&#39;, &#39;-&#39;), (&#39;chr14&#39;, &#39;+&#39;), (&#39;chr14&#39;, &#39;-&#39;), (&#39;chr15&#39;, &#39;+&#39;), (&#39;chr15&#39;, &#39;-&#39;), (&#39;chr16&#39;, &#39;+&#39;), (&#39;chr16&#39;, &#39;-&#39;), (&#39;chr17&#39;, &#39;+&#39;), (&#39;chr17&#39;, &#39;-&#39;), (&#39;chr18&#39;, &#39;+&#39;), (&#39;chr18&#39;, &#39;-&#39;), (&#39;chr19&#39;, &#39;+&#39;), (&#39;chr19&#39;, &#39;-&#39;), (&#39;chr20&#39;, &#39;+&#39;), (&#39;chr20&#39;, &#39;-&#39;), (&#39;chr21&#39;, &#39;+&#39;), (&#39;chr21&#39;, &#39;-&#39;), (&#39;chr22&#39;, &#39;+&#39;), (&#39;chr22&#39;, &#39;-&#39;), (&#39;chrX&#39;, &#39;+&#39;), (&#39;chrX&#39;, &#39;-&#39;), (&#39;chrY&#39;, &#39;+&#39;), (&#39;chrY&#39;, &#39;-&#39;)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">unstrand</span><span class="p">()</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;, &#39;chr4&#39;, &#39;chr5&#39;, &#39;chr6&#39;, &#39;chr7&#39;, &#39;chr8&#39;, &#39;chr9&#39;, &#39;chr10&#39;, &#39;chr11&#39;, &#39;chr12&#39;, &#39;chr13&#39;, &#39;chr14&#39;, &#39;chr15&#39;, &#39;chr16&#39;, &#39;chr17&#39;, &#39;chr18&#39;, &#39;chr19&#39;, &#39;chr20&#39;, &#39;chr21&#39;, &#39;chr22&#39;, &#39;chrX&#39;, &#39;chrY&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.k_nearest">
<span class="sig-name descname"><span class="pre">k_nearest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ties</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_b'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply_strand_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.k_nearest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.k_nearest" title="Permalink to this definition"></a></dt>
<dd><p>Find k nearest intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to find nearest interval in.</p></li>
<li><p><strong>k</strong> (<em>int</em><em> or </em><em>list/array/Series</em><em> of </em><em>int</em>) – Number of closest to return. If iterable, must be same length as PyRanges.</p></li>
<li><p><strong>ties</strong> (<em>{None</em><em>, </em><em>&quot;first&quot;</em><em>, </em><em>&quot;last&quot;</em><em>, </em><em>&quot;different&quot;}</em><em>, </em><em>default None</em>) – How to resolve ties, i.e. closest intervals with equal distance. None means that the k nearest intervals are kept.
“first” means that the first tie is kept, “last” meanst that the last is kept.
“different” means that all nearest intervals with the k unique nearest distances are kept.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are stranded,
otherwise ignore the strand information.</p></li>
<li><p><strong>overlap</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to include overlaps.</p></li>
<li><p><strong>how</strong> (<em>{None</em><em>, </em><em>&quot;upstream&quot;</em><em>, </em><em>&quot;downstream&quot;}</em><em>, </em><em>default None</em><em>, </em><em>i.e. both directions</em>) – Whether to only look for nearest in one direction. Always with respect to the PyRanges
it is called on.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>default &quot;_b&quot;</em>) – Suffix to give columns with shared name in other.</p></li>
<li><p><strong>apply_strand_suffix</strong> (<em>bool</em><em>, </em><em>default None</em>) – If first pyranges is unstranded, but the second is not, the first will be given a strand column.
apply_strand_suffix makes the added strand column a regular data column instead by adding a suffix.</p></li>
</ul>
</dd>
</dl>
<p>nb_cpu: int, default 1</p>
<blockquote>
<div><p>How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A PyRanges with columns of nearest interval horizontally appended.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>nearest also exists, and is more performant.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.new_position" title="pyranges.pyranges.PyRanges.new_position"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.new_position</span></code></a></dt><dd><p>give joined PyRanges new coordinates</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.nearest" title="pyranges.pyranges.PyRanges.nearest"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.nearest</span></code></a></dt><dd><p>find nearest intervals</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |</span>
<span class="go">| chr1         |         8 |         9 | +            |</span>
<span class="go">| chr1         |         5 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         2 | +            |</span>
<span class="go">| chr1         |         6 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int32) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |         6 |         7 | -            |          1 |</span>
<span class="go">| chr1         |         3 |         6 | +            |         1 |         2 | +            |         -2 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         6 |         7 | -            |         -2 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         1 |         2 | +            |         -7 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         1 |         2 | +            |          4 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 6 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;upstream&quot;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int32) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |         1 |         2 | +            |         -2 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         6 |         7 | -            |         -2 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         1 |         2 | +            |         -7 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int32) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |         6 |         7 | -            |          1 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         6 |         7 | -            |         -2 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         1 |         2 | +            |         -7 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">+</span> <span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">+</span> <span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span>
<span class="gp">... </span>      <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">+</span> <span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span> <span class="o">+</span> <span class="p">[</span><span class="mi">11</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">)}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="p">,</span> <span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d1</span><span class="p">),</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d2</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |         5 |         6 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |        ID |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         1 |         3 |         0 |</span>
<span class="go">|            1 |         1 |         3 |         1 |</span>
<span class="go">|            1 |         5 |         7 |         2 |</span>
<span class="go">|            1 |         5 |         7 |         3 |</span>
<span class="go">|            1 |         9 |        11 |         4 |</span>
<span class="go">|            1 |         9 |        11 |         5 |</span>
<span class="go">+--------------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 6 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="go">|            1 |         5 |         6 |         5 |         7 |         2 |          0 |</span>
<span class="go">|            1 |         5 |         6 |         5 |         7 |         3 |          0 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ties</span><span class="o">=</span><span class="s2">&quot;different&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="go">|            1 |         5 |         6 |         5 |         7 |         2 |          0 |</span>
<span class="go">|            1 |         5 |         6 |         5 |         7 |         3 |          0 |</span>
<span class="go">|            1 |         5 |         6 |         1 |         3 |         1 |         -3 |</span>
<span class="go">|            1 |         5 |         6 |         1 |         3 |         0 |         -3 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">ties</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="go">|            1 |         5 |         6 |         5 |         7 |         2 |          0 |</span>
<span class="go">|            1 |         5 |         6 |         1 |         3 |         1 |         -3 |</span>
<span class="go">|            1 |         5 |         6 |         9 |        11 |         4 |          4 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">k_nearest</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End |   Start_b |     End_b |        ID |   Distance |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int32) |   (int32) |   (int64) |    (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------+------------|</span>
<span class="go">|            1 |         5 |         6 |         1 |         3 |         1 |         -3 |</span>
<span class="go">|            1 |         5 |         6 |         1 |         3 |         0 |         -3 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.lengths">
<span class="sig-name descname"><span class="pre">lengths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">as_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.lengths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.lengths" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of each interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>as_dict</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return lengths as Series or dict of Series per key.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Series or dict of Series with the lengths of each interval.</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.lengths" title="pyranges.pyranges.PyRanges.lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.lengths</span></code></a></dt><dd><p>return the intervals lengths</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">lengths</span><span class="p">()</span>
<span class="go">0    3</span>
<span class="go">1    1</span>
<span class="go">2    2</span>
<span class="go">dtype: int32</span>
</pre></div>
</div>
<p>To find the length of the genome covered by the intervals, use merge first:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">Length</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">lengths</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |    Length |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------+-----------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |         3 |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |         1 |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |         2 |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.make_stranded">
<span class="sig-name descname"><span class="pre">make_stranded</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'to_plus'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#pyranges.pyranges.PyRanges.make_stranded" title="Permalink to this definition"></a></dt>
<dd><p>Returns a PyRanges copy which is ensured to be stranded.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>method</strong> (<em>{&quot;to_plus&quot;</em><em>, </em><em>&quot;remove&quot;}</em><em>, </em><em>default &quot;to_plus&quot;</em>) – If “to_plus” (default), if a Strand column is present and contains any invalid values
(anything other than “+” or “-“), they are turned into “+”.
If “remove”, rows with invalid Strand value are omitted</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A stranded PyRanges</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.max_disjoint">
<span class="sig-name descname"><span class="pre">max_disjoint</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.max_disjoint"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.max_disjoint" title="Permalink to this definition"></a></dt>
<dd><p>Find the maximal disjoint set of intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Find the max disjoint set separately for each strand.</p></li>
<li><p><strong>slack</strong> (<em>int</em><em>, </em><em>default 0</em>) – Consider intervals within a distance of slack to be overlapping.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with maximal disjoint set of intervals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">max_disjoint</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.merge">
<span class="sig-name descname"><span class="pre">merge</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.merge"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.merge" title="Permalink to this definition"></a></dt>
<dd><p>Merge overlapping intervals into one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Only merge intervals on same strand.</p></li>
<li><p><strong>count</strong> (<em>bool</em><em>, </em><em>default False</em>) – Count intervals in each superinterval.</p></li>
<li><p><strong>count_col</strong> (<em>str</em><em>, </em><em>default &quot;Count&quot;</em>) – Name of column with counts.</p></li>
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>default None</em>) – Only merge intervals with equal values in these columns.</p></li>
<li><p><strong>slack</strong> (<em>int</em><em>, </em><em>default 0</em>) – Allow this many nucleotides between each interval to merge.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with superintervals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To avoid losing metadata, use cluster instead. If you want to perform a reduction function
on the metadata, use pandas groupby.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.cluster" title="pyranges.pyranges.PyRanges.cluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.cluster</span></code></a></dt><dd><p>annotate overlapping intervals with common ID</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()[[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="go">| 1            | gene         | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | transcript   | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | exon         | 11868     | 12227     | +            | DDX11L1     |</span>
<span class="go">| 1            | exon         | 12612     | 12721     | +            | DDX11L1     |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="go">| 1            | gene         | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | transcript   | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1179364   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1173055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">count_col</span><span class="o">=</span><span class="s2">&quot;Count&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | Count     |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (int32)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------|</span>
<span class="go">| 1            | 11868     | 14409     | +            | 12        |</span>
<span class="go">| 1            | 29553     | 31109     | +            | 11        |</span>
<span class="go">| 1            | 52472     | 53312     | +            | 3         |</span>
<span class="go">| 1            | 57597     | 64116     | +            | 7         |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...       |</span>
<span class="go">| 1            | 1062207   | 1063288   | -            | 4         |</span>
<span class="go">| 1            | 1070966   | 1074306   | -            | 10        |</span>
<span class="go">| 1            | 1081817   | 1116361   | -            | 319       |</span>
<span class="go">| 1            | 1173055   | 1179555   | -            | 4         |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 62 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+--------------+-----------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | Feature      | Count     |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (category)   | (int32)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+--------------+-----------|</span>
<span class="go">| 1            | 65564     | 65573     | +            | CDS          | 1         |</span>
<span class="go">| 1            | 69036     | 70005     | +            | CDS          | 2         |</span>
<span class="go">| 1            | 924431    | 924948    | +            | CDS          | 1         |</span>
<span class="go">| 1            | 925921    | 926013    | +            | CDS          | 11        |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...          | ...       |</span>
<span class="go">| 1            | 1062207   | 1063288   | -            | transcript   | 1         |</span>
<span class="go">| 1            | 1070966   | 1074306   | -            | transcript   | 1         |</span>
<span class="go">| 1            | 1081817   | 1116361   | -            | transcript   | 19        |</span>
<span class="go">| 1            | 1173055   | 1179555   | -            | transcript   | 1         |</span>
<span class="go">+--------------+-----------+-----------+--------------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 748 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">],</span> <span class="n">count</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+--------------+-------------+-----------+</span>
<span class="go">| Chromosome   | Start     | End       | Strand       | Feature      | gene_name   | Count     |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (category)   | (category)   | (object)    | (int32)   |</span>
<span class="go">|--------------+-----------+-----------+--------------+--------------+-------------+-----------|</span>
<span class="go">| 1            | 1020172   | 1020373   | +            | CDS          | AGRN        | 1         |</span>
<span class="go">| 1            | 1022200   | 1022462   | +            | CDS          | AGRN        | 2         |</span>
<span class="go">| 1            | 1034555   | 1034703   | +            | CDS          | AGRN        | 2         |</span>
<span class="go">| 1            | 1035276   | 1035324   | +            | CDS          | AGRN        | 4         |</span>
<span class="go">| ...          | ...       | ...       | ...          | ...          | ...         | ...       |</span>
<span class="go">| 1            | 347981    | 348366    | -            | transcript   | RPL23AP24   | 1         |</span>
<span class="go">| 1            | 1173055   | 1179555   | -            | transcript   | TTLL10-AS1  | 1         |</span>
<span class="go">| 1            | 14403     | 29570     | -            | transcript   | WASH7P      | 1         |</span>
<span class="go">| 1            | 185216    | 195411    | -            | transcript   | WASH9P      | 1         |</span>
<span class="go">+--------------+-----------+-----------+--------------+--------------+-------------+-----------+</span>
<span class="go">Stranded PyRanges object has 807 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.mp">
<span class="sig-name descname"><span class="pre">mp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.mp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.mp" title="Permalink to this definition"></a></dt>
<dd><p>Merge location and print.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.mpc">
<span class="sig-name descname"><span class="pre">mpc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.mpc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.mpc" title="Permalink to this definition"></a></dt>
<dd><p>Merge location, print and return self.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.msp">
<span class="sig-name descname"><span class="pre">msp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.msp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.msp" title="Permalink to this definition"></a></dt>
<dd><p>Sort on location, merge location info and print.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.mspc">
<span class="sig-name descname"><span class="pre">mspc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.mspc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.mspc" title="Permalink to this definition"></a></dt>
<dd><p>Sort on location, merge location, print and return self.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.nearest">
<span class="sig-name descname"><span class="pre">nearest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'_b'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">apply_strand_suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.nearest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.nearest" title="Permalink to this definition"></a></dt>
<dd><p>Find closest interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to find nearest interval in.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>overlap</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to include overlaps.</p></li>
<li><p><strong>how</strong> (<em>{None</em><em>, </em><em>&quot;upstream&quot;</em><em>, </em><em>&quot;downstream&quot;}</em><em>, </em><em>default None</em><em>, </em><em>i.e. both directions</em>) – Whether to only look for nearest in one direction. Always with respect to the PyRanges
it is called on.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>default &quot;_b&quot;</em>) – Suffix to give columns with shared name in other.</p></li>
<li><p><strong>apply_strand_suffix</strong> (<em>bool</em><em>, </em><em>default None</em>) – If first pyranges is unstranded, but the second is not, the first will be given the strand column of the second.
apply_strand_suffix makes the added strand column a regular data column instead by adding a suffix.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PyRanges with columns representing nearest interval horizontally appended.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>A k_nearest also exists, but is less performant.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.new_position" title="pyranges.pyranges.PyRanges.new_position"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.new_position</span></code></a></dt><dd><p>give joined PyRanges new coordinates</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.k_nearest" title="pyranges.pyranges.PyRanges.k_nearest"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.k_nearest</span></code></a></dt><dd><p>find k nearest intervals</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |</span>
<span class="go">| chr1         |         8 |         9 | +            |</span>
<span class="go">| chr1         |         5 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         2 | +            |</span>
<span class="go">| chr1         |         6 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">nearest</span><span class="p">(</span><span class="n">f2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |         6 |         7 | -            |          1 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         6 |         7 | -            |          2 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">nearest</span><span class="p">(</span><span class="n">f2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;upstream&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |   Start_b |     End_b | Strand_b     |   Distance |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int32) |   (int32) | (category)   |    (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |         1 |         2 | +            |          2 |</span>
<span class="go">| chr1         |         8 |         9 | +            |         6 |         7 | -            |          2 |</span>
<span class="go">| chr1         |         5 |         7 | -            |         6 |         7 | -            |          0 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+-----------+--------------+------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.new_position">
<span class="sig-name descname"><span class="pre">new_position</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.new_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.new_position" title="Permalink to this definition"></a></dt>
<dd><p>Give new position.</p>
<p>The operation join produces a PyRanges with two pairs of start coordinates and two pairs of
end coordinates. This operation uses these to give the PyRanges a new position.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>new_pos</strong> (<em>{&quot;union&quot;</em><em>, </em><em>&quot;intersection&quot;</em><em>, </em><em>&quot;swap&quot;}</em>) – Change of coordinates.</p></li>
<li><p><strong>columns</strong> (<em>tuple</em><em> of </em><em>str</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – The name of the coordinate columns. By default uses the two first columns containing
“Start” and the two first columns containing “End”.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.join" title="pyranges.pyranges.PyRanges.join"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.join</span></code></a></dt><dd><p>combine two PyRanges horizontally with SQL-style joins.</p>
</dd>
</dl>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>PyRanges with new coordinates.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         3 |         6 |</span>
<span class="go">| chr1         |         8 |         9 |</span>
<span class="go">| chr1         |         5 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>                    <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         1 |         4 |</span>
<span class="go">| chr1         |         6 |         7 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">j</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">j</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |   Start_b |     End_b |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------|</span>
<span class="go">| chr1         |         3 |         6 |         1 |         4 |</span>
<span class="go">| chr1         |         5 |         7 |         6 |         7 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">j</span><span class="o">.</span><span class="n">new_position</span><span class="p">(</span><span class="s2">&quot;swap&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |   Start_b |     End_b |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------|</span>
<span class="go">| chr1         |         1 |         4 |         3 |         6 |</span>
<span class="go">| chr1         |         6 |         7 |         5 |         7 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">j</span><span class="o">.</span><span class="n">new_position</span><span class="p">(</span><span class="s2">&quot;union&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mp</span><span class="p">()</span>
<span class="go">+--------------------+-----------+-----------+</span>
<span class="go">| - Position -       |   Start_b |     End_b |</span>
<span class="go">| (Multiple types)   |   (int32) |   (int32) |</span>
<span class="go">|--------------------+-----------+-----------|</span>
<span class="go">| chr1 1-6           |         1 |         4 |</span>
<span class="go">| chr1 5-7           |         6 |         7 |</span>
<span class="go">+--------------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">j</span><span class="o">.</span><span class="n">new_position</span><span class="p">(</span><span class="s2">&quot;intersection&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mp</span><span class="p">()</span>
<span class="go">+--------------------+-----------+-----------+</span>
<span class="go">| - Position -       |   Start_b |     End_b |</span>
<span class="go">| (Multiple types)   |   (int32) |   (int32) |</span>
<span class="go">|--------------------+-----------+-----------|</span>
<span class="go">| chr1 1-4           |         1 |         4 |</span>
<span class="go">| chr1 6-7           |         6 |         7 |</span>
<span class="go">+--------------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">j2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;B&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;D&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">j2</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |         A |         B |         C |         D |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (int64) |   (int64) |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         3 |         4 |         1 |         3 |         2 |         5 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">j2</span><span class="o">.</span><span class="n">new_position</span><span class="p">(</span><span class="s2">&quot;intersection&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">))</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |         A |         B |         C |         D |</span>
<span class="go">|   (category) |   (int32) |   (int32) |   (int64) |   (int64) |   (int64) |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+-----------+-----------+-----------+-----------|</span>
<span class="go">|            1 |         2 |         3 |         1 |         3 |         2 |         5 |</span>
<span class="go">+--------------+-----------+-----------+-----------+-----------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.overlap">
<span class="sig-name descname"><span class="pre">overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'first'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">invert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.overlap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.overlap" title="Permalink to this definition"></a></dt>
<dd><p>Return overlapping intervals.</p>
<p>Returns the intervals in self which overlap with those in other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to find overlaps with.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>how</strong> (<em>{&quot;first&quot;</em><em>, </em><em>&quot;containment&quot;</em><em>, </em><em>False</em><em>, </em><em>None}</em><em>, </em><em>default &quot;first&quot;</em>) – What intervals to report. By default reports every interval in self with overlap once.
“containment” reports all intervals where the overlapping is contained within it.</p></li>
<li><p><strong>invert</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the intervals without overlaps.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PyRanges with overlapping intervals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.intersect" title="pyranges.pyranges.PyRanges.intersect"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.intersect</span></code></a></dt><dd><p>report overlapping subintervals</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.set_intersect" title="pyranges.pyranges.PyRanges.set_intersect"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.set_intersect</span></code></a></dt><dd><p>set-intersect PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | ID         |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|         chr1 |         1 |         3 | a          |</span>
<span class="go">|         chr1 |         4 |         9 | b          |</span>
<span class="go">|         chr1 |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         2 |         3 |</span>
<span class="go">| chr1         |         2 |         9 |</span>
<span class="go">| chr1         |         9 |        10 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         1 |         3 | a          |</span>
<span class="go">| chr1         |         4 |         9 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         1 |         3 | a          |</span>
<span class="go">| chr1         |         1 |         3 | a          |</span>
<span class="go">| chr1         |         4 |         9 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;containment&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         4 |         9 | b          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">invert</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.pc">
<span class="sig-name descname"><span class="pre">pc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.pc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.pc" title="Permalink to this definition"></a></dt>
<dd><p>Print and return self.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.print">
<span class="sig-name descname"><span class="pre">print</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merge_position</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.print"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.print" title="Permalink to this definition"></a></dt>
<dd><p>Print the PyRanges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>default 8</em>) – The number of rows to print.</p></li>
<li><p><strong>merge_postion</strong> (<em>bool</em><em>, </em><em>default False</em>) – Print location in same column to save screen space.</p></li>
<li><p><strong>sort</strong> (<em>bool</em><em> or </em><em>str</em><em>, </em><em>default False</em>) – Sort the PyRanges before printing. Will print chromosomsomes or strands interleaved on
sort columns.</p></li>
<li><p><strong>formatting</strong> (<em>dict</em><em>, </em><em>default None</em>) – Formatting options per column.</p></li>
<li><p><strong>chain</strong> (<em>False</em>) – Return the PyRanges. Useful to print intermediate results in call chains.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.pc" title="pyranges.pyranges.PyRanges.pc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.pc</span></code></a></dt><dd><p>print chain</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.sp" title="pyranges.pyranges.PyRanges.sp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.sp</span></code></a></dt><dd><p>sort print</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.mp" title="pyranges.pyranges.PyRanges.mp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.mp</span></code></a></dt><dd><p>merge print</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.spc" title="pyranges.pyranges.PyRanges.spc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.spc</span></code></a></dt><dd><p>sort print chain</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.mpc" title="pyranges.pyranges.PyRanges.mpc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.mpc</span></code></a></dt><dd><p>merge print chain</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.msp" title="pyranges.pyranges.PyRanges.msp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.msp</span></code></a></dt><dd><p>merge sort print</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.mspc" title="pyranges.pyranges.PyRanges.mspc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.mspc</span></code></a></dt><dd><p>merge sort print chain</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.rp" title="pyranges.pyranges.PyRanges.rp"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.rp</span></code></a></dt><dd><p>raw print dictionary of DataFrames</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5000</span><span class="p">],</span>
<span class="gp">... </span>     <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7000</span><span class="p">],</span> <span class="s1">&#39;Name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;i1&#39;</span><span class="p">,</span> <span class="s1">&#39;i3&#39;</span><span class="p">,</span> <span class="s1">&#39;i2&#39;</span><span class="p">],</span>
<span class="gp">... </span>     <span class="s1">&#39;Score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">1.1</span><span class="p">,</span> <span class="mf">2.3987</span><span class="p">,</span> <span class="mf">5.9999995</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |       Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (float64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-------------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | i1         |      1.1    | +            |</span>
<span class="go">| chr1         |         8 |         9 | i3         |      2.3987 | +            |</span>
<span class="go">| chr1         |      5000 |      7000 | i2         |      6      | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">print</span><span class="p">(</span><span class="n">formatting</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{:,}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;Score&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="p">})</span>
<span class="go">+--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |       Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (float64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-------------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | i1         |         1.1 | +            |</span>
<span class="go">| chr1         |         8 |         9 | i3         |         2.4 | +            |</span>
<span class="go">| chr1         |     5,000 |      7000 | i2         |         6   | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-------------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">print</span><span class="p">(</span><span class="n">merge_position</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># gr.mp()</span>
<span class="go">+--------------------+------------+-------------+</span>
<span class="go">| - Position -       | Name       |       Score |</span>
<span class="go">| (Multiple types)   | (object)   |   (float64) |</span>
<span class="go">|--------------------+------------+-------------|</span>
<span class="go">| chr1 3-6 +         | i1         |      1.1    |</span>
<span class="go">| chr1 8-9 +         | i3         |      2.3987 |</span>
<span class="go">| chr1 5000-7000 -   | i2         |      6      |</span>
<span class="go">+--------------------+------------+-------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chipseq</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chipseq</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="go">| chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<p>To interleave strands in output, use print with <cite>sort=True</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chipseq</span><span class="o">.</span><span class="n">print</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span> <span class="c1"># chipseq.sp()</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 1325303   | 1325328   | U0         | 0         | -            |</span>
<span class="go">| chr1         | 1541598   | 1541623   | U0         | 0         | +            |</span>
<span class="go">| chr1         | 1599121   | 1599146   | U0         | 0         | +            |</span>
<span class="go">| chr1         | 1820285   | 1820310   | U0         | 0         | -            |</span>
<span class="go">| chr1         | 2448322   | 2448347   | U0         | 0         | -            |</span>
<span class="go">| chr1         | 3046141   | 3046166   | U0         | 0         | -            |</span>
<span class="go">| chr1         | 3437168   | 3437193   | U0         | 0         | -            |</span>
<span class="go">| chr1         | 3504032   | 3504057   | U0         | 0         | +            |</span>
<span class="go">| chr1         | 3637087   | 3637112   | U0         | 0         | -            |</span>
<span class="go">| chr1         | 3681903   | 3681928   | U0         | 0         | -            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 15548022  | 15548047  | U0         | 0         | +            |</span>
<span class="go">| chrY         | 16045242  | 16045267  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 16495497  | 16495522  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 21559181  | 21559206  | U0         | 0         | +            |</span>
<span class="go">| chrY         | 21707662  | 21707687  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 21751211  | 21751236  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 21910706  | 21910731  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 22054002  | 22054027  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 22210637  | 22210662  | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome, Start, End and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">()</span><span class="o">.</span><span class="n">print</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   | Start     | End       |</span>
<span class="go">| (category)   | (int32)   | (int32)   |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         | 0         | 249250621 |</span>
<span class="go">| chr2         | 0         | 243199373 |</span>
<span class="go">| chr3         | 0         | 198022430 |</span>
<span class="go">| chr4         | 0         | 191154276 |</span>
<span class="go">| ...          | ...       | ...       |</span>
<span class="go">| chr22        | 0         | 51304566  |</span>
<span class="go">| chrM         | 0         | 16571     |</span>
<span class="go">| chrX         | 0         | 155270560 |</span>
<span class="go">| chrY         | 0         | 59373566  |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 25 rows and 3 columns from 25 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.rp">
<span class="sig-name descname"><span class="pre">rp</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.rp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.rp" title="Permalink to this definition"></a></dt>
<dd><p>Print dict of DataFrames.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.rpc">
<span class="sig-name descname"><span class="pre">rpc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.rpc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.rpc" title="Permalink to this definition"></a></dt>
<dd><p>Print dict of DataFrames and return self.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.sample">
<span class="sig-name descname"><span class="pre">sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.sample"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.sample" title="Permalink to this definition"></a></dt>
<dd><p>Subsample arbitrary rows of PyRanges.</p>
<p>If n is larger than length of PyRanges, replace must be True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em><em>, </em><em>default 8</em>) – Number of rows to return</p></li>
<li><p><strong>replace</strong> (<em>bool</em><em>, </em><em>False</em>) – Reuse rows.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr2         |  76564764 |  76564789 | U0         |         0 | +            |</span>
<span class="go">| chr3         | 185739979 | 185740004 | U0         |         0 | -            |</span>
<span class="go">| chr20        |  40373657 |  40373682 | U0         |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="mi">10001</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Cannot take a larger sample than population when &#39;replace=False&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.set_intersect">
<span class="sig-name descname"><span class="pre">set_intersect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">how</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.set_intersect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.set_intersect" title="Permalink to this definition"></a></dt>
<dd><p>Return set-theoretical intersection.</p>
<p>Like intersect, but both PyRanges are merged first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to set-intersect.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>how</strong> (<em>{None</em><em>, </em><em>&quot;first&quot;</em><em>, </em><em>&quot;last&quot;</em><em>, </em><em>&quot;containment&quot;}</em><em>, </em><em>default None</em><em>, </em><em>i.e. all</em>) – What intervals to report. By default reports all overlapping intervals. “containment”
reports intervals where the overlapping is contained within it.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PyRanges with overlapping subintervals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.intersect" title="pyranges.pyranges.PyRanges.intersect"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.intersect</span></code></a></dt><dd><p>find overlapping subintervals</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.overlap" title="pyranges.pyranges.PyRanges.overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.overlap</span></code></a></dt><dd><p>report overlapping intervals</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | ID         |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|         chr1 |         1 |         3 | a          |</span>
<span class="go">|         chr1 |         4 |         9 | b          |</span>
<span class="go">|         chr1 |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|         chr1 |         2 |         3 |</span>
<span class="go">|         chr1 |         2 |         9 |</span>
<span class="go">|         chr1 |         9 |        10 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">set_intersect</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         2 |         3 |</span>
<span class="go">| chr1         |         4 |         9 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<p>In this simple unstranded case, this is the same as the below:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">merge</span><span class="p">()</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">gr2</span><span class="o">.</span><span class="n">merge</span><span class="p">())</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         2 |         3 |</span>
<span class="go">| chr1         |         4 |         9 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">set_intersect</span><span class="p">(</span><span class="n">gr2</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;containment&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         4 |         9 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.set_union">
<span class="sig-name descname"><span class="pre">set_union</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.set_union"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.set_union" title="Permalink to this definition"></a></dt>
<dd><p>Return set-theoretical union.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a>) – PyRanges to do union with.</p></li>
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A PyRanges with the union of intervals.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.set_intersect" title="pyranges.pyranges.PyRanges.set_intersect"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.set_intersect</span></code></a></dt><dd><p>set-theoretical intersection</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.overlap" title="pyranges.pyranges.PyRanges.overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.overlap</span></code></a></dt><dd><p>report overlapping intervals</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | ID         |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">|         chr1 |         1 |         3 | a          |</span>
<span class="go">|         chr1 |         4 |         9 | b          |</span>
<span class="go">|         chr1 |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|         chr1 |         2 |         3 |</span>
<span class="go">|         chr1 |         2 |         9 |</span>
<span class="go">|         chr1 |         9 |        10 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">set_union</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         1 |        11 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.sort" title="Permalink to this definition"></a></dt>
<dd><p>Sort by position or columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>default None</em>) – Column(s) to sort by. Default is Start and End.
Special value “5” can be provided to sort by 5’: intervals on + strand are sorted in ascending order, while
those on - strand are sorted in descending order.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since a PyRanges contains multiple DataFrames, the sorting only happens within dataframes.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sorted PyRanges</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../multioverlap/index.html#module-pyranges.multioverlap" title="pyranges.multioverlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.multioverlap</span></code></a></dt><dd><p>find overlaps with multiple PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">between</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand     |</span>
<span class="go">| (object)     |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         3 |         6 | +          |</span>
<span class="go">| chr1         |         6 |         8 | +          |</span>
<span class="go">| chr1         |         8 |         9 | +          |</span>
<span class="go">| chr1         |         5 |         7 | -          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (object)     |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         3 |         5 |</span>
<span class="go">| chr1         |         5 |         6 |</span>
<span class="go">| chr1         |         6 |         7 |</span>
<span class="go">| chr1         |         8 |         9 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 4 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">between</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (object)     |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         3 |         5 |</span>
<span class="go">| chr1         |         5 |         6 |</span>
<span class="go">| chr1         |         6 |         7 |</span>
<span class="go">| chr1         |         7 |         8 |</span>
<span class="go">| chr1         |         8 |         9 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 5 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.sp">
<span class="sig-name descname"><span class="pre">sp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.sp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.sp" title="Permalink to this definition"></a></dt>
<dd><p>Sort on location and print.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.spc">
<span class="sig-name descname"><span class="pre">spc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formatting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.spc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.spc" title="Permalink to this definition"></a></dt>
<dd><p>Sort on location, print and return self.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.print" title="pyranges.pyranges.PyRanges.print"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.print</span></code></a></dt><dd><p>print PyRanges.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.slack">
<span class="sig-name descname"><span class="pre">slack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">slack</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.slack"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.slack" title="Permalink to this definition"></a></dt>
<dd><p>Deprecated: this function has been moved to Pyranges.extend</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.spliced_subsequence">
<span class="sig-name descname"><span class="pre">spliced_subsequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.spliced_subsequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.spliced_subsequence" title="Permalink to this definition"></a></dt>
<dd><p>Get subsequences of the intervals, using coordinates mapping to spliced transcripts (without introns)</p>
<p>The returned intervals are subregions of self, cut according to specifications.
Start and end are relative to the 5’ end: 0 means the leftmost nucleotide for + strand
intervals, while it means the rightmost one for - strand.
This method also allows to manipulate groups of intervals (e.g. exons belonging to same transcripts)
through the ‘by’ argument. When using it, start and end refer to the spliced transcript coordinates,
meaning that introns are ignored in the count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em>) – Start of subregion, 0-based and included, counting from the 5’ end.
Use a negative int to count from the 3’  (e.g. -1 is the last nucleotide)</p></li>
<li><p><strong>end</strong> (<em>int</em><em>, </em><em>default None</em>) – End of subregion, 0-based and excluded, counting from the 5’ end.
Use a negative int to count from the 3’  (e.g. -1 is the last nucleotide)
If None, the existing 3’ end is returned.</p></li>
<li><p><strong>by</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>default None</em>) – intervals are grouped by this/these ID column(s) beforehand, e.g. exons belonging to same transcripts</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>strand<span class="classifier">bool, default None, i.e. auto</span></dt><dd><p>Whether strand is considered when interpreting the start and end arguments of this function.
If True, counting is from the 5’ end, which is the leftmost coordinate for + strand and the rightmost for - strand.
If False, all intervals are processed like they reside on the + strand.
If None (default), strand is considered if the PyRanges is stranded.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Subregion of self, subsequenced as specified by arguments</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the request goes out of bounds (e.g. requesting 100 nts for a 90nt region), only the existing portion is returned</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.subsequence" title="pyranges.pyranges.PyRanges.subsequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subsequence</span></code></a></dt><dd><p>analogous to this method, but input coordinates refer to the unspliced transcript</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>  <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">70</span><span class="p">,</span> <span class="mi">140</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">11</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">80</span><span class="p">,</span> <span class="mi">152</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;transcript_id&quot;</span><span class="p">:[</span><span class="s2">&quot;t1&quot;</span><span class="p">,</span> <span class="s2">&quot;t1&quot;</span><span class="p">,</span> <span class="s2">&quot;t2&quot;</span><span class="p">,</span> <span class="s2">&quot;t2&quot;</span><span class="p">,</span> <span class="s2">&quot;t3&quot;</span><span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |         1 |        11 | t1              |</span>
<span class="go">|            1 | +            |        40 |        60 | t1              |</span>
<span class="go">|            2 | -            |        10 |        25 | t2              |</span>
<span class="go">|            2 | -            |        70 |        80 | t2              |</span>
<span class="go">|            3 | +            |       140 |       152 | t3              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<blockquote>
<div><p># Get the first 15 nucleotides of <em>each spliced transcript</em>, grouping exons by transcript_id:</p>
</div></blockquote>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">spliced_subsequence</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int64) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |         1 |        11 | t1              |</span>
<span class="go">|            1 | +            |        40 |        45 | t1              |</span>
<span class="go">|            2 | -            |        70 |        80 | t2              |</span>
<span class="go">|            2 | -            |        20 |        25 | t2              |</span>
<span class="go">|            3 | +            |       140 |       152 | t3              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<blockquote>
<div><p># Get the last 20 nucleotides of each spliced transcript:</p>
</div></blockquote>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">spliced_subsequence</span><span class="p">(</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int64) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |        40 |        60 | t1              |</span>
<span class="go">|            2 | -            |        70 |        75 | t2              |</span>
<span class="go">|            2 | -            |        10 |        25 | t2              |</span>
<span class="go">|            3 | +            |       140 |       152 | t3              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 5 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<blockquote>
<div><p># Get region from 25 to 60 of each spliced transcript, or their existing subportion:</p>
</div></blockquote>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">spliced_subsequence</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |        55 |        60 | t1              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 1 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<blockquote>
<div><p># Get region of each spliced transcript which excludes their first and last 3 nucleotides:</p>
</div></blockquote>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">spliced_subsequence</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int64) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |         4 |        11 | t1              |</span>
<span class="go">|            1 | +            |        40 |        57 | t1              |</span>
<span class="go">|            2 | -            |        70 |        77 | t2              |</span>
<span class="go">|            2 | -            |        13 |        25 | t2              |</span>
<span class="go">|            3 | +            |       143 |       149 | t3              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.split">
<span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">between</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.split"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.split" title="Permalink to this definition"></a></dt>
<dd><p>Split into non-overlapping intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to ignore strand information if PyRanges is stranded.</p></li>
<li><p><strong>between</strong> (<em>bool</em><em>, </em><em>default False</em>) – Include lengths between intervals.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with intervals split at overlap points.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../multioverlap/index.html#module-pyranges.multioverlap" title="pyranges.multioverlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.multioverlap</span></code></a></dt><dd><p>find overlaps with multiple PyRanges</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">12</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | +            |</span>
<span class="go">| chr1         |         5 |         9 | +            |</span>
<span class="go">| chr1         |         5 |         7 | -            |</span>
<span class="go">| chr1         |        11 |        12 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand     |</span>
<span class="go">| (object)     |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         3 |         5 | +          |</span>
<span class="go">| chr1         |         5 |         6 | +          |</span>
<span class="go">| chr1         |         6 |         9 | +          |</span>
<span class="go">| chr1         |         5 |         7 | -          |</span>
<span class="go">| chr1         |        11 |        12 | -          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">between</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand     |</span>
<span class="go">| (object)     |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         3 |         5 | +          |</span>
<span class="go">| chr1         |         5 |         6 | +          |</span>
<span class="go">| chr1         |         6 |         9 | +          |</span>
<span class="go">| chr1         |         5 |         7 | -          |</span>
<span class="go">| chr1         |         7 |        11 | -          |</span>
<span class="go">| chr1         |        11 |        12 | -          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Stranded PyRanges object has 6 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (object)     |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         3 |         5 |</span>
<span class="go">| chr1         |         5 |         6 |</span>
<span class="go">| chr1         |         6 |         7 |</span>
<span class="go">| chr1         |         7 |         9 |</span>
<span class="go">| chr1         |        11 |        12 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 5 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">between</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (object)     |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         3 |         5 |</span>
<span class="go">| chr1         |         5 |         6 |</span>
<span class="go">| chr1         |         6 |         7 |</span>
<span class="go">| chr1         |         7 |         9 |</span>
<span class="go">| chr1         |         9 |        11 |</span>
<span class="go">| chr1         |        11 |        12 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 6 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.subset">
<span class="sig-name descname"><span class="pre">subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.subset" title="Permalink to this definition"></a></dt>
<dd><p>Return a subset of the rows.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>f</strong> (<em>function</em>) – Function which returns boolean Series equal to length of df.</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use
chromosome/strand pairs if the PyRanges is stranded.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>PyRanges can also be subsetted directly with a boolean Series. This function is slightly
faster, but more cumbersome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>PyRanges subset on rows.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">f1</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         3 |         6 | interval1  |         0 | +            |</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">Start</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<p>Also possible:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="p">[</span><span class="n">gr</span><span class="o">.</span><span class="n">Start</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">]</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         |         8 |         9 | interval3  |         0 | +            |</span>
<span class="go">| chr1         |         5 |         7 | interval2  |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.subsequence">
<span class="sig-name descname"><span class="pre">subsequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.subsequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.subsequence" title="Permalink to this definition"></a></dt>
<dd><p>Get subsequences of the intervals.</p>
<p>The returned intervals are subregions of self, cut according to specifications.
Start and end are relative to the 5’ end: 0 means the leftmost nucleotide for + strand
intervals, while it means the rightmost one for - strand.
This method also allows to manipulate groups of intervals (e.g. exons belonging to same transcripts)
through the ‘by’ argument. When using it, start and end refer to the unspliced transcript coordinates,
meaning that introns are included in the count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em>) – Start of subregion, 0-based and included, counting from the 5’ end.
Use a negative int to count from the 3’  (e.g. -1 is the last nucleotide)</p></li>
<li><p><strong>end</strong> (<em>int</em><em>, </em><em>default None</em>) – <p>End of subregion, 0-based and excluded, counting from the 5’ end.
Use a negative int to count from the 3’  (e.g. -1 is the last nucleotide)</p>
<p>If None, the existing 3’ end is returned.</p>
</p></li>
<li><p><strong>by</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>default None</em>) – intervals are grouped by this/these ID column(s) beforehand, e.g. exons belonging to same transcripts</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether strand is considered when interpreting the start and end arguments of this function.
If True, counting is from the 5’ end, which is the leftmost coordinate for + strand and the rightmost for - strand.
If False, all intervals are processed like they reside on the + strand.
If None (default), strand is considered if the PyRanges is stranded.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Subregion of self, subsequenced as specified by arguments</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If the request goes out of bounds (e.g. requesting 100 nts for a 90nt region), only the existing portion is returned</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.spliced_subsequence" title="pyranges.pyranges.PyRanges.spliced_subsequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">spliced_subsequence</span></code></a></dt><dd><p>analogous to this method, but intronic regions are not counted, so that input coordinates refer to the spliced transcript</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>  <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;Strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">140</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">20</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="mi">155</span><span class="p">],</span>
<span class="gp">... </span>                  <span class="s2">&quot;transcript_id&quot;</span><span class="p">:[</span><span class="s2">&quot;t1&quot;</span><span class="p">,</span> <span class="s2">&quot;t1&quot;</span><span class="p">,</span> <span class="s2">&quot;t2&quot;</span><span class="p">,</span> <span class="s2">&quot;t2&quot;</span><span class="p">,</span> <span class="s2">&quot;t3&quot;</span><span class="p">]</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">|   Chromosome | Strand       |     Start |       End | transcript_id   |</span>
<span class="go">|   (category) | (category)   |   (int32) |   (int32) | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+-----------------|</span>
<span class="go">|            1 | +            |         1 |        20 | t1              |</span>
<span class="go">|            1 | +            |        40 |        60 | t1              |</span>
<span class="go">|            2 | -            |         2 |        13 | t2              |</span>
<span class="go">|            2 | -            |        30 |        45 | t2              |</span>
<span class="go">|            3 | +            |       140 |       155 | t3              |</span>
<span class="go">+--------------+--------------+-----------+-----------+-----------------+</span>
<span class="go">Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<p># Get the first 10 nucleotides (at the 5’) of <em>each interval</em> (each line of the dataframe):
&gt;&gt;&gt; p.subsequence(0, 10)
+————–+————–+———–+———–+—————–+
|   Chromosome | Strand       |     Start |       End | transcript_id   |
|   (category) | (category)   |   (int32) |   (int32) | (object)        |
<a href="#id1"><span class="problematic" id="id2">|--------------+--------------+-----------+-----------+-----------------|</span></a>
|            1 | +            |         1 |        11 | t1              |
|            1 | +            |        40 |        50 | t1              |
|            2 | -            |         3 |        13 | t2              |
|            2 | -            |        35 |        45 | t2              |
|            3 | +            |       140 |       150 | t3              |
+————–+————–+———–+———–+—————–+
Stranded PyRanges object has 5 rows and 5 columns from 3 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.</p>
<p># Get the first 10 nucleotides of <em>each transcript</em>, grouping exons by transcript_id:
&gt;&gt;&gt; p.subsequence(0, 10, by=’transcript_id’)
+————–+————–+———–+———–+—————–+
|   Chromosome | Strand       |     Start |       End | transcript_id   |
|   (category) | (category)   |   (int32) |   (int32) | (object)        |
<a href="#id3"><span class="problematic" id="id4">|--------------+--------------+-----------+-----------+-----------------|</span></a>
|            1 | +            |         1 |        11 | t1              |
|            2 | -            |        35 |        45 | t2              |
|            3 | +            |       140 |       150 | t3              |
+————–+————–+———–+———–+—————–+
Stranded PyRanges object has 3 rows and 5 columns from 3 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.</p>
<p># Get the last 20 nucleotides of each transcript:
&gt;&gt;&gt; p.subsequence(-20, by=’transcript_id’)
+————–+————–+———–+———–+—————–+
|   Chromosome | Strand       |     Start |       End | transcript_id   |
|   (category) | (category)   |   (int32) |   (int32) | (object)        |
<a href="#id5"><span class="problematic" id="id6">|--------------+--------------+-----------+-----------+-----------------|</span></a>
|            1 | +            |        40 |        60 | t1              |
|            2 | -            |         2 |        13 | t2              |
|            3 | +            |       140 |       155 | t3              |
+————–+————–+———–+———–+—————–+
Stranded PyRanges object has 3 rows and 5 columns from 3 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.</p>
<p># Get region from 30 to 330 of each transcript, or their existing subportion:
&gt;&gt;&gt; p.subsequence(30, 300, by=’transcript_id’)
+————–+————–+———–+———–+—————–+
|   Chromosome | Strand       |     Start |       End | transcript_id   |
|   (category) | (category)   |   (int32) |   (int32) | (object)        |
<a href="#id7"><span class="problematic" id="id8">|--------------+--------------+-----------+-----------+-----------------|</span></a>
|            1 | +            |        40 |        60 | t1              |
|            2 | -            |         2 |        13 | t2              |
+————–+————–+———–+———–+—————–+
Stranded PyRanges object has 2 rows and 5 columns from 2 chromosomes.
For printing, the PyRanges was sorted on Chromosome and Strand.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.subtract">
<span class="sig-name descname"><span class="pre">subtract</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strandedness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.subtract"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.subtract" title="Permalink to this definition"></a></dt>
<dd><p>Subtract intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strandedness</strong> (<em>{None</em><em>, </em><em>&quot;same&quot;</em><em>, </em><em>&quot;opposite&quot;</em><em>, </em><em>False}</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to compare PyRanges on the same strand, the opposite or ignore strand
information. The default, None, means use “same” if both PyRanges are strande,
otherwise ignore the strand information.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.overlap" title="pyranges.PyRanges.overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.overlap</span></code></a></dt><dd><p>use with invert=True to return all intervals without overlap</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span>
<span class="gp">... </span>                   <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         1 |         3 | a          |</span>
<span class="go">| chr1         |         4 |         9 | b          |</span>
<span class="go">| chr1         |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr2</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         2 |         3 |</span>
<span class="go">| chr1         |         2 |         9 |</span>
<span class="go">| chr1         |         9 |        10 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">subtract</span><span class="p">(</span><span class="n">gr2</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">| Chromosome   |     Start |       End | ID         |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------|</span>
<span class="go">| chr1         |         1 |         2 | a          |</span>
<span class="go">| chr1         |        10 |        11 | c          |</span>
<span class="go">+--------------+-----------+-----------+------------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">to_stdout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.summary" title="Permalink to this definition"></a></dt>
<dd><p>Return info.</p>
<p>Count refers to the number of intervals, the rest to the lengths.</p>
<p>The column “pyrange” describes the data as is. “coverage_forward” and “coverage_reverse”
describe the data after strand-specific merging of overlapping intervals.
“coverage_unstranded” describes the data after merging, without considering the strands.</p>
<p>The row “count” is the number of intervals and “sum” is their total length. The rest describe the lengths of the
intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>to_stdout</strong> (<em>bool</em><em>, </em><em>default True</em>) – Print summary.</p></li>
<li><p><strong>return_df</strong> (<em>bool</em><em>, </em><em>default False</em>) – Return df with summary.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None or DataFrame with summary.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()[[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_id         |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)        |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-----------------|</span>
<span class="go">| 1            | gene         | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | transcript   | 11868     | 14409     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | exon         | 11868     | 12227     | +            | ENSG00000223972 |</span>
<span class="go">| 1            | exon         | 12612     | 12721     | +            | ENSG00000223972 |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...             |</span>
<span class="go">| 1            | gene         | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">| 1            | transcript   | 1173055   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">| 1            | exon         | 1179364   | 1179555   | -            | ENSG00000205231 |</span>
<span class="go">| 1            | exon         | 1173055   | 1176396   | -            | ENSG00000205231 |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-----------------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="go">+-------+------------------+--------------------+--------------------+-----------------------+</span>
<span class="go">|       |          pyrange |   coverage_forward |   coverage_reverse |   coverage_unstranded |</span>
<span class="go">|-------+------------------+--------------------+--------------------+-----------------------|</span>
<span class="go">| count |   2446           |               39   |               23   |                  32   |</span>
<span class="go">| mean  |   2291.92        |             7058.1 |            30078.6 |               27704.2 |</span>
<span class="go">| std   |  11906.9         |            10322.3 |            59467.7 |               67026.9 |</span>
<span class="go">| min   |      1           |               83   |              154   |                  83   |</span>
<span class="go">| 25%   |     90           |             1051   |             1204   |                1155   |</span>
<span class="go">| 50%   |    138           |             2541   |             6500   |                6343   |</span>
<span class="go">| 75%   |    382.25        |             7168   |            23778   |               20650.8 |</span>
<span class="go">| max   | 241726           |            43065   |           241726   |              291164   |</span>
<span class="go">| sum   |      5.60603e+06 |           275266   |           691807   |              886534   |</span>
<span class="go">+-------+------------------+--------------------+--------------------+-----------------------+</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">summary</span><span class="p">(</span><span class="n">return_df</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">to_stdout</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">            pyrange  coverage_forward  coverage_reverse  coverage_unstranded</span>
<span class="go">count  2.446000e+03         39.000000         23.000000            32.000000</span>
<span class="go">mean   2.291918e+03       7058.102564      30078.565217         27704.187500</span>
<span class="go">std    1.190685e+04      10322.309347      59467.695265         67026.868647</span>
<span class="go">min    1.000000e+00         83.000000        154.000000            83.000000</span>
<span class="go">25%    9.000000e+01       1051.000000       1204.000000          1155.000000</span>
<span class="go">50%    1.380000e+02       2541.000000       6500.000000          6343.000000</span>
<span class="go">75%    3.822500e+02       7168.000000      23778.000000         20650.750000</span>
<span class="go">max    2.417260e+05      43065.000000     241726.000000        291164.000000</span>
<span class="go">sum    5.606031e+06     275266.000000     691807.000000        886534.000000</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.tail">
<span class="sig-name descname"><span class="pre">tail</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.tail"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.tail" title="Permalink to this definition"></a></dt>
<dd><p>Return the n last rows.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 8</em>) – Return n rows.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PyRanges with the n last rows.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.head" title="pyranges.pyranges.PyRanges.head"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.head</span></code></a></dt><dd><p>return the first rows</p>
</dd>
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.sample" title="pyranges.pyranges.PyRanges.sample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.sample</span></code></a></dt><dd><p>return random rows</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="go">| chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chrY         |  13517892 |  13517917 | U0         |         0 | -            |</span>
<span class="go">| chrY         |   8010951 |   8010976 | U0         |         0 | -            |</span>
<span class="go">| chrY         |   7405376 |   7405401 | U0         |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.tile">
<span class="sig-name descname"><span class="pre">tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tile_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.tile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.tile" title="Permalink to this definition"></a></dt>
<dd><p>Return overlapping genomic tiles.</p>
<p>The genome is divided into bookended tiles of length <cite>tile_size</cite> and one is returned per
overlapping interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tile_size</strong> (<em>int</em>) – Length of the tiles.</p></li>
<li><p><strong>overlap</strong> (<em>bool</em><em>, </em><em>default False</em>) – Add column of nucleotide overlap to each tile.</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use
chromosome/strand pairs if the PyRanges is stranded.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tiled PyRanges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.window" title="pyranges.PyRanges.window"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.window</span></code></a></dt><dd><p>divide intervals into windows</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()[[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="go">| 1            | gene         | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | transcript   | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | exon         | 11868     | 12227     | +            | DDX11L1     |</span>
<span class="go">| 1            | exon         | 12612     | 12721     | +            | DDX11L1     |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="go">| 1            | gene         | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | transcript   | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1179364   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1173055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="mi">200</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="go">| 1            | gene         | 11800     | 12000     | +            | DDX11L1     |</span>
<span class="go">| 1            | gene         | 12000     | 12200     | +            | DDX11L1     |</span>
<span class="go">| 1            | gene         | 12200     | 12400     | +            | DDX11L1     |</span>
<span class="go">| 1            | gene         | 12400     | 12600     | +            | DDX11L1     |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="go">| 1            | exon         | 1175600   | 1175800   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1175800   | 1176000   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1176000   | 1176200   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1176200   | 1176400   | -            | TTLL10-AS1  |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">Stranded PyRanges object has 30,538 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+---------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_name   | TileOverlap   |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    | (int32)       |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-------------+---------------|</span>
<span class="go">| 1            | gene         | 11800     | 11900     | +            | DDX11L1     | 32            |</span>
<span class="go">| 1            | gene         | 11900     | 12000     | +            | DDX11L1     | 100           |</span>
<span class="go">| 1            | gene         | 12000     | 12100     | +            | DDX11L1     | 100           |</span>
<span class="go">| 1            | gene         | 12100     | 12200     | +            | DDX11L1     | 100           |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...         | ...           |</span>
<span class="go">| 1            | exon         | 1176000   | 1176100   | -            | TTLL10-AS1  | 100           |</span>
<span class="go">| 1            | exon         | 1176100   | 1176200   | -            | TTLL10-AS1  | 100           |</span>
<span class="go">| 1            | exon         | 1176200   | 1176300   | -            | TTLL10-AS1  | 100           |</span>
<span class="go">| 1            | exon         | 1176300   | 1176400   | -            | TTLL10-AS1  | 96            |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+---------------+</span>
<span class="go">Stranded PyRanges object has 58,516 rows and 7 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_example">
<span class="sig-name descname"><span class="pre">to_example</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_example"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_example" title="Permalink to this definition"></a></dt>
<dd><p>Return as dict.</p>
<p>Used for easily creating examples for copy and pasting.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 10</em>) – Number of rows. Half is taken from the start, the other half from the end.</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.from_dict</span></code></dt><dd><p>create PyRanges from dict</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">chipseq</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   | Start     | End       | Name       | Score     | Strand       |</span>
<span class="go">| (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 216711011 | 216711036 | U0         | 0         | +            |</span>
<span class="go">| chr1         | 144227079 | 144227104 | U0         | 0         | +            |</span>
<span class="go">| ...          | ...       | ...       | ...        | ...       | ...          |</span>
<span class="go">| chrY         | 15224235  | 15224260  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 13517892  | 13517917  | U0         | 0         | -            |</span>
<span class="go">| chrY         | 8010951   | 8010976   | U0         | 0         | -            |</span>
<span class="go">| chrY         | 7405376   | 7405401   | U0         | 0         | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">gr</span><span class="o">.</span><span class="n">to_example</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">{&#39;Chromosome&#39;: [&#39;chr1&#39;, &#39;chr1&#39;, &#39;chrY&#39;, &#39;chrY&#39;], &#39;Start&#39;: [212609534, 169887529, 8010951, 7405376], &#39;End&#39;: [212609559, 169887554, 8010976, 7405401], &#39;Name&#39;: [&#39;U0&#39;, &#39;U0&#39;, &#39;U0&#39;, &#39;U0&#39;], &#39;Score&#39;: [0, 0, 0, 0], &#39;Strand&#39;: [&#39;+&#39;, &#39;+&#39;, &#39;-&#39;, &#39;-&#39;]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Name       |     Score | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+--------------|</span>
<span class="go">| chr1         | 212609534 | 212609559 | U0         |         0 | +            |</span>
<span class="go">| chr1         | 169887529 | 169887554 | U0         |         0 | +            |</span>
<span class="go">| chrY         |   8010951 |   8010976 | U0         |         0 | -            |</span>
<span class="go">| chrY         |   7405376 |   7405401 | U0         |         0 | -            |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 4 rows and 6 columns from 2 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.three_end">
<span class="sig-name descname"><span class="pre">three_end</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.three_end"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.three_end" title="Permalink to this definition"></a></dt>
<dd><p>Return the 3’-end.</p>
<p>The 3’-end is the start of intervals on the reverse strand and the end of intervals on the
forward strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>PyRanges with the 3’.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.five_end" title="pyranges.pyranges.PyRanges.five_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.five_end</span></code></a></dt><dd><p>return the five prime end</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span>  <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         5 | +            |</span>
<span class="go">| chr1         |         6 |         8 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">three_end</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         5 |         6 | +            |</span>
<span class="go">| chr1         |         6 |         7 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'infer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Write to bed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em><em>, </em><em>default None</em>) – Where to write. If None, returns string representation.</p></li>
<li><p><strong>keep</strong> (<em>bool</em><em>, </em><em>default True</em>) – Whether to keep all columns, not just Chromosome, Start, End,
Name, Score, Strand when writing.</p></li>
<li><p><strong>compression</strong> (<em>str</em><em>, </em><em>compression type to use</em><em>, </em><em>by default infer based on extension.</em>) – See pandas.DataFree.to_csv for more info.</p></li>
<li><p><strong>chain</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the PyRanges after writing.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span>  <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">],</span> <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |      Gene |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------|</span>
<span class="go">| chr1         |         1 |         5 | +            |         1 |</span>
<span class="go">| chr1         |         6 |         8 | -            |         2 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_bed</span><span class="p">())</span>
<span class="go">chr1    1       5       .       .       +       1</span>
<span class="go">chr1    6       8       .       .       -       2</span>
</pre></div>
</div>
<p>Does not include noncanonical bed-column <cite>Gene</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_bed</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="go">chr1    1       5       .       .       +</span>
<span class="go">chr1    6       8       .       .       -</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_bed</span><span class="p">(</span><span class="s2">&quot;test.bed&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |      Gene |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------|</span>
<span class="go">| chr1         |         1 |         5 | +            |         1 |</span>
<span class="go">| chr1         |         6 |         8 | -            |         2 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;test.bed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="go">[&#39;chr1\t1\t5\t.\t.\t+\t1\n&#39;, &#39;chr1\t6\t8\t.\t.\t-\t2\n&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_sizes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rpm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">divide</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dryrun</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_bigwig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Write regular or value coverage to bigwig.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To create one bigwig per strand, subset the PyRanges first.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – Where to write bigwig.</p></li>
<li><p><strong>chromosome_sizes</strong> (<a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges"><em>PyRanges</em></a><em> or </em><em>dict</em>) – If dict: map of chromosome names to chromosome length.</p></li>
<li><p><strong>rpm</strong> (<em>True</em>) – Whether to normalize data by dividing by total number of intervals and multiplying by
1e6.</p></li>
<li><p><strong>divide</strong> (<em>bool</em><em>, </em><em>default False</em>) – (Only useful with value_col) Divide value coverage by regular coverage and take log2.</p></li>
<li><p><strong>value_col</strong> (<em>str</em><em>, </em><em>default None</em>) – Name of column to compute coverage of.</p></li>
<li><p><strong>dryrun</strong> (<em>bool</em><em>, </em><em>default False</em>) – Return data that would be written without writing bigwigs.</p></li>
<li><p><strong>chain</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the PyRanges after writing.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Requires pybigwig to be installed.</p>
<p>If you require more control over the normalization process, use pyranges.to_bigwig()</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.to_bigwig" title="pyranges.to_bigwig"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.to_bigwig</span></code></a></dt><dd><p>write pandas DataFrame to bigwig.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span>  <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;Value&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |     Value |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |   (int64) |</span>
<span class="go">|--------------+-----------+-----------+--------------+-----------|</span>
<span class="go">| chr1         |         1 |         7 | +            |        10 |</span>
<span class="go">| chr1         |         4 |         8 | -            |        20 |</span>
<span class="go">| chr1         |         6 |        10 | -            |        30 |</span>
<span class="go">+--------------+-----------+-----------+--------------+-----------+</span>
<span class="go">Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_bigwig</span><span class="p">(</span><span class="n">dryrun</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">rpm</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-------------+</span>
<span class="go">| Chromosome   |     Start |       End |       Score |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (float64) |</span>
<span class="go">|--------------+-----------+-----------+-------------|</span>
<span class="go">| chr1         |         1 |         4 |           1 |</span>
<span class="go">| chr1         |         4 |         6 |           2 |</span>
<span class="go">| chr1         |         6 |         7 |           3 |</span>
<span class="go">| chr1         |         7 |         8 |           2 |</span>
<span class="go">| chr1         |         8 |        10 |           1 |</span>
<span class="go">+--------------+-----------+-----------+-------------+</span>
<span class="go">Unstranded PyRanges object has 5 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_bigwig</span><span class="p">(</span><span class="n">dryrun</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">rpm</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">value_col</span><span class="o">=</span><span class="s2">&quot;Value&quot;</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-------------+</span>
<span class="go">| Chromosome   |     Start |       End |       Score |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (float64) |</span>
<span class="go">|--------------+-----------+-----------+-------------|</span>
<span class="go">| chr1         |         1 |         4 |          10 |</span>
<span class="go">| chr1         |         4 |         6 |          30 |</span>
<span class="go">| chr1         |         6 |         7 |          60 |</span>
<span class="go">| chr1         |         7 |         8 |          50 |</span>
<span class="go">| chr1         |         8 |        10 |          30 |</span>
<span class="go">+--------------+-----------+-----------+-------------+</span>
<span class="go">Unstranded PyRanges object has 5 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_bigwig</span><span class="p">(</span><span class="n">dryrun</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">rpm</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">value_col</span><span class="o">=</span><span class="s2">&quot;Value&quot;</span><span class="p">,</span> <span class="n">divide</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+-------------+</span>
<span class="go">| Chromosome   |     Start |       End |       Score |</span>
<span class="go">| (category)   |   (int32) |   (int32) |   (float64) |</span>
<span class="go">|--------------+-----------+-----------+-------------|</span>
<span class="go">| chr1         |         0 |         1 |   nan       |</span>
<span class="go">| chr1         |         1 |         4 |     3.32193 |</span>
<span class="go">| chr1         |         4 |         6 |     3.90689 |</span>
<span class="go">| chr1         |         6 |         7 |     4.32193 |</span>
<span class="go">| chr1         |         7 |         8 |     4.64386 |</span>
<span class="go">| chr1         |         8 |        10 |     4.90689 |</span>
<span class="go">+--------------+-----------+-----------+-------------+</span>
<span class="go">Unstranded PyRanges object has 6 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_csv">
<span class="sig-name descname"><span class="pre">to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">','</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'infer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_csv" title="Permalink to this definition"></a></dt>
<dd><p>Write to comma- or other value-separated file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em><em>, </em><em>default None</em><em>, </em><em>i.e. return string representation.</em>) – Where to write file.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>default &quot;</em><em>,</em><em>&quot;</em>) – String of length 1. Field delimiter for the output file.</p></li>
<li><p><strong>header</strong> (<em>bool</em><em>, </em><em>default True</em>) – Write out the column names.</p></li>
<li><p><strong>compression</strong> (<em>{‘infer’</em><em>, </em><em>‘gzip’</em><em>, </em><em>‘bz2’</em><em>, </em><em>‘zip’</em><em>, </em><em>‘xz’</em><em>, </em><em>None}</em><em>, </em><em>default &quot;infer&quot;</em>) – Which compression to use. Uses file extension to infer by default.</p></li>
<li><p><strong>chain</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the PyRanges after writing.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The output encodes intervals just like PyRanges: 0-based, Start included and End excluded.</p>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;Feature&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">Chromosome      Start   End     Feature</span>
<span class="go">1       1       4       gene</span>
<span class="go">1       3       6       exon</span>
<span class="go">1       5       9       exon</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_gff3">
<span class="sig-name descname"><span class="pre">to_gff3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'infer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_gff3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_gff3" title="Permalink to this definition"></a></dt>
<dd><p>Write to General Feature Format 3.</p>
<p>The GFF format consists of a tab-separated file without header.
GFF contains a fixed amount of columns, indicated below (names before “:”).
For each of these, PyRanges will use the corresponding column (names after “:”).</p>
<p><code class="docutils literal notranslate"><span class="pre">seqname:</span> <span class="pre">Chromosome</span>
<span class="pre">source:</span> <span class="pre">Source</span>
<span class="pre">type:</span> <span class="pre">Feature</span>
<span class="pre">start:</span> <span class="pre">Start</span>
<span class="pre">end:</span> <span class="pre">End</span>
<span class="pre">score:</span> <span class="pre">Score</span>
<span class="pre">strand:</span> <span class="pre">Strand</span>
<span class="pre">phase:</span> <span class="pre">Frame</span>
<span class="pre">attribute:</span> <span class="pre">auto-filled</span></code></p>
<p>Columns which are not mapped to GFF columns are appended as a field
in the attribute string (i.e. the last field).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em><em>, </em><em>default None</em><em>, </em><em>i.e. return string representation.</em>) – Where to write file.</p></li>
<li><p><strong>compression</strong> (<em>{‘infer’</em><em>, </em><em>‘gzip’</em><em>, </em><em>‘bz2’</em><em>, </em><em>‘zip’</em><em>, </em><em>‘xz’</em><em>, </em><em>None}</em><em>, </em><em>default &quot;infer&quot;</em>) – Which compression to use. Uses file extension to infer by default.</p></li>
<li><p><strong>chain</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the PyRanges after writing.</p></li>
<li><p><strong>map_cols</strong> (<em>dict</em><em>, </em><em>default None</em>) – Override mapping between GFF and PyRanges fields for any number of columns.
Format: <code class="docutils literal notranslate"><span class="pre">{gff_column</span> <span class="pre">:</span> <span class="pre">pyranges_column}</span></code>
If a mapping is found for the “attribute”` column, this not auto-filled</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Nonexisting columns will be added with a ‘.’ to represent the missing values.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_gff3" title="pyranges.read_gff3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_gff3</span></code></a></dt><dd><p>read GFF3 files</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.to_gtf</span></code></dt><dd><p>write to GTF format</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;Feature&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gff3</span><span class="p">())</span>
<span class="go">1       .       gene    2       4       .       .       .</span>
<span class="go">1       .       exon    4       6       .       .       .</span>
<span class="go">1       .       exon    6       9       .       .       .</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">Gene</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">function</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;a b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;def&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gff3</span><span class="p">())</span>
<span class="go">1       .       gene    2       4       .       .       .       Gene=1;function=a b</span>
<span class="go">1       .       exon    4       6       .       .       .       Gene=2;function=c</span>
<span class="go">1       .       exon    6       9       .       .       .       Gene=3;function=def</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">the_frame</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">tag</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;mRNA&#39;</span><span class="p">,</span> <span class="s1">&#39;CDS&#39;</span><span class="p">,</span> <span class="s1">&#39;CDS&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+------------+-------------+------------+</span>
<span class="go">|   Chromosome |     Start |       End | Feature    |      Gene | function   |   the_frame | tag        |</span>
<span class="go">|   (category) |   (int32) |   (int32) | (object)   |   (int64) | (object)   |     (int64) | (object)   |</span>
<span class="go">|--------------+-----------+-----------+------------+-----------+------------+-------------+------------|</span>
<span class="go">|            1 |         1 |         4 | gene       |         1 | a b        |           0 | mRNA       |</span>
<span class="go">|            1 |         3 |         6 | exon       |         2 | c          |           2 | CDS        |</span>
<span class="go">|            1 |         5 |         9 | exon       |         3 | def        |           1 | CDS        |</span>
<span class="go">+--------------+-----------+-----------+------------+-----------+------------+-------------+------------+</span>
<span class="go">Unstranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gff3</span><span class="p">(</span><span class="n">map_cols</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;phase&#39;</span><span class="p">:</span><span class="s1">&#39;the_frame&#39;</span><span class="p">,</span> <span class="s1">&#39;feature&#39;</span><span class="p">:</span><span class="s1">&#39;tag&#39;</span><span class="p">}))</span>
<span class="go">1       .       mRNA    2       4       .       .       0       Gene=1;function=a b</span>
<span class="go">1       .       CDS     4       6       .       .       2       Gene=2;function=c</span>
<span class="go">1       .       CDS     6       9       .       .       1       Gene=3;function=def</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gff3</span><span class="p">(</span><span class="n">map_cols</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;attribute&#39;</span><span class="p">:</span><span class="s1">&#39;Gene&#39;</span><span class="p">}))</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_gtf">
<span class="sig-name descname"><span class="pre">to_gtf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compression</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'infer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_gtf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_gtf" title="Permalink to this definition"></a></dt>
<dd><p>Write to Gene Transfer Format.</p>
<p>The GTF format consists of a tab-separated file without header.
It contains a fixed amount of columns, indicated below (names before “:”).
For each of these, PyRanges will use the corresponding column (names after “:”).</p>
<p><code class="docutils literal notranslate"><span class="pre">seqname:</span> <span class="pre">Chromosome</span>
<span class="pre">source:</span> <span class="pre">Source</span>
<span class="pre">type:</span> <span class="pre">Feature</span>
<span class="pre">start:</span> <span class="pre">Start</span>
<span class="pre">end:</span> <span class="pre">End</span>
<span class="pre">score:</span> <span class="pre">Score</span>
<span class="pre">strand:</span> <span class="pre">Strand</span>
<span class="pre">frame:</span> <span class="pre">Frame</span>
<span class="pre">attribute:</span> <span class="pre">auto-filled</span></code></p>
<p>Columns which are not mapped to GTF columns are appended as a field
in the attribute string (i.e. the last field).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em><em>, </em><em>default None</em><em>, </em><em>i.e. return string representation.</em>) – Where to write file.</p></li>
<li><p><strong>compression</strong> (<em>{‘infer’</em><em>, </em><em>‘gzip’</em><em>, </em><em>‘bz2’</em><em>, </em><em>‘zip’</em><em>, </em><em>‘xz’</em><em>, </em><em>None}</em><em>, </em><em>default &quot;infer&quot;</em>) – Which compression to use. Uses file extension to infer by default.</p></li>
<li><p><strong>chain</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return the PyRanges after writing.</p></li>
<li><p><strong>map_cols</strong> (<em>dict</em><em>, </em><em>default None</em>) – Override mapping between GTF and PyRanges fields for any number of columns.
Format: <code class="docutils literal notranslate"><span class="pre">{gtf_column</span> <span class="pre">:</span> <span class="pre">pyranges_column}</span></code>
If a mapping is found for the “attribute”` column, this not auto-filled</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Nonexisting columns will be added with a ‘.’ to represent the missing values.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.read_gtf" title="pyranges.read_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.read_gtf</span></code></a></dt><dd><p>read GTF files</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.to_gff3</span></code></dt><dd><p>write to GFF3 format</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="s2">&quot;Feature&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gtf</span><span class="p">())</span>
<span class="go">1       .       gene    2       4       .       .       .</span>
<span class="go">1       .       exon    4       6       .       .       .</span>
<span class="go">1       .       exon    6       9       .       .       .</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Tim&quot;</span><span class="p">,</span> <span class="s2">&quot;Eric&quot;</span><span class="p">,</span> <span class="s2">&quot;Endre&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">prices</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Cheap&quot;</span><span class="p">,</span> <span class="s2">&quot;Premium&quot;</span><span class="p">,</span> <span class="s2">&quot;Fine European&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gtf</span><span class="p">())</span>
<span class="go">1       .       gene    2       4       .       .       .       name &quot;Tim&quot;; prices &quot;Cheap&quot;;</span>
<span class="go">1       .       exon    4       6       .       .       .       name &quot;Eric&quot;; prices &quot;Premium&quot;;</span>
<span class="go">1       .       exon    6       9       .       .       .       name &quot;Endre&quot;; prices &quot;Fine European&quot;;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">gr</span><span class="o">.</span><span class="n">to_gtf</span><span class="p">(</span><span class="n">map_cols</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;feature&quot;</span><span class="p">:</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;attribute&quot;</span><span class="p">:</span><span class="s2">&quot;prices&quot;</span><span class="p">}))</span>
<span class="go">1       .       Tim     2       4       .       .       .       prices &quot;Cheap&quot;;</span>
<span class="go">1       .       Eric    4       6       .       .       .       prices &quot;Premium&quot;;</span>
<span class="go">1       .       Endre   6       9       .       .       .       prices &quot;Fine European&quot;;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.to_rle">
<span class="sig-name descname"><span class="pre">to_rle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rpm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nb_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.to_rle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.to_rle" title="Permalink to this definition"></a></dt>
<dd><p>Return as RleDict.</p>
<p>Create collection of Rles representing the coverage or other numerical value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value_col</strong> (<em>str</em><em>, </em><em>default None</em>) – Numerical column to create RleDict from.</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to treat strands serparately.</p></li>
<li><p><strong>rpm</strong> (<em>bool</em><em>, </em><em>default False</em>) – Normalize by multiplying with <cite>1e6/(number_intervals)</cite>.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Rle with coverage or other info from the PyRanges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pyrle.RleDict</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span>     <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="s1">&#39;Score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mf">3.14</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_rle</span><span class="p">()</span>
<span class="go">chr1 +</span>
<span class="go">--</span>
<span class="go">+--------+-----+-----+-----+-----+</span>
<span class="go">| Runs   | 3   | 3   | 2   | 1   |</span>
<span class="go">|--------+-----+-----+-----+-----|</span>
<span class="go">| Values | 0.0 | 1.0 | 0.0 | 1.0 |</span>
<span class="go">+--------+-----+-----+-----+-----+</span>
<span class="go">Rle of length 9 containing 4 elements (avg. length 2.25)</span>

<span class="go">chr1 -</span>
<span class="go">--</span>
<span class="go">+--------+-----+-----+</span>
<span class="go">| Runs   | 5   | 2   |</span>
<span class="go">|--------+-----+-----|</span>
<span class="go">| Values | 0.0 | 1.0 |</span>
<span class="go">+--------+-----+-----+</span>
<span class="go">Rle of length 7 containing 2 elements (avg. length 3.5)</span>
<span class="go">RleDict object with 2 chromosomes/strand pairs.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_rle</span><span class="p">(</span><span class="n">value_col</span><span class="o">=</span><span class="s2">&quot;Score&quot;</span><span class="p">)</span>
<span class="go">chr1 +</span>
<span class="go">--</span>
<span class="go">+--------+-----+-----+-----+-----+</span>
<span class="go">| Runs   | 3   | 3   | 2   | 1   |</span>
<span class="go">|--------+-----+-----+-----+-----|</span>
<span class="go">| Values | 0.0 | 0.1 | 0.0 | 5.0 |</span>
<span class="go">+--------+-----+-----+-----+-----+</span>
<span class="go">Rle of length 9 containing 4 elements (avg. length 2.25)</span>

<span class="go">chr1 -</span>
<span class="go">--</span>
<span class="go">+--------+-----+------+</span>
<span class="go">| Runs   | 5   | 2    |</span>
<span class="go">|--------+-----+------|</span>
<span class="go">| Values | 0.0 | 3.14 |</span>
<span class="go">+--------+-----+------+</span>
<span class="go">Rle of length 7 containing 2 elements (avg. length 3.5)</span>
<span class="go">RleDict object with 2 chromosomes/strand pairs.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_rle</span><span class="p">(</span><span class="n">value_col</span><span class="o">=</span><span class="s2">&quot;Score&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">chr1</span>
<span class="go">+--------+-----+-----+------+------+-----+-----+</span>
<span class="go">| Runs   | 3   | 2   | 1    | 1    | 1   | 1   |</span>
<span class="go">|--------+-----+-----+------+------+-----+-----|</span>
<span class="go">| Values | 0.0 | 0.1 | 3.24 | 3.14 | 0.0 | 5.0 |</span>
<span class="go">+--------+-----+-----+------+------+-----+-----+</span>
<span class="go">Rle of length 9 containing 6 elements (avg. length 1.5)</span>
<span class="go">Unstranded RleDict object with 1 chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">to_rle</span><span class="p">(</span><span class="n">rpm</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">chr1 +</span>
<span class="go">--</span>
<span class="go">+--------+-----+-------------------+-----+-------------------+</span>
<span class="go">| Runs   | 3   | 3                 | 2   | 1                 |</span>
<span class="go">|--------+-----+-------------------+-----+-------------------|</span>
<span class="go">| Values | 0.0 | 333333.3333333333 | 0.0 | 333333.3333333333 |</span>
<span class="go">+--------+-----+-------------------+-----+-------------------+</span>
<span class="go">Rle of length 9 containing 4 elements (avg. length 2.25)</span>

<span class="go">chr1 -</span>
<span class="go">--</span>
<span class="go">+--------+-----+-------------------+</span>
<span class="go">| Runs   | 5   | 2                 |</span>
<span class="go">|--------+-----+-------------------|</span>
<span class="go">| Values | 0.0 | 333333.3333333333 |</span>
<span class="go">+--------+-----+-------------------+</span>
<span class="go">Rle of length 7 containing 2 elements (avg. length 3.5)</span>
<span class="go">RleDict object with 2 chromosomes/strand pairs.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.unstrand">
<span class="sig-name descname"><span class="pre">unstrand</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.unstrand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.unstrand" title="Permalink to this definition"></a></dt>
<dd><p>Remove strand.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Removes Strand column even if PyRanges is not stranded.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#pyranges.pyranges.PyRanges.stranded" title="pyranges.pyranges.PyRanges.stranded"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PyRanges.stranded</span></code></a></dt><dd><p>whether PyRanges contains valid strand info.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span>  <span class="p">{</span><span class="s1">&#39;Chromosome&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">],</span> <span class="s1">&#39;Start&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;End&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="s1">&#39;Strand&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">| Chromosome   |     Start |       End | Strand       |</span>
<span class="go">| (category)   |   (int32) |   (int32) | (category)   |</span>
<span class="go">|--------------+-----------+-----------+--------------|</span>
<span class="go">| chr1         |         1 |         5 | +            |</span>
<span class="go">| chr1         |         6 |         8 | -            |</span>
<span class="go">+--------------+-----------+-----------+--------------+</span>
<span class="go">Stranded PyRanges object has 2 rows and 4 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">unstrand</span><span class="p">()</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">| Chromosome   |     Start |       End |</span>
<span class="go">| (category)   |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">| chr1         |         1 |         5 |</span>
<span class="go">| chr1         |         6 |         8 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.values">
<span class="sig-name descname"><span class="pre">values</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.values" title="Permalink to this definition"></a></dt>
<dd><p>Return the underlying DataFrames.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.window">
<span class="sig-name descname"><span class="pre">window</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.window"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.window" title="Permalink to this definition"></a></dt>
<dd><p>Return overlapping genomic windows.</p>
<p>Windows of length <cite>window_size</cite> are returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>window_size</strong> (<em>int</em>) – Length of the windows.</p></li>
<li><p><strong>strand</strong> (<em>bool</em><em>, </em><em>default None</em><em>, </em><em>i.e. auto</em>) – Whether to do operations on chromosome/strand pairs or chromosomes. If None, will use
chromosome/strand pairs if the PyRanges is stranded.</p></li>
<li><p><strong>nb_cpu</strong> (<em>int</em><em>, </em><em>default 1</em>) – How many cpus to use. Can at most use 1 per chromosome or chromosome/strand tuple.
Will only lead to speedups on large datasets.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to pass as keyword arguments to <cite>f</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tiled PyRanges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#pyranges.pyranges.PyRanges" title="pyranges.pyranges.PyRanges">PyRanges</a></p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="../index.html#pyranges.PyRanges.tile" title="pyranges.PyRanges.tile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pyranges.PyRanges.tile</span></code></a></dt><dd><p>divide intervals into adjacent tiles.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;Start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">895</span><span class="p">],</span> <span class="s2">&quot;End&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1259</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |       895 |      1259 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 1 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">window</span><span class="p">(</span><span class="mi">200</span><span class="p">)</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">|   Chromosome |     Start |       End |</span>
<span class="go">|   (category) |   (int32) |   (int32) |</span>
<span class="go">|--------------+-----------+-----------|</span>
<span class="go">|            1 |       895 |      1095 |</span>
<span class="go">|            1 |      1095 |      1259 |</span>
<span class="go">+--------------+-----------+-----------+</span>
<span class="go">Unstranded PyRanges object has 2 rows and 3 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ensembl_gtf</span><span class="p">()[[</span><span class="s2">&quot;Feature&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="go">| 1            | gene         | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | transcript   | 11868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | exon         | 11868     | 12227     | +            | DDX11L1     |</span>
<span class="go">| 1            | exon         | 12612     | 12721     | +            | DDX11L1     |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="go">| 1            | gene         | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | transcript   | 1173055   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1179364   | 1179555   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1173055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">Stranded PyRanges object has 2,446 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gr</span><span class="o">.</span><span class="n">window</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">| Chromosome   | Feature      | Start     | End       | Strand       | gene_name   |</span>
<span class="go">| (category)   | (category)   | (int32)   | (int32)   | (category)   | (object)    |</span>
<span class="go">|--------------+--------------+-----------+-----------+--------------+-------------|</span>
<span class="go">| 1            | gene         | 11868     | 12868     | +            | DDX11L1     |</span>
<span class="go">| 1            | gene         | 12868     | 13868     | +            | DDX11L1     |</span>
<span class="go">| 1            | gene         | 13868     | 14409     | +            | DDX11L1     |</span>
<span class="go">| 1            | transcript   | 11868     | 12868     | +            | DDX11L1     |</span>
<span class="go">| ...          | ...          | ...       | ...       | ...          | ...         |</span>
<span class="go">| 1            | exon         | 1173055   | 1174055   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1174055   | 1175055   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1175055   | 1176055   | -            | TTLL10-AS1  |</span>
<span class="go">| 1            | exon         | 1176055   | 1176396   | -            | TTLL10-AS1  |</span>
<span class="go">+--------------+--------------+-----------+-----------+--------------+-------------+</span>
<span class="go">Stranded PyRanges object has 7,516 rows and 6 columns from 1 chromosomes.</span>
<span class="go">For printing, the PyRanges was sorted on Chromosome and Strand.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__getstate__">
<span class="sig-name descname"><span class="pre">__getstate__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__getstate__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__getstate__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyranges.pyranges.PyRanges.__setstate__">
<span class="sig-name descname"><span class="pre">__setstate__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/pyranges/pyranges.html#PyRanges.__setstate__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyranges.pyranges.PyRanges.__setstate__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../multithreaded/index.html" class="btn btn-neutral float-left" title="pyranges.multithreaded" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../readers/index.html" class="btn btn-neutral float-right" title="pyranges.readers" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Endre Bakken Stovner and Marco Mariotti.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>